{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook is looking at metadata from SRA. I am wanting to develop a programmatic approach to download metadata information, but needs to be consistent with Zhenxia and Miegs analysis. I am also wanting to figure exactly how Zhenxia and the Miegs did this.\n",
    "\n",
    "There are two programmatic approaches:\n",
    "* SRAdb\n",
    "* Biopython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', 999)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true,
    "heading_collapsed": true
   },
   "source": [
    "# SRAdb\n",
    "\n",
    "SRAdb is nice because it parses everything into a table, but the counts don't match with my expectations. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false,
    "hidden": true
   },
   "outputs": [],
   "source": [
    "import rpy2\n",
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "## Download SQLite database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "hidden": true
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "cd ../../output/\n",
    "if [ ! -e SRAmetadb.sqlite ]; then \n",
    "    wget 'http://dl.dropbox.com/u/51653511/SRAmetadb.sqlite.gz'\n",
    "    gunzip SRAmetadb.sqlite.gz\n",
    "fi"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "## Query database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false,
    "hidden": true
   },
   "outputs": [],
   "source": [
    "%%R \n",
    "library(SRAdb)\n",
    "sra_con = dbConnect(SQLite(), '../../output/SRAmetadb.sqlite')\n",
    "rs = getSRA(search_terms='\"Drosophila melanogaster\"[Organism]', out_types=c('run', 'sample'), sra_con=sra_con)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "## Print number of rows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false,
    "hidden": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([10356,    21], dtype=int32)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%R dim(rs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Biopyton\n",
    "\n",
    "Biopython's counts match what I expect, but the results are in raw XML and need a parser. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from Bio import Entrez\n",
    "Entrez.email = 'justin.fear@nih.gov'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get List of Records"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'22344'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "handle = Entrez.esearch(db='sra', term='\"Drosophila melanogaster\"[Orgn]')\n",
    "records = Entrez.read(handle)\n",
    "records['Count']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get Full XML of Subset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<?xml version=\"1.0\" ?>\n",
      "<EXPERIMENT_PACKAGE_SET>\n",
      "<EXPERIMENT_PACKAGE><EXPERIMENT alias=\"E-MTAB-1976:C1\" accession=\"ERX330955\" broker_name=\"ArrayExpress\"><IDENTIFIERS><PRIMARY_ID>ERX330955</PRIMARY_ID><SUBMITTER_ID namespace=\"UNIVERSITY OF MANCHESTER\">E-MTAB-1976:C1</SUBMITTER_ID><SUBMITTER_ID namespace=\"University of Manchester\">E-MTAB-1976:C1</SUBMITTER_ID></IDENTIFIERS><TITLE>AB SOLiD 4 System sequencing; Deignan</TITLE><STUDY_REF refname=\"E-MTAB-1976\" refcenter=\"University of Manchester\" accession=\"ERP004143\"><IDENTIFIERS><PRIMARY_ID>ERP004143</PRIMARY_ID><SUBMITTER_ID namespace=\"University of Manchester\">E-MTAB-1976</SUBMITTER_ID></IDENTIFIERS></STUDY_REF><DESIGN><DESIGN_DESCRIPTION>Deignan</DESIGN_DESCRIPTION><SAMPLE_DESCRIPTOR refname=\"E-MTAB-1976:C1\" accession=\"ERS362580\" refcenter=\"University of Manchester\"><IDENTIFIERS><PRIMARY_ID>ERS362580</PRIMARY_ID><EXTERNAL_ID namespace=\"BioSample\">SAMEA2225913</EXTERNAL_ID><SUBMITTER_ID namespace=\"University of Manchester\">E-MTAB-1976:C1</SUBMITTER_ID></IDENTIFIERS></SAMPLE_DESCRIPTOR><LIBRARY_DESCRIPTOR><LIBRARY_NAME>C1</LIBRARY_NAME><LIBRARY_STRATEGY>RNA-Seq</LIBRARY_STRATEGY><LIBRARY_SOURCE>TRANSCRIPTOMIC</LIBRARY_SOURCE><LIBRARY_SELECTION>cDNA</LIBRARY_SELECTION><LIBRARY_LAYOUT><SINGLE/></LIBRARY_LAYOUT><LIBRARY_CONSTRUCTION_PROTOCOL>The Decepentaplegic receptor Thickviens (Tkv) was over-expressed using the UASp-GAL4 system. pUASp-tkvQD was obtained from this original stock yw;pUASpTkvQD/TM6B (Casanueva and Ferguson, 2004; Development, 131(9): 1881-90). Gal4-Tub:VP16 was obtained from the original stock, Bloomington 7062. At the 3-3.5h time point after egg lay (AEL) embryos over-expressing TkvQD, derived from Gal4-Tub:VP16/+ ; P{UASp-TkvQD}/+ adults were collected. As an experimental control, 3-3.5h embryos derived from Gal4-Tub:VP16/+ ; + were also collected. RNA was extracted from 30ul of Drosophila embryos per sample using Trizol (Invitrogen) SOLiDTM Total RNA-Seq Kit (PN 4445374)</LIBRARY_CONSTRUCTION_PROTOCOL></LIBRARY_DESCRIPTOR></DESIGN><PLATFORM><ABI_SOLID><INSTRUMENT_MODEL>AB SOLiD 4 System</INSTRUMENT_MODEL></ABI_SOLID></PLATFORM><EXPERIMENT_ATTRIBUTES><EXPERIMENT_ATTRIBUTE><TAG>Experimental Factor: genotype</TAG><VALUE>Gal4-Tub:VP16/+</VALUE></EXPERIMENT_ATTRIBUTE></EXPERIMENT_ATTRIBUTES></EXPERIMENT><SUBMISSION alias=\"E-MTAB-1976\" broker_name=\"ArrayExpress\" lab_name=\"ArrayExpress\" accession=\"ERA259843\"><IDENTIFIERS><PRIMARY_ID>ERA259843</PRIMARY_ID><SUBMITTER_ID namespace=\"UNIVERSITY OF MANCHESTER\">E-MTAB-1976</SUBMITTER_ID><SUBMITTER_ID namespace=\"University of Manchester\">E-MTAB-1976</SUBMITTER_ID></IDENTIFIERS><TITLE>Submitted by UNIVERSITY OF MANCHESTER on 21-OCT-2013</TITLE></SUBMISSION><Organization type=\"center\"><Name abbr=\"University Of Manchester\">University Of Manchester</Name><Contact email=\"john.archer@manchester.ac.uk\"><Name><First>John</First><Last>Archer</Last><Middle/></Name></Contact></Organization><STUDY alias=\"E-MTAB-1976\" accession=\"ERP004143\" center_name=\"UNIVERSITY OF MANCHESTER\" broker_name=\"ArrayExpress\"><IDENTIFIERS><PRIMARY_ID>ERP004143</PRIMARY_ID><EXTERNAL_ID namespace=\"BioProject\">PRJEB4822</EXTERNAL_ID><SUBMITTER_ID namespace=\"UNIVERSITY OF MANCHESTER\">E-MTAB-1976</SUBMITTER_ID><SUBMITTER_ID namespace=\"University of Manchester\">E-MTAB-1976</SUBMITTER_ID></IDENTIFIERS><DESCRIPTOR><STUDY_TITLE>Deignan</STUDY_TITLE><STUDY_TYPE existing_study_type=\"Transcriptome Analysis\"/><STUDY_ABSTRACT>The goal of the RNA-Seq experiment was to identify novel target genes of the Decapentaplegic (Dpp) signalling pathway in the early Drosophila embryo. This was thus achieved by identifying genes that were differentially expressed in response to ectopic Dpp signalling (over-expression of Thickviens (Tkv)) in comparison to the control samples.</STUDY_ABSTRACT><STUDY_DESCRIPTION>The goal of the RNA-Seq experiment was to identify novel target genes of the Decapentaplegic (Dpp) signalling pathway in the early Drosophila embryo. This was thus achieved by identifying genes that were differentially expressed in response to ectopic Dpp signalling (over-expression of Thickviens (Tkv)) in comparison to the control samples.</STUDY_DESCRIPTION></DESCRIPTOR><STUDY_LINKS><STUDY_LINK><URL_LINK><LABEL>E-MTAB-1976 in ArrayExpress</LABEL><URL>http://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-1976</URL></URL_LINK></STUDY_LINK></STUDY_LINKS><STUDY_ATTRIBUTES><STUDY_ATTRIBUTE><TAG>ArrayExpress</TAG><VALUE>E-MTAB-1976</VALUE></STUDY_ATTRIBUTE></STUDY_ATTRIBUTES></STUDY><SAMPLE alias=\"E-MTAB-1976:C1\" accession=\"ERS362580\" center_name=\"UNIVERSITY OF MANCHESTER\" broker_name=\"ArrayExpress\"><IDENTIFIERS><PRIMARY_ID>ERS362580</PRIMARY_ID><EXTERNAL_ID namespace=\"BioSample\">SAMEA2225913</EXTERNAL_ID><SUBMITTER_ID namespace=\"UNIVERSITY OF MANCHESTER\">E-MTAB-1976:C1</SUBMITTER_ID><SUBMITTER_ID namespace=\"University of Manchester\">E-MTAB-1976:C1</SUBMITTER_ID></IDENTIFIERS><TITLE>C1</TITLE><SAMPLE_NAME><TAXON_ID>7227</TAXON_ID><SCIENTIFIC_NAME>Drosophila melanogaster</SCIENTIFIC_NAME><COMMON_NAME>fruit fly</COMMON_NAME></SAMPLE_NAME><DESCRIPTION>Protocols: The Decepentaplegic receptor Thickviens (Tkv) was over-expressed using the UASp-GAL4 system. pUASp-tkvQD was obtained from this original stock yw;pUASpTkvQD/TM6B (Casanueva and Ferguson, 2004; Development, 131(9): 1881-90). Gal4-Tub:VP16 was obtained from the original stock, Bloomington 7062. At the 3-3.5h time point after egg lay (AEL) embryos over-expressing TkvQD, derived from Gal4-Tub:VP16/+ ; P{UASp-TkvQD}/+ adults were collected. As an experimental control, 3-3.5h embryos derived from Gal4-Tub:VP16/+ ; + were also collected. RNA was extracted from 30ul of Drosophila embryos per sample using Trizol (Invitrogen) SOLiDTM Total RNA-Seq Kit (PN 4445374)</DESCRIPTION><SAMPLE_ATTRIBUTES><SAMPLE_ATTRIBUTE><TAG>organism</TAG><VALUE>Drosophila melanogaster</VALUE></SAMPLE_ATTRIBUTE><SAMPLE_ATTRIBUTE><TAG>developmental stage</TAG><VALUE>larval stage</VALUE></SAMPLE_ATTRIBUTE><SAMPLE_ATTRIBUTE><TAG>genotype</TAG><VALUE>Gal4-Tub:VP16/+</VALUE></SAMPLE_ATTRIBUTE><SAMPLE_ATTRIBUTE><TAG>age</TAG><VALUE>3-3.5 hour</VALUE></SAMPLE_ATTRIBUTE></SAMPLE_ATTRIBUTES></SAMPLE><RUN_SET><RUN alias=\"E-MTAB-1976:C1_LD.fastq.gz\" run_center=\"Genomic Technolgies Core Facility, Faculty of Life Sciences, University of Manchester\" accession=\"ERR358181\" broker_name=\"ArrayExpress\" load_done=\"false\" unavailable=\"true\" published=\"2016-06-16 08:04:14\" is_public=\"true\" cluster_name=\"public\" static_data_available=\"1\"><IDENTIFIERS><PRIMARY_ID>ERR358181</PRIMARY_ID><SUBMITTER_ID namespace=\"UNIVERSITY OF MANCHESTER\">E-MTAB-1976:C1_LD.fastq.gz</SUBMITTER_ID><SUBMITTER_ID namespace=\"University of Manchester\">E-MTAB-1976:C1_LD.fastq.gz</SUBMITTER_ID></IDENTIFIERS><TITLE>AB SOLiD 4 System sequencing; Deignan</TITLE><EXPERIMENT_REF refname=\"E-MTAB-1976:C1\" refcenter=\"University of Manchester\" accession=\"ERX330955\"><IDENTIFIERS><PRIMARY_ID>ERX330955</PRIMARY_ID><SUBMITTER_ID namespace=\"University of Manchester\">E-MTAB-1976:C1</SUBMITTER_ID></IDENTIFIERS></EXPERIMENT_REF></RUN></RUN_SET></EXPERIMENT_PACKAGE>\n",
      "</EXPERIMENT_PACKAGE_SET>\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "h = Entrez.efetch(db='sra', id=records['IdList'][:1])\n",
    "r = h.read()\n",
    "\n",
    "print(r)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SRA Website\n",
    "\n",
    "The gold standard is downloading results directly from SRA's website. This is what Zhenxia and the Miegs did. I still want to verify how they made their tables."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(13993, 29)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Assay_Type_s</th>\n",
       "      <th>AssemblyName_s</th>\n",
       "      <th>BioProject_s</th>\n",
       "      <th>BioSample_s</th>\n",
       "      <th>Center_Name_s</th>\n",
       "      <th>Experiment_s</th>\n",
       "      <th>InsertSize_l</th>\n",
       "      <th>LibraryLayout_s</th>\n",
       "      <th>LibrarySelection_s</th>\n",
       "      <th>LibrarySource_s</th>\n",
       "      <th>Library_Name_s</th>\n",
       "      <th>LoadDate_s</th>\n",
       "      <th>MBases_l</th>\n",
       "      <th>MBytes_l</th>\n",
       "      <th>Platform_s</th>\n",
       "      <th>ReleaseDate_s</th>\n",
       "      <th>SRA_Sample_s</th>\n",
       "      <th>SRA_Study_s</th>\n",
       "      <th>Sample_Name_s</th>\n",
       "      <th>Sex_s</th>\n",
       "      <th>genotype_s</th>\n",
       "      <th>source_name_s</th>\n",
       "      <th>strain_s</th>\n",
       "      <th>tissue_s</th>\n",
       "      <th>Consent_s</th>\n",
       "      <th>Organism_s</th>\n",
       "      <th>g1k_analysis_group_s</th>\n",
       "      <th>g1k_pop_code_s</th>\n",
       "      <th>source_s</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Run_s</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR000909</th>\n",
       "      <td>FL-cDNA</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "      <td>PRJNA74619</td>\n",
       "      <td>SAMN00000114</td>\n",
       "      <td>WUGSC</td>\n",
       "      <td>SRX000185</td>\n",
       "      <td>0</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>2305754913</td>\n",
       "      <td>2012-01-19</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>LS454</td>\n",
       "      <td>2008-04-09</td>\n",
       "      <td>SRS000346</td>\n",
       "      <td>SRP000119</td>\n",
       "      <td>SRA000293_sample</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "      <td>public</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR001337</th>\n",
       "      <td>EST</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "      <td>PRJNA107113</td>\n",
       "      <td>SAMN00000196</td>\n",
       "      <td>GEO</td>\n",
       "      <td>SRX000316</td>\n",
       "      <td>0</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>WT_females_beta-eliminated</td>\n",
       "      <td>2014-05-25</td>\n",
       "      <td>54</td>\n",
       "      <td>362</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2008-06-27</td>\n",
       "      <td>SRS000430</td>\n",
       "      <td>SRP000181</td>\n",
       "      <td>GSM278573</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "      <td>public</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR001338</th>\n",
       "      <td>EST</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "      <td>PRJNA107113</td>\n",
       "      <td>SAMN00000208</td>\n",
       "      <td>GEO</td>\n",
       "      <td>SRX000328</td>\n",
       "      <td>0</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>IR_non-beta-eliminated</td>\n",
       "      <td>2014-05-25</td>\n",
       "      <td>115</td>\n",
       "      <td>796</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2008-06-27</td>\n",
       "      <td>SRS000442</td>\n",
       "      <td>SRP000181</td>\n",
       "      <td>GSM278707</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "      <td>public</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR001339</th>\n",
       "      <td>EST</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "      <td>PRJNA107113</td>\n",
       "      <td>SAMN00000207</td>\n",
       "      <td>GEO</td>\n",
       "      <td>SRX000327</td>\n",
       "      <td>0</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>WT_females_non-beta-eliminated</td>\n",
       "      <td>2014-05-25</td>\n",
       "      <td>144</td>\n",
       "      <td>1008</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2008-06-27</td>\n",
       "      <td>SRS000441</td>\n",
       "      <td>SRP000181</td>\n",
       "      <td>GSM278706</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "      <td>public</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR001340</th>\n",
       "      <td>EST</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "      <td>PRJNA107113</td>\n",
       "      <td>SAMN00000198</td>\n",
       "      <td>GEO</td>\n",
       "      <td>SRX000318</td>\n",
       "      <td>0</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>IR_beta-eliminated</td>\n",
       "      <td>2014-05-25</td>\n",
       "      <td>63</td>\n",
       "      <td>425</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2008-06-27</td>\n",
       "      <td>SRS000432</td>\n",
       "      <td>SRP000181</td>\n",
       "      <td>GSM278575</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "      <td>public</td>\n",
       "      <td>Drosophila melanogaster</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "      <td>&lt;not provided&gt;</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Assay_Type_s  AssemblyName_s BioProject_s   BioSample_s  \\\n",
       "Run_s                                                               \n",
       "SRR000909      FL-cDNA  <not provided>   PRJNA74619  SAMN00000114   \n",
       "SRR001337          EST  <not provided>  PRJNA107113  SAMN00000196   \n",
       "SRR001338          EST  <not provided>  PRJNA107113  SAMN00000208   \n",
       "SRR001339          EST  <not provided>  PRJNA107113  SAMN00000207   \n",
       "SRR001340          EST  <not provided>  PRJNA107113  SAMN00000198   \n",
       "\n",
       "          Center_Name_s Experiment_s  InsertSize_l LibraryLayout_s  \\\n",
       "Run_s                                                                \n",
       "SRR000909         WUGSC    SRX000185             0          SINGLE   \n",
       "SRR001337           GEO    SRX000316             0          SINGLE   \n",
       "SRR001338           GEO    SRX000328             0          SINGLE   \n",
       "SRR001339           GEO    SRX000327             0          SINGLE   \n",
       "SRR001340           GEO    SRX000318             0          SINGLE   \n",
       "\n",
       "          LibrarySelection_s LibrarySource_s                  Library_Name_s  \\\n",
       "Run_s                                                                          \n",
       "SRR000909               cDNA  TRANSCRIPTOMIC                      2305754913   \n",
       "SRR001337             RANDOM  TRANSCRIPTOMIC      WT_females_beta-eliminated   \n",
       "SRR001338             RANDOM  TRANSCRIPTOMIC          IR_non-beta-eliminated   \n",
       "SRR001339             RANDOM  TRANSCRIPTOMIC  WT_females_non-beta-eliminated   \n",
       "SRR001340             RANDOM  TRANSCRIPTOMIC              IR_beta-eliminated   \n",
       "\n",
       "           LoadDate_s  MBases_l  MBytes_l Platform_s ReleaseDate_s  \\\n",
       "Run_s                                                                \n",
       "SRR000909  2012-01-19         0         0      LS454    2008-04-09   \n",
       "SRR001337  2014-05-25        54       362   ILLUMINA    2008-06-27   \n",
       "SRR001338  2014-05-25       115       796   ILLUMINA    2008-06-27   \n",
       "SRR001339  2014-05-25       144      1008   ILLUMINA    2008-06-27   \n",
       "SRR001340  2014-05-25        63       425   ILLUMINA    2008-06-27   \n",
       "\n",
       "          SRA_Sample_s SRA_Study_s     Sample_Name_s           Sex_s  \\\n",
       "Run_s                                                                  \n",
       "SRR000909    SRS000346   SRP000119  SRA000293_sample  <not provided>   \n",
       "SRR001337    SRS000430   SRP000181         GSM278573  <not provided>   \n",
       "SRR001338    SRS000442   SRP000181         GSM278707  <not provided>   \n",
       "SRR001339    SRS000441   SRP000181         GSM278706  <not provided>   \n",
       "SRR001340    SRS000432   SRP000181         GSM278575  <not provided>   \n",
       "\n",
       "               genotype_s   source_name_s        strain_s        tissue_s  \\\n",
       "Run_s                                                                       \n",
       "SRR000909  <not provided>  <not provided>  <not provided>  <not provided>   \n",
       "SRR001337  <not provided>  <not provided>  <not provided>  <not provided>   \n",
       "SRR001338  <not provided>  <not provided>  <not provided>  <not provided>   \n",
       "SRR001339  <not provided>  <not provided>  <not provided>  <not provided>   \n",
       "SRR001340  <not provided>  <not provided>  <not provided>  <not provided>   \n",
       "\n",
       "          Consent_s               Organism_s g1k_analysis_group_s  \\\n",
       "Run_s                                                               \n",
       "SRR000909    public  Drosophila melanogaster       <not provided>   \n",
       "SRR001337    public  Drosophila melanogaster       <not provided>   \n",
       "SRR001338    public  Drosophila melanogaster       <not provided>   \n",
       "SRR001339    public  Drosophila melanogaster       <not provided>   \n",
       "SRR001340    public  Drosophila melanogaster       <not provided>   \n",
       "\n",
       "           g1k_pop_code_s        source_s  \n",
       "Run_s                                      \n",
       "SRR000909  <not provided>  <not provided>  \n",
       "SRR001337  <not provided>  <not provided>  \n",
       "SRR001338  <not provided>  <not provided>  \n",
       "SRR001339  <not provided>  <not provided>  \n",
       "SRR001340  <not provided>  <not provided>  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# see ../../data/SRA/README for info about where this file came from\n",
    "dfSRA = pd.read_csv('../../data/SRA/SraRunTable.txt', sep='\\t')\n",
    "dfSRA.set_index('Run_s', inplace=True)\n",
    "print(dfSRA.shape)\n",
    "dfSRA.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Assay_Type_s', 'AssemblyName_s', 'BioProject_s', 'BioSample_s',\n",
       "       'Center_Name_s', 'Experiment_s', 'InsertSize_l', 'LibraryLayout_s',\n",
       "       'LibrarySelection_s', 'LibrarySource_s', 'Library_Name_s', 'LoadDate_s',\n",
       "       'MBases_l', 'MBytes_l', 'Platform_s', 'ReleaseDate_s', 'Run_s',\n",
       "       'SRA_Sample_s', 'SRA_Study_s', 'Sample_Name_s', 'Sex_s', 'genotype_s',\n",
       "       'source_name_s', 'strain_s', 'tissue_s', 'Consent_s', 'Organism_s',\n",
       "       'g1k_analysis_group_s', 'g1k_pop_code_s', 'source_s'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfSRA.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['<not provided>', 'whole organism', 'heads', 'head', 'testis',\n",
       "       'ovary', 'whole fly', 'testes', 'ovaries', 'mixed-stage embryos',\n",
       "       'female heads', 'embryo-derived cell-line', 'Adult ovaries',\n",
       "       'Male body', 'CNS-derived cell-line', 'dorsal mesothoracic disc',\n",
       "       'antenna disc-derived cell-line', 'ventral prothoracic disc',\n",
       "       'Female heads', 'fat body', 'midgut', 'Ovary', 'Whole flies',\n",
       "       'brain', 'whole flies', 'fly', 'Immortalized cells',\n",
       "       'Adult fly heads', 'salivary gland', 'male whole body',\n",
       "       'female whole body', 'central nervous system', 'whole brains',\n",
       "       'mushroom bodies', 'Head', 'Digestive System', 'Accessory Gland',\n",
       "       'Carcass', 'Imaginal Disc', 'Central Nervous System', 'Fat',\n",
       "       'Ovaries', 'Testes', 'Salivary Gland', 'neuron', 'muscle',\n",
       "       'whole-embryo', 'eggs', 'whole body without testis', 'whole body',\n",
       "       'Brain', 'dissected testis', 'ovary stem cells',\n",
       "       'embryonic/larval hemocyte', 'Heads', 'Wing imaginal discs',\n",
       "       'eye-antennal imaginal disc', 'wing imaginal disc', 'embryo',\n",
       "       'abdomen', 'late pupae heads', 'Wing imaginal disc',\n",
       "       'wing imaginal discs', 'eye imaginal discs', 'head & thorax',\n",
       "       '24hr embryo', 'Ovarian Somatic Cells (OSC)', 'fly ovaries',\n",
       "       'Testis', 'WholeBody', 'whole 0-2 hour embryo', 'whole larvae',\n",
       "       'whole embryo', 'T3 leg disc', 'haltere disc', 'wing disc',\n",
       "       'T3 pharate leg', 'pharate haltere', 'pharate wing', 'intestine',\n",
       "       'whole first-instar larvae', 'whole_animal', 'gut', 'S2 Cells',\n",
       "       'Whole body', 'fly Heads', 'S2 cells', 'Salivary glands',\n",
       "       'Body (head removed)', 'retina',\n",
       "       'brain (ventral ganglion w/optic lobes)',\n",
       "       'brain (w/antennal lobes)', 'cell_line-1', 'not applicable',\n",
       "       'fly heads', 'wing discs', 'eye-antennal discs',\n",
       "       '3rd instar larvae wing discs', 'eye-antennal disc',\n",
       "       'Ovary Germarium to stage 7', 'Ovary stage 9 to 10',\n",
       "       'Ovary stage 11 + Egg chambers', 'virgin female carcass',\n",
       "       'virgin female ovary', 'virgin male carcass', 'virgin testes',\n",
       "       'larval brain', 'wing', 'FACS Sorted midgut cells', 'Whole animal',\n",
       "       'Whole Body', 'ovary follicle cells', 'Spermatozoa', 'heart',\n",
       "       'Periphery (entire body except head)', 'whole animal',\n",
       "       'S2 cells (late embryo cell line)', 'Full heads',\n",
       "       'eye/antennal imaginal discs', 'thoracic muscle', 'follicle',\n",
       "       'Whole organism', 'whole', 'cell line', 'Whole embryo',\n",
       "       'unfertilized egg', 'optical lobe', 'Not Applicable',\n",
       "       'adult midgut', 'Fly Heads', 'multi-dendritic sensory neurons',\n",
       "       'whole-animal', 'Terminal Organ Ganglia', 'Dorsal Organ Ganglia',\n",
       "       'dissected midgut (30x)', 'embryonic',\n",
       "       'brain without ventral nerve cord', 'fat bodies', 'imaginal disc',\n",
       "       'antennae', 'whole embryos', 'prothoracic leg disc',\n",
       "       'circulatory system', 'antennal disc', 'mixed imaginal discs',\n",
       "       'whole dechorionated embryos', 'anterior ovariole',\n",
       "       'thorax plus abdomen', 'whole larvae (n \\x98 30)',\n",
       "       'whole brown pre-pupae (n \\x98 30)'], dtype=object)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# List of tissue types, note that there is some cleaning that needs done\n",
    "dfSRA.tissue_s.unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Zhen-Xia's List"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(12719, 57)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BioSample</th>\n",
       "      <th>ReleaseDate</th>\n",
       "      <th>LoadDate</th>\n",
       "      <th>avgLength</th>\n",
       "      <th>size_MB</th>\n",
       "      <th>Experiment</th>\n",
       "      <th>LibraryName</th>\n",
       "      <th>LibraryStrategy</th>\n",
       "      <th>LibrarySelection</th>\n",
       "      <th>LibrarySource</th>\n",
       "      <th>LibraryLayout</th>\n",
       "      <th>Platform</th>\n",
       "      <th>Model</th>\n",
       "      <th>SRAStudy</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>Sample</th>\n",
       "      <th>SampleName</th>\n",
       "      <th>CenterName</th>\n",
       "      <th>Submission</th>\n",
       "      <th>sex</th>\n",
       "      <th>dev_stage</th>\n",
       "      <th>tissue</th>\n",
       "      <th>genotype</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>ExperimentTitle</th>\n",
       "      <th>StudyTitle</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>Total_reads</th>\n",
       "      <th>Unmapped_reads</th>\n",
       "      <th>non.unique</th>\n",
       "      <th>unique</th>\n",
       "      <th>Non.splice_reads</th>\n",
       "      <th>Splice_reads</th>\n",
       "      <th>ambiguous</th>\n",
       "      <th>no_feature</th>\n",
       "      <th>reads_mapped2genes</th>\n",
       "      <th>reads_mapped2intergenic</th>\n",
       "      <th>reads_mapped2ERCC</th>\n",
       "      <th>mRNA</th>\n",
       "      <th>ncRNA</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>rRNA</th>\n",
       "      <th>snoRNA</th>\n",
       "      <th>snRNA</th>\n",
       "      <th>tRNA</th>\n",
       "      <th>exp_genes</th>\n",
       "      <th>layout</th>\n",
       "      <th>failed</th>\n",
       "      <th>F</th>\n",
       "      <th>R</th>\n",
       "      <th>TIN_median</th>\n",
       "      <th>median_coverage_1.20</th>\n",
       "      <th>median_coverage_21.40</th>\n",
       "      <th>median_coverage_41.60</th>\n",
       "      <th>median_coverage_61.80</th>\n",
       "      <th>median_coverage_81.100</th>\n",
       "      <th>correlation</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DRR001177</th>\n",
       "      <td>SAMD00016337</td>\n",
       "      <td>2011-12-11</td>\n",
       "      <td>2014-05-31</td>\n",
       "      <td>36</td>\n",
       "      <td>832</td>\n",
       "      <td>DRX000774</td>\n",
       "      <td>SRig10098</td>\n",
       "      <td>OTHER</td>\n",
       "      <td>size fractionation</td>\n",
       "      <td>OTHER</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina Genome Analyzer IIx</td>\n",
       "      <td>DRP000426</td>\n",
       "      <td>PRJDB2780</td>\n",
       "      <td>DRS000734</td>\n",
       "      <td>SAMD00016337</td>\n",
       "      <td>RIKEN_OSC</td>\n",
       "      <td>DRA000418</td>\n",
       "      <td>male</td>\n",
       "      <td>embryo</td>\n",
       "      <td>whole_organism</td>\n",
       "      <td>AGO2- and negative control-IP short RNA-seq in...</td>\n",
       "      <td>S2</td>\n",
       "      <td>SRig10098</td>\n",
       "      <td>Chromatin associated RNAi components take part...</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>38551396</td>\n",
       "      <td>24102998</td>\n",
       "      <td>11711140</td>\n",
       "      <td>2737258</td>\n",
       "      <td>2737191</td>\n",
       "      <td>67</td>\n",
       "      <td>63300</td>\n",
       "      <td>542516</td>\n",
       "      <td>1687592</td>\n",
       "      <td>443596</td>\n",
       "      <td>254</td>\n",
       "      <td>432466</td>\n",
       "      <td>1182063</td>\n",
       "      <td>54793</td>\n",
       "      <td>5219</td>\n",
       "      <td>9665</td>\n",
       "      <td>165</td>\n",
       "      <td>3221</td>\n",
       "      <td>13691</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>0.1378</td>\n",
       "      <td>0.4379</td>\n",
       "      <td>0.4243</td>\n",
       "      <td>6.250000</td>\n",
       "      <td>0.080266</td>\n",
       "      <td>0.079628</td>\n",
       "      <td>0.216993</td>\n",
       "      <td>0.341371</td>\n",
       "      <td>0.604689</td>\n",
       "      <td>0.693731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRR001178</th>\n",
       "      <td>SAMD00016336</td>\n",
       "      <td>2011-12-11</td>\n",
       "      <td>2014-05-31</td>\n",
       "      <td>36</td>\n",
       "      <td>791</td>\n",
       "      <td>DRX000775</td>\n",
       "      <td>SRig10099</td>\n",
       "      <td>OTHER</td>\n",
       "      <td>size fractionation</td>\n",
       "      <td>OTHER</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina Genome Analyzer IIx</td>\n",
       "      <td>DRP000426</td>\n",
       "      <td>PRJDB2780</td>\n",
       "      <td>DRS000735</td>\n",
       "      <td>SAMD00016336</td>\n",
       "      <td>RIKEN_OSC</td>\n",
       "      <td>DRA000418</td>\n",
       "      <td>male</td>\n",
       "      <td>embryo</td>\n",
       "      <td>whole_organism</td>\n",
       "      <td>AGO2- and negative control-IP short RNA-seq in...</td>\n",
       "      <td>S2</td>\n",
       "      <td>SRig10099</td>\n",
       "      <td>Chromatin associated RNAi components take part...</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>36613554</td>\n",
       "      <td>24010803</td>\n",
       "      <td>9923336</td>\n",
       "      <td>2679415</td>\n",
       "      <td>2679391</td>\n",
       "      <td>24</td>\n",
       "      <td>98459</td>\n",
       "      <td>515561</td>\n",
       "      <td>1643139</td>\n",
       "      <td>422006</td>\n",
       "      <td>250</td>\n",
       "      <td>657744</td>\n",
       "      <td>889254</td>\n",
       "      <td>56460</td>\n",
       "      <td>4831</td>\n",
       "      <td>30900</td>\n",
       "      <td>509</td>\n",
       "      <td>3441</td>\n",
       "      <td>13634</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>0.1688</td>\n",
       "      <td>0.4294</td>\n",
       "      <td>0.4018</td>\n",
       "      <td>7.945317</td>\n",
       "      <td>0.064579</td>\n",
       "      <td>0.081030</td>\n",
       "      <td>0.143947</td>\n",
       "      <td>0.199425</td>\n",
       "      <td>0.726678</td>\n",
       "      <td>0.693217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRR014222</th>\n",
       "      <td>SAMD00012584</td>\n",
       "      <td>2015-03-23</td>\n",
       "      <td>2015-03-23</td>\n",
       "      <td>101</td>\n",
       "      <td>10621</td>\n",
       "      <td>DRX012753</td>\n",
       "      <td>PCR2085</td>\n",
       "      <td>EST</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>DRP001250</td>\n",
       "      <td>PRJDB2557</td>\n",
       "      <td>DRS012553</td>\n",
       "      <td>DRS012553</td>\n",
       "      <td>ASGRC-WM</td>\n",
       "      <td>DRA001188</td>\n",
       "      <td>unknown</td>\n",
       "      <td>adult</td>\n",
       "      <td>whole_organism</td>\n",
       "      <td>Clontech Catalog number: 636222 Lot number: 10...</td>\n",
       "      <td>mix</td>\n",
       "      <td>SMART method</td>\n",
       "      <td>mRNA 5 end sequencing</td>\n",
       "      <td>whole_organism</td>\n",
       "      <td>164160619</td>\n",
       "      <td>66074703</td>\n",
       "      <td>3403887</td>\n",
       "      <td>94682029</td>\n",
       "      <td>73561370</td>\n",
       "      <td>21120659</td>\n",
       "      <td>8827565</td>\n",
       "      <td>279985</td>\n",
       "      <td>85422641</td>\n",
       "      <td>151838</td>\n",
       "      <td>0</td>\n",
       "      <td>64520581</td>\n",
       "      <td>3803807</td>\n",
       "      <td>14974</td>\n",
       "      <td>17058229</td>\n",
       "      <td>9806</td>\n",
       "      <td>13724</td>\n",
       "      <td>1520</td>\n",
       "      <td>15046</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>0.0103</td>\n",
       "      <td>0.9856</td>\n",
       "      <td>0.0041</td>\n",
       "      <td>33.202892</td>\n",
       "      <td>0.513654</td>\n",
       "      <td>0.034357</td>\n",
       "      <td>0.005381</td>\n",
       "      <td>0.068175</td>\n",
       "      <td>0.212674</td>\n",
       "      <td>0.798120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRR014223</th>\n",
       "      <td>SAMD00012584</td>\n",
       "      <td>2015-03-23</td>\n",
       "      <td>2015-03-23</td>\n",
       "      <td>101</td>\n",
       "      <td>10314</td>\n",
       "      <td>DRX012754</td>\n",
       "      <td>PCR2072</td>\n",
       "      <td>EST</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>DRP001250</td>\n",
       "      <td>PRJDB2557</td>\n",
       "      <td>DRS012553</td>\n",
       "      <td>DRS012553</td>\n",
       "      <td>ASGRC-WM</td>\n",
       "      <td>DRA001188</td>\n",
       "      <td>unknown</td>\n",
       "      <td>adult</td>\n",
       "      <td>whole_organism</td>\n",
       "      <td>Clontech Catalog number: 636222 Lot number: 10...</td>\n",
       "      <td>mix</td>\n",
       "      <td>Ligation method</td>\n",
       "      <td>mRNA 5 end sequencing</td>\n",
       "      <td>whole_organism</td>\n",
       "      <td>151593644</td>\n",
       "      <td>31527578</td>\n",
       "      <td>10224830</td>\n",
       "      <td>109841236</td>\n",
       "      <td>38175966</td>\n",
       "      <td>71665270</td>\n",
       "      <td>6497407</td>\n",
       "      <td>181333</td>\n",
       "      <td>103064996</td>\n",
       "      <td>97500</td>\n",
       "      <td>0</td>\n",
       "      <td>102778950</td>\n",
       "      <td>252968</td>\n",
       "      <td>5790</td>\n",
       "      <td>15696</td>\n",
       "      <td>1402</td>\n",
       "      <td>10024</td>\n",
       "      <td>166</td>\n",
       "      <td>14014</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>0.1035</td>\n",
       "      <td>0.8922</td>\n",
       "      <td>0.0043</td>\n",
       "      <td>17.325735</td>\n",
       "      <td>0.663610</td>\n",
       "      <td>0.096432</td>\n",
       "      <td>0.011929</td>\n",
       "      <td>0.004769</td>\n",
       "      <td>0.032353</td>\n",
       "      <td>0.657597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRR014224</th>\n",
       "      <td>SAMD00012584</td>\n",
       "      <td>2015-03-23</td>\n",
       "      <td>2015-03-23</td>\n",
       "      <td>126</td>\n",
       "      <td>12218</td>\n",
       "      <td>DRX012755</td>\n",
       "      <td>PCR2123</td>\n",
       "      <td>EST</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>DRP001250</td>\n",
       "      <td>PRJDB2557</td>\n",
       "      <td>DRS012553</td>\n",
       "      <td>DRS012553</td>\n",
       "      <td>ASGRC-WM</td>\n",
       "      <td>DRA001188</td>\n",
       "      <td>unknown</td>\n",
       "      <td>adult</td>\n",
       "      <td>whole_organism</td>\n",
       "      <td>Clontech Catalog number: 636222 Lot number: 10...</td>\n",
       "      <td>mix</td>\n",
       "      <td>CapSMART and Non-CapSMART method</td>\n",
       "      <td>mRNA 5 end sequencing</td>\n",
       "      <td>whole_organism</td>\n",
       "      <td>150847202</td>\n",
       "      <td>63307153</td>\n",
       "      <td>3283707</td>\n",
       "      <td>84256342</td>\n",
       "      <td>61657329</td>\n",
       "      <td>22599013</td>\n",
       "      <td>7208452</td>\n",
       "      <td>163524</td>\n",
       "      <td>76794434</td>\n",
       "      <td>89932</td>\n",
       "      <td>0</td>\n",
       "      <td>50509969</td>\n",
       "      <td>9533770</td>\n",
       "      <td>8403</td>\n",
       "      <td>16719380</td>\n",
       "      <td>6275</td>\n",
       "      <td>16301</td>\n",
       "      <td>336</td>\n",
       "      <td>14664</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>0.0066</td>\n",
       "      <td>0.9926</td>\n",
       "      <td>0.0009</td>\n",
       "      <td>34.692736</td>\n",
       "      <td>0.645220</td>\n",
       "      <td>0.088640</td>\n",
       "      <td>0.004934</td>\n",
       "      <td>0.095900</td>\n",
       "      <td>0.309740</td>\n",
       "      <td>0.785452</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              BioSample ReleaseDate    LoadDate  avgLength  size_MB  \\\n",
       "SRR                                                                   \n",
       "DRR001177  SAMD00016337  2011-12-11  2014-05-31         36      832   \n",
       "DRR001178  SAMD00016336  2011-12-11  2014-05-31         36      791   \n",
       "DRR014222  SAMD00012584  2015-03-23  2015-03-23        101    10621   \n",
       "DRR014223  SAMD00012584  2015-03-23  2015-03-23        101    10314   \n",
       "DRR014224  SAMD00012584  2015-03-23  2015-03-23        126    12218   \n",
       "\n",
       "          Experiment LibraryName LibraryStrategy    LibrarySelection  \\\n",
       "SRR                                                                    \n",
       "DRR001177  DRX000774   SRig10098           OTHER  size fractionation   \n",
       "DRR001178  DRX000775   SRig10099           OTHER  size fractionation   \n",
       "DRR014222  DRX012753     PCR2085             EST                cDNA   \n",
       "DRR014223  DRX012754     PCR2072             EST                cDNA   \n",
       "DRR014224  DRX012755     PCR2123             EST                cDNA   \n",
       "\n",
       "            LibrarySource LibraryLayout  Platform  \\\n",
       "SRR                                                 \n",
       "DRR001177           OTHER        SINGLE  ILLUMINA   \n",
       "DRR001178           OTHER        SINGLE  ILLUMINA   \n",
       "DRR014222  TRANSCRIPTOMIC        SINGLE  ILLUMINA   \n",
       "DRR014223  TRANSCRIPTOMIC        SINGLE  ILLUMINA   \n",
       "DRR014224  TRANSCRIPTOMIC        SINGLE  ILLUMINA   \n",
       "\n",
       "                                  Model   SRAStudy BioProject     Sample  \\\n",
       "SRR                                                                        \n",
       "DRR001177  Illumina Genome Analyzer IIx  DRP000426  PRJDB2780  DRS000734   \n",
       "DRR001178  Illumina Genome Analyzer IIx  DRP000426  PRJDB2780  DRS000735   \n",
       "DRR014222           Illumina HiSeq 2500  DRP001250  PRJDB2557  DRS012553   \n",
       "DRR014223           Illumina HiSeq 2500  DRP001250  PRJDB2557  DRS012553   \n",
       "DRR014224           Illumina HiSeq 2500  DRP001250  PRJDB2557  DRS012553   \n",
       "\n",
       "             SampleName CenterName Submission      sex dev_stage  \\\n",
       "SRR                                                                \n",
       "DRR001177  SAMD00016337  RIKEN_OSC  DRA000418     male    embryo   \n",
       "DRR001178  SAMD00016336  RIKEN_OSC  DRA000418     male    embryo   \n",
       "DRR014222     DRS012553   ASGRC-WM  DRA001188  unknown     adult   \n",
       "DRR014223     DRS012553   ASGRC-WM  DRA001188  unknown     adult   \n",
       "DRR014224     DRS012553   ASGRC-WM  DRA001188  unknown     adult   \n",
       "\n",
       "                   tissue                                           genotype  \\\n",
       "SRR                                                                            \n",
       "DRR001177  whole_organism  AGO2- and negative control-IP short RNA-seq in...   \n",
       "DRR001178  whole_organism  AGO2- and negative control-IP short RNA-seq in...   \n",
       "DRR014222  whole_organism  Clontech Catalog number: 636222 Lot number: 10...   \n",
       "DRR014223  whole_organism  Clontech Catalog number: 636222 Lot number: 10...   \n",
       "DRR014224  whole_organism  Clontech Catalog number: 636222 Lot number: 10...   \n",
       "\n",
       "          cell_type                   ExperimentTitle  \\\n",
       "SRR                                                     \n",
       "DRR001177        S2                         SRig10098   \n",
       "DRR001178        S2                         SRig10099   \n",
       "DRR014222       mix                      SMART method   \n",
       "DRR014223       mix                   Ligation method   \n",
       "DRR014224       mix  CapSMART and Non-CapSMART method   \n",
       "\n",
       "                                                  StudyTitle     sample_type  \\\n",
       "SRR                                                                            \n",
       "DRR001177  Chromatin associated RNAi components take part...       cell_line   \n",
       "DRR001178  Chromatin associated RNAi components take part...       cell_line   \n",
       "DRR014222                              mRNA 5 end sequencing  whole_organism   \n",
       "DRR014223                              mRNA 5 end sequencing  whole_organism   \n",
       "DRR014224                              mRNA 5 end sequencing  whole_organism   \n",
       "\n",
       "           Total_reads  Unmapped_reads  non.unique     unique  \\\n",
       "SRR                                                             \n",
       "DRR001177     38551396        24102998    11711140    2737258   \n",
       "DRR001178     36613554        24010803     9923336    2679415   \n",
       "DRR014222    164160619        66074703     3403887   94682029   \n",
       "DRR014223    151593644        31527578    10224830  109841236   \n",
       "DRR014224    150847202        63307153     3283707   84256342   \n",
       "\n",
       "           Non.splice_reads  Splice_reads  ambiguous  no_feature  \\\n",
       "SRR                                                                \n",
       "DRR001177           2737191            67      63300      542516   \n",
       "DRR001178           2679391            24      98459      515561   \n",
       "DRR014222          73561370      21120659    8827565      279985   \n",
       "DRR014223          38175966      71665270    6497407      181333   \n",
       "DRR014224          61657329      22599013    7208452      163524   \n",
       "\n",
       "           reads_mapped2genes  reads_mapped2intergenic  reads_mapped2ERCC  \\\n",
       "SRR                                                                         \n",
       "DRR001177             1687592                   443596                254   \n",
       "DRR001178             1643139                   422006                250   \n",
       "DRR014222            85422641                   151838                  0   \n",
       "DRR014223           103064996                    97500                  0   \n",
       "DRR014224            76794434                    89932                  0   \n",
       "\n",
       "                mRNA    ncRNA  pseudogene      rRNA  snoRNA  snRNA  tRNA  \\\n",
       "SRR                                                                        \n",
       "DRR001177     432466  1182063       54793      5219    9665    165  3221   \n",
       "DRR001178     657744   889254       56460      4831   30900    509  3441   \n",
       "DRR014222   64520581  3803807       14974  17058229    9806  13724  1520   \n",
       "DRR014223  102778950   252968        5790     15696    1402  10024   166   \n",
       "DRR014224   50509969  9533770        8403  16719380    6275  16301   336   \n",
       "\n",
       "           exp_genes  layout  failed       F       R  TIN_median  \\\n",
       "SRR                                                                \n",
       "DRR001177      13691  SINGLE  0.1378  0.4379  0.4243    6.250000   \n",
       "DRR001178      13634  SINGLE  0.1688  0.4294  0.4018    7.945317   \n",
       "DRR014222      15046  SINGLE  0.0103  0.9856  0.0041   33.202892   \n",
       "DRR014223      14014  SINGLE  0.1035  0.8922  0.0043   17.325735   \n",
       "DRR014224      14664  SINGLE  0.0066  0.9926  0.0009   34.692736   \n",
       "\n",
       "           median_coverage_1.20  median_coverage_21.40  median_coverage_41.60  \\\n",
       "SRR                                                                             \n",
       "DRR001177              0.080266               0.079628               0.216993   \n",
       "DRR001178              0.064579               0.081030               0.143947   \n",
       "DRR014222              0.513654               0.034357               0.005381   \n",
       "DRR014223              0.663610               0.096432               0.011929   \n",
       "DRR014224              0.645220               0.088640               0.004934   \n",
       "\n",
       "           median_coverage_61.80  median_coverage_81.100  correlation  \n",
       "SRR                                                                    \n",
       "DRR001177               0.341371                0.604689     0.693731  \n",
       "DRR001178               0.199425                0.726678     0.693217  \n",
       "DRR014222               0.068175                0.212674     0.798120  \n",
       "DRR014223               0.004769                0.032353     0.657597  \n",
       "DRR014224               0.095900                0.309740     0.785452  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Zhenxia gave me this dataset, it is names `.xls` but it is really a `tsv`.\n",
    "dfZ = pd.read_table('../../data/zhenxia/metadata_quality_control.xls')\n",
    "dfZ.set_index('SRR', inplace=True)\n",
    "print(dfZ.shape)\n",
    "dfZ.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "# Jean and Danielle's"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": true,
    "hidden": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(9368, 291)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 1</th>\n",
       "      <th>Genes significantly expressed</th>\n",
       "      <th>% fragments on strand plus of genes</th>\n",
       "      <th>Unnamed: 4</th>\n",
       "      <th>RunId</th>\n",
       "      <th>Run accession in SRA</th>\n",
       "      <th>Title</th>\n",
       "      <th>Other title</th>\n",
       "      <th>Sorting_title 1</th>\n",
       "      <th>Sorting title 2</th>\n",
       "      <th>Sample (summarized from biosample or manual)</th>\n",
       "      <th>Stage and experimental summary</th>\n",
       "      <th>System or Tissue</th>\n",
       "      <th>Laboratory, Author when available, submission date</th>\n",
       "      <th>Species</th>\n",
       "      <th>Project</th>\n",
       "      <th>Project description</th>\n",
       "      <th>Reference</th>\n",
       "      <th>Unnamed: 19</th>\n",
       "      <th>Sequencing protocol</th>\n",
       "      <th>Design</th>\n",
       "      <th>Sequencing platform and Machine type</th>\n",
       "      <th>Experiment type</th>\n",
       "      <th>Sequencing date</th>\n",
       "      <th>Submission date</th>\n",
       "      <th>Library preparation</th>\n",
       "      <th>RNA Integrity Number</th>\n",
       "      <th>Details</th>\n",
       "      <th>Corresponding microarray</th>\n",
       "      <th>Spots, bases, average read length (in SRA)</th>\n",
       "      <th>Single or Paired end</th>\n",
       "      <th>Unnamed: 32</th>\n",
       "      <th>Average read length (nt)</th>\n",
       "      <th>Fragment multiplicity</th>\n",
       "      <th>Million distinct sequences</th>\n",
       "      <th>Million raw reads</th>\n",
       "      <th>Megabases sequenced</th>\n",
       "      <th>%A</th>\n",
       "      <th>%T</th>\n",
       "      <th>%G</th>\n",
       "      <th>%C</th>\n",
       "      <th>%N</th>\n",
       "      <th>%GC</th>\n",
       "      <th>Unnamed: 44</th>\n",
       "      <th>Exit adaptor of read 1</th>\n",
       "      <th>Exit adaptor of read 2</th>\n",
       "      <th>PolyA index : fragments per million with at least 3 A13 motifs shifted by at least 6 bases</th>\n",
       "      <th>Telomere index : fragments per million with at least 3 TTAGGG/CCCTAA motifs</th>\n",
       "      <th>Telomere index allowing G/C substitution</th>\n",
       "      <th>Noise in telomere index : FPM with at least 3 AATCCC/GGGATT motifs</th>\n",
       "      <th>Unnamed: 51</th>\n",
       "      <th>Million reads aligned on all targets per CPU hour</th>\n",
       "      <th>Maximum RAM used (GB)</th>\n",
       "      <th>Number of blocks</th>\n",
       "      <th>Unnamed: 55</th>\n",
       "      <th>Million reads aligned on any target</th>\n",
       "      <th>Average length aligned per read (nt)</th>\n",
       "      <th>Mb aligned on any target</th>\n",
       "      <th>%Reads aligned on any target</th>\n",
       "      <th>%Mb aligned on any target</th>\n",
       "      <th>% length aligned on average</th>\n",
       "      <th>Unnamed: 62</th>\n",
       "      <th>Raw reads</th>\n",
       "      <th>Well mapped reads</th>\n",
       "      <th>Reads mapping uniquely  to a single locus (genomic site) and maximum 1 gene</th>\n",
       "      <th>Reads mapping uniquely to a single locus, but to 2 genes in antisense</th>\n",
       "      <th>Reads mapping uniquely to a single locus</th>\n",
       "      <th>Reads mapping to 2 to 9 sites</th>\n",
       "      <th>Reads rejected because they map to 10 sites or more</th>\n",
       "      <th>Reads rejected by alignment quality filter</th>\n",
       "      <th>Reads rejected before alignment because of low entropy (&lt;= 16 bp)</th>\n",
       "      <th>Reads with no insert (adaptor only)</th>\n",
       "      <th>Reads unaligned from half aligned fragments</th>\n",
       "      <th>Reads of good quality from unaligned fragments (missing genome, microbiome or mapping difficulty)</th>\n",
       "      <th>% well mapped reads</th>\n",
       "      <th>% reads mapping uniquely  to a single locus (genomic site) and maximum 1 gene</th>\n",
       "      <th>% reads mapping uniquely to a single locus, but to 2 genes in antisense</th>\n",
       "      <th>% reads mapping uniquely to a single locus</th>\n",
       "      <th>% reads mapping to 2 to 9 sites</th>\n",
       "      <th>% reads rejected because they map to 10 sites or more</th>\n",
       "      <th>% reads rejected by alignment quality filter</th>\n",
       "      <th>% reads rejected before alignment because of low entropy (&lt;= 16 bp)</th>\n",
       "      <th>% reads with no insert (adaptor only)</th>\n",
       "      <th>% reads unaligned from half aligned fragments</th>\n",
       "      <th>% reads of good quality from unaligned fragments (missing genome, microbiome or mapping difficulty)</th>\n",
       "      <th>Unnamed: 86</th>\n",
       "      <th>Reads aligned in PhiX DNA spike-in</th>\n",
       "      <th>Reads aligned in ERCC RNA spike-in</th>\n",
       "      <th>Reads aligned in ribosomal RNA</th>\n",
       "      <th>Reads aligned in mitochondria</th>\n",
       "      <th>Reads aligned in other small genes</th>\n",
       "      <th>Reads aligned in AceView</th>\n",
       "      <th>Reads aligned in RefSeq</th>\n",
       "      <th>Reads aligned in Encode</th>\n",
       "      <th>Reads aligned in Magic</th>\n",
       "      <th>Reads aligned in genome</th>\n",
       "      <th>Reads aligned in any previously annotated transcripts</th>\n",
       "      <th>Reads inside introns, new exons and intergenic</th>\n",
       "      <th>Reads aligned in microbiome</th>\n",
       "      <th>Reads aligned in viruses</th>\n",
       "      <th>Reads aligned on the imaginary genome (mapping specificity control)</th>\n",
       "      <th>Reads supporting known exons-junctions</th>\n",
       "      <th>Unnamed: 103</th>\n",
       "      <th>% Reads aligned in PhiX DNA spike-in</th>\n",
       "      <th>% Reads aligned in ERCC RNA spike-in</th>\n",
       "      <th>% Reads aligned in ribosomal RNA</th>\n",
       "      <th>% Reads aligned in mitochondria</th>\n",
       "      <th>% Reads aligned in other small genes</th>\n",
       "      <th>% Reads aligned in AceView</th>\n",
       "      <th>% Reads aligned in RefSeq</th>\n",
       "      <th>% Reads aligned in Encode</th>\n",
       "      <th>% Reads aligned in Magic</th>\n",
       "      <th>% Reads aligned in genome</th>\n",
       "      <th>% Reads aligned in any previously annotated transcripts</th>\n",
       "      <th>% Reads inside introns, new exons and intergenic</th>\n",
       "      <th>% Reads aligned in microbiome</th>\n",
       "      <th>% Reads aligned in viruses</th>\n",
       "      <th>% Reads aligned on the imaginary genome (mapping specificity control)</th>\n",
       "      <th>% Reads supporting known exons-junctions</th>\n",
       "      <th>Unnamed: 120</th>\n",
       "      <th>Mb aligned in PhiX DNA spike-in</th>\n",
       "      <th>Mb aligned in ERCC RNA spike-in</th>\n",
       "      <th>Mb aligned in ribosomal RNA</th>\n",
       "      <th>Mb aligned in mitochondria</th>\n",
       "      <th>Mb aligned in other small genes</th>\n",
       "      <th>Mb aligned in AceView</th>\n",
       "      <th>Mb aligned in RefSeq</th>\n",
       "      <th>Mb aligned in Encode</th>\n",
       "      <th>Mb aligned in Magic</th>\n",
       "      <th>Mb aligned in genome</th>\n",
       "      <th>Mb aligned in any previously annotated transcripts</th>\n",
       "      <th>Mb inside introns, new exons and intergenic</th>\n",
       "      <th>Mb aligned in microbiome</th>\n",
       "      <th>Mb aligned in viruses</th>\n",
       "      <th>Mb aligned on the imaginary genome (mapping specificity control)</th>\n",
       "      <th>Mb supporting known exons-junctions</th>\n",
       "      <th>Unnamed: 137</th>\n",
       "      <th>Number of aligned_fragments</th>\n",
       "      <th>Compatible pairs</th>\n",
       "      <th>Non compatible pairs</th>\n",
       "      <th>Paired end fragments with only one read aligned</th>\n",
       "      <th>% Compatible pairs</th>\n",
       "      <th>% Non compatible pairs</th>\n",
       "      <th>% Paired end fragments with only one read aligned</th>\n",
       "      <th>1% fragments in library are shorter than (nt)</th>\n",
       "      <th>5% fragments in library are shorter than (nt)</th>\n",
       "      <th>mode of fragment lengths</th>\n",
       "      <th>median fragment lengths</th>\n",
       "      <th>average of fragment lengths</th>\n",
       "      <th>5% fragments in library are longer than (nt)</th>\n",
       "      <th>1% fragments in library are longer than (nt)</th>\n",
       "      <th>Unnamed: 152</th>\n",
       "      <th>Average length after clipping adaptors and barcodes (nt)</th>\n",
       "      <th>Average length aligned (nt)</th>\n",
       "      <th>Average length aligned in read 1 (nt)</th>\n",
       "      <th>Average length aligned in read 2 (paired end)</th>\n",
       "      <th>Average length aligned on RefSeq (nt)</th>\n",
       "      <th>Average length aligned on EBI (nt)</th>\n",
       "      <th>Average length aligned on AceView (nt)</th>\n",
       "      <th>Average length aligned on magic (nt)</th>\n",
       "      <th>Average length aligned on Other (nt)</th>\n",
       "      <th>Average length aligned on Genome (nt)</th>\n",
       "      <th>Average length aligned on Mitochondria (nt)</th>\n",
       "      <th>Average length aligned on rRNA (nt)</th>\n",
       "      <th>Average length aligned on Chloroplast (nt)</th>\n",
       "      <th>Average length aligned on small RNAs (nt)</th>\n",
       "      <th>Average length aligned on DNA spikeIn (nt)</th>\n",
       "      <th>Average length aligned on RNA spikeIn (nt)</th>\n",
       "      <th>Average length aligned on Imaginary genome specificity control (nt)</th>\n",
       "      <th>Unnamed: 170</th>\n",
       "      <th>Average % length aligned (nt)</th>\n",
       "      <th>Average % length aligned in read 1 (nt)</th>\n",
       "      <th>Average % length aligned in read 2 (paired end)</th>\n",
       "      <th>Average % length aligned on RefSeq</th>\n",
       "      <th>Average % length aligned on EBI</th>\n",
       "      <th>Average % length aligned on AceView</th>\n",
       "      <th>Average % length aligned on magic</th>\n",
       "      <th>Average % length aligned on Other</th>\n",
       "      <th>Average % length aligned on Genome</th>\n",
       "      <th>Average % length aligned on Mitochondria</th>\n",
       "      <th>Average % length aligned on rRNA</th>\n",
       "      <th>Average length aligned on Chloroplast</th>\n",
       "      <th>Average length aligned on small RNAs</th>\n",
       "      <th>Average length aligned on DNA spikeIn</th>\n",
       "      <th>Average length aligned on RNA spikeIn</th>\n",
       "      <th>Average length aligned on Imaginary genome specificity control</th>\n",
       "      <th>Unnamed: 187</th>\n",
       "      <th>Reads mapping on plus strand of rRNA</th>\n",
       "      <th>Reads mapping on minus strand of rRNA</th>\n",
       "      <th>Reads mapping on both strands of rRNA</th>\n",
       "      <th>% reads mapping on plus strand of rRNA</th>\n",
       "      <th>% reads mapping on minus strand of rRNA</th>\n",
       "      <th>% reads mapping on both strand of rRNA</th>\n",
       "      <th>Reads mapping on plus strand of RefSeq</th>\n",
       "      <th>Reads mapping on minus strand of RefSeq</th>\n",
       "      <th>Reads mapping on both strands of RefSeq</th>\n",
       "      <th>% reads mapping on plus strand of RefSeq</th>\n",
       "      <th>% reads mapping on minus strand of RefSeq</th>\n",
       "      <th>% reads mapping on both strand of RefSeq</th>\n",
       "      <th>Reads mapping on plus strand of EBI</th>\n",
       "      <th>Reads mapping on minus strand of EBI</th>\n",
       "      <th>Reads mapping on both strands of EBI</th>\n",
       "      <th>% reads mapping on plus strand of EBI</th>\n",
       "      <th>% reads mapping on minus strand of EBI</th>\n",
       "      <th>% reads mapping on both strand of EBI</th>\n",
       "      <th>Reads mapping on plus strand of av</th>\n",
       "      <th>Reads mapping on minus strand of av</th>\n",
       "      <th>Reads mapping on both strands of av</th>\n",
       "      <th>% reads mapping on plus strand of av</th>\n",
       "      <th>% reads mapping on minus strand of av</th>\n",
       "      <th>% reads mapping on both strand of av</th>\n",
       "      <th>Reads mapping on plus strand of genome</th>\n",
       "      <th>Reads mapping on minus strand of genome</th>\n",
       "      <th>Reads mapping on both strands of genome</th>\n",
       "      <th>% reads mapping on plus strand of genome</th>\n",
       "      <th>% reads mapping on minus strand of genome</th>\n",
       "      <th>% reads mapping on both strand of genome</th>\n",
       "      <th>Reads mapping on plus strand of SpikeIn</th>\n",
       "      <th>Reads mapping on minus strand of SpikeIn</th>\n",
       "      <th>Reads mapping on both strands of SpikeIn</th>\n",
       "      <th>% reads mapping on plus strand of SpikeIn</th>\n",
       "      <th>% reads mapping on minus strand of SpikeIn</th>\n",
       "      <th>% reads mapping on both strand of SpikeIn</th>\n",
       "      <th>Unnamed: 224</th>\n",
       "      <th>Megabases uniquely aligned and used for mismatch counts</th>\n",
       "      <th>Total number of mismatches</th>\n",
       "      <th>Mismatches per kb aligned</th>\n",
       "      <th>Transitions</th>\n",
       "      <th>Transversions</th>\n",
       "      <th>1, 2 or 3 bp insertions not in polymers</th>\n",
       "      <th>1, 2 or 3 bp deletions not in polymers</th>\n",
       "      <th>1, 2 or 3 bp insertions in polymers</th>\n",
       "      <th>1, 2 or 3 bp deletions in polymers</th>\n",
       "      <th>A&gt;G</th>\n",
       "      <th>T&gt;C</th>\n",
       "      <th>G&gt;A</th>\n",
       "      <th>C&gt;T</th>\n",
       "      <th>A&gt;T</th>\n",
       "      <th>T&gt;A</th>\n",
       "      <th>G&gt;C</th>\n",
       "      <th>C&gt;G</th>\n",
       "      <th>A&gt;C</th>\n",
       "      <th>T&gt;G</th>\n",
       "      <th>G&gt;T</th>\n",
       "      <th>C&gt;A</th>\n",
       "      <th>Single Insertions</th>\n",
       "      <th>Single Deletions</th>\n",
       "      <th>Double insertions</th>\n",
       "      <th>Double deletions</th>\n",
       "      <th>Triple insertions</th>\n",
       "      <th>Triple deletions</th>\n",
       "      <th>Transitions per kb</th>\n",
       "      <th>Transversions per kb</th>\n",
       "      <th>Insertions per kb</th>\n",
       "      <th>Deletions per kb</th>\n",
       "      <th>% transitions</th>\n",
       "      <th>% transversions</th>\n",
       "      <th>% insertions not in polymers</th>\n",
       "      <th>% deletions not in polymers</th>\n",
       "      <th>% insertions in polymers</th>\n",
       "      <th>% deletions in polymers</th>\n",
       "      <th>Unnamed: 262</th>\n",
       "      <th>Accessible transcript length, estimated on transcripts longer than 8kb, usually limited by the 3' biais in poly-A selected experiments</th>\n",
       "      <th>Number of well expressed transcripts longer than 8kb</th>\n",
       "      <th>Average coverage cumulated over these transcripts</th>\n",
       "      <th>Unnamed: 266</th>\n",
       "      <th>Intergenic kb</th>\n",
       "      <th>Intergenic density</th>\n",
       "      <th>Unnamed: 269</th>\n",
       "      <th>Zero index</th>\n",
       "      <th>Low index</th>\n",
       "      <th>Cross over index</th>\n",
       "      <th>Genes touched</th>\n",
       "      <th>Genes with reliable index</th>\n",
       "      <th>Genes with index &gt;= 10</th>\n",
       "      <th>Genes with index &gt;= 12</th>\n",
       "      <th>Genes with index &gt;= 15</th>\n",
       "      <th>Genes with index &gt;= 18</th>\n",
       "      <th>Genes expressed in at least one run</th>\n",
       "      <th>RefSeq genes significantly expressed in at least one run</th>\n",
       "      <th>RefSeq genes significantly expressed in majority of runs</th>\n",
       "      <th>ENCODE genes significantly expressed in at least one run</th>\n",
       "      <th>ENCODE genes significantly expressed in majority of runs</th>\n",
       "      <th>AceView genes significantly expressed in at least one run</th>\n",
       "      <th>AceView genes significantly expressed in majority of runs</th>\n",
       "      <th>AceView transcripts significantly expressed in  at least one run</th>\n",
       "      <th>Mb_aligned</th>\n",
       "      <th>Mb_in_genes</th>\n",
       "      <th>Mb_in_genes_with_GeneId_minus_high_genes</th>\n",
       "      <th>Mb_in_high_genes</th>\n",
       "      <th>Intergenic</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>### Run</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR332342</th>\n",
       "      <td>NaN</td>\n",
       "      <td>10772.0</td>\n",
       "      <td>51.057</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SRR332342</td>\n",
       "      <td>SRR332342</td>\n",
       "      <td>Wholeflies w1118 Female R3 Illumina, 5 day mat...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RF1pw</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Wholeflies w1118 Female R3 Illumina, 5 day mat...</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Whole_organism</td>\n",
       "      <td>Malone JH, Cho DY, Mattiuzzo NR, Artieri CG, J...</td>\n",
       "      <td>Drosophila melanogaster (taxid:7227)</td>\n",
       "      <td>RNA-Seq on Illumina platform of Drosophila Dro...</td>\n",
       "      <td>RNA-Seq reads of DrosDel deficiency (Df/+) and...</td>\n",
       "      <td>PMID: 22531030 : Mediation of Drosophila autos...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RNA, Unspecified_RNA, RNA-Seq, cDNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Illumina, Genome Analyzer II</td>\n",
       "      <td>sraRNA, sraUnspecified_RNA, RNA-Seq, cDNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2012-04-05</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>14403905 spots, 518540580 bases, average read ...</td>\n",
       "      <td>Single-end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>f1:36</td>\n",
       "      <td>2.60</td>\n",
       "      <td>5.534</td>\n",
       "      <td>14.404</td>\n",
       "      <td>518.471</td>\n",
       "      <td>25.21</td>\n",
       "      <td>24.26</td>\n",
       "      <td>25.59</td>\n",
       "      <td>24.94</td>\n",
       "      <td>0.01</td>\n",
       "      <td>50.52</td>\n",
       "      <td>NaN</td>\n",
       "      <td>agatcggaagagct</td>\n",
       "      <td>NaN</td>\n",
       "      <td>69.29</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>35.51</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13.201</td>\n",
       "      <td>35.17</td>\n",
       "      <td>464.254</td>\n",
       "      <td>91.65</td>\n",
       "      <td>89.54</td>\n",
       "      <td>97.69</td>\n",
       "      <td>NaN</td>\n",
       "      <td>14403900</td>\n",
       "      <td>13200970</td>\n",
       "      <td>12808340</td>\n",
       "      <td>117307</td>\n",
       "      <td>12925647</td>\n",
       "      <td>275322</td>\n",
       "      <td>24910</td>\n",
       "      <td>107511</td>\n",
       "      <td>1933</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1068577</td>\n",
       "      <td>91.65</td>\n",
       "      <td>88.92</td>\n",
       "      <td>0.814</td>\n",
       "      <td>89.74</td>\n",
       "      <td>1.91</td>\n",
       "      <td>0.173</td>\n",
       "      <td>0.746</td>\n",
       "      <td>0.0134</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7.42</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "      <td>42</td>\n",
       "      <td>5723251</td>\n",
       "      <td>86672</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7061319</td>\n",
       "      <td>0</td>\n",
       "      <td>6859694</td>\n",
       "      <td>12869513</td>\n",
       "      <td>331452</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "      <td>279034</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0000347</td>\n",
       "      <td>0.000292</td>\n",
       "      <td>39.73</td>\n",
       "      <td>0.602</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>49.024</td>\n",
       "      <td>0</td>\n",
       "      <td>47.62</td>\n",
       "      <td>89.35</td>\n",
       "      <td>2.30</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000174</td>\n",
       "      <td>1.94</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>203.06</td>\n",
       "      <td>3.08</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>246.61</td>\n",
       "      <td>0.0</td>\n",
       "      <td>241.21</td>\n",
       "      <td>452.69</td>\n",
       "      <td>11559.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>35.72</td>\n",
       "      <td>35.17</td>\n",
       "      <td>35.17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>34.92</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>35.16</td>\n",
       "      <td>35.53</td>\n",
       "      <td>35.48</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>98.46</td>\n",
       "      <td>98.46</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>98.03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>98.710</td>\n",
       "      <td>98.830</td>\n",
       "      <td>98.970</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2954663.0</td>\n",
       "      <td>2768588.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>51.6300</td>\n",
       "      <td>48.3700</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3541738.0</td>\n",
       "      <td>3395040.0</td>\n",
       "      <td>124541.0</td>\n",
       "      <td>50.157</td>\n",
       "      <td>48.08</td>\n",
       "      <td>1.76</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3431228.0</td>\n",
       "      <td>3428466.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>50.02010</td>\n",
       "      <td>49.97990</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>439</td>\n",
       "      <td>587469</td>\n",
       "      <td>1.33820</td>\n",
       "      <td>287857</td>\n",
       "      <td>281897</td>\n",
       "      <td>410</td>\n",
       "      <td>4186</td>\n",
       "      <td>7043</td>\n",
       "      <td>6076</td>\n",
       "      <td>90084</td>\n",
       "      <td>80492</td>\n",
       "      <td>51011</td>\n",
       "      <td>66270</td>\n",
       "      <td>32086</td>\n",
       "      <td>38624</td>\n",
       "      <td>18388</td>\n",
       "      <td>21903</td>\n",
       "      <td>59697</td>\n",
       "      <td>36659</td>\n",
       "      <td>30693</td>\n",
       "      <td>43847</td>\n",
       "      <td>2210</td>\n",
       "      <td>8138</td>\n",
       "      <td>5228</td>\n",
       "      <td>1988</td>\n",
       "      <td>15</td>\n",
       "      <td>136</td>\n",
       "      <td>0.65571</td>\n",
       "      <td>0.64213</td>\n",
       "      <td>0.00093</td>\n",
       "      <td>0.00954</td>\n",
       "      <td>49.00</td>\n",
       "      <td>47.99</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.71</td>\n",
       "      <td>1.20</td>\n",
       "      <td>1.03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4800.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>345.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>377</td>\n",
       "      <td>0.0107</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.3</td>\n",
       "      <td>9.4</td>\n",
       "      <td>9.9</td>\n",
       "      <td>12795.0</td>\n",
       "      <td>10772.0</td>\n",
       "      <td>10193.0</td>\n",
       "      <td>8737.0</td>\n",
       "      <td>3973.0</td>\n",
       "      <td>428.0</td>\n",
       "      <td>10772.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10772.0</td>\n",
       "      <td>10772.0</td>\n",
       "      <td>10772.0</td>\n",
       "      <td>10772.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>464.254</td>\n",
       "      <td>246.611</td>\n",
       "      <td>229.655</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1577515</th>\n",
       "      <td>NaN</td>\n",
       "      <td>10038.0</td>\n",
       "      <td>50.920</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SRR1577515</td>\n",
       "      <td>SRR1577515</td>\n",
       "      <td>Df(2L)ED50001/+ female rep1, whole animal, w[1...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RF2pw</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Df(2L)ED50001/+ female rep1, whole animal, w[1...</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Whole_organism</td>\n",
       "      <td>Lee H, Russell S, Oliver B, NIDDK, NIH, Bethes...</td>\n",
       "      <td>Drosophila melanogaster (taxid:7227)</td>\n",
       "      <td>Expression profiling pooled Drosophila melanog...</td>\n",
       "      <td>We performed mRNA transcriptional profiling on...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RNA, Unspecified_RNA, RNA-Seq, cDNA</td>\n",
       "      <td>Flies were homogenized in 500 ul of RNAlater s...</td>\n",
       "      <td>Illumina, HiSeq 2000</td>\n",
       "      <td>sraRNA, sraUnspecified_RNA, RNA-Seq, cDNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6196373 spots, 470924348 bases, average read l...</td>\n",
       "      <td>Single-end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>f1:76</td>\n",
       "      <td>1.78</td>\n",
       "      <td>3.477</td>\n",
       "      <td>6.196</td>\n",
       "      <td>470.920</td>\n",
       "      <td>24.38</td>\n",
       "      <td>24.17</td>\n",
       "      <td>25.62</td>\n",
       "      <td>25.64</td>\n",
       "      <td>0.19</td>\n",
       "      <td>51.26</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>110.87</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>14.32</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.162</td>\n",
       "      <td>74.71</td>\n",
       "      <td>460.342</td>\n",
       "      <td>99.44</td>\n",
       "      <td>97.75</td>\n",
       "      <td>98.30</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6196373</td>\n",
       "      <td>6161529</td>\n",
       "      <td>5884460</td>\n",
       "      <td>123491</td>\n",
       "      <td>6007951</td>\n",
       "      <td>153578</td>\n",
       "      <td>5852</td>\n",
       "      <td>460</td>\n",
       "      <td>63</td>\n",
       "      <td>9605</td>\n",
       "      <td>0</td>\n",
       "      <td>18864</td>\n",
       "      <td>99.438</td>\n",
       "      <td>94.97</td>\n",
       "      <td>1.99</td>\n",
       "      <td>96.96</td>\n",
       "      <td>2.48</td>\n",
       "      <td>0.0944</td>\n",
       "      <td>0.00742</td>\n",
       "      <td>0.00102</td>\n",
       "      <td>0.155</td>\n",
       "      <td>0</td>\n",
       "      <td>0.304</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6890</td>\n",
       "      <td>19275</td>\n",
       "      <td>408392</td>\n",
       "      <td>319951</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5293878</td>\n",
       "      <td>0</td>\n",
       "      <td>4912715</td>\n",
       "      <td>6040827</td>\n",
       "      <td>113812</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>582715</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.111</td>\n",
       "      <td>0.311</td>\n",
       "      <td>6.59</td>\n",
       "      <td>5.16</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>85.44</td>\n",
       "      <td>0</td>\n",
       "      <td>79.28</td>\n",
       "      <td>97.49</td>\n",
       "      <td>1.84</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000113</td>\n",
       "      <td>9.40</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.52</td>\n",
       "      <td>1.45</td>\n",
       "      <td>30.79</td>\n",
       "      <td>24.06</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>395.17</td>\n",
       "      <td>0.0</td>\n",
       "      <td>368.00</td>\n",
       "      <td>451.42</td>\n",
       "      <td>8406.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>75.82</td>\n",
       "      <td>74.71</td>\n",
       "      <td>74.71</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>74.65</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>74.91</td>\n",
       "      <td>75.21</td>\n",
       "      <td>75.39</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>74.79</td>\n",
       "      <td>75.35</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>98.54</td>\n",
       "      <td>98.54</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>98.50</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>98.830</td>\n",
       "      <td>99.026</td>\n",
       "      <td>99.223</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>98.450</td>\n",
       "      <td>99.184</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>206934.0</td>\n",
       "      <td>201458.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>50.6700</td>\n",
       "      <td>49.3300</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2631597.0</td>\n",
       "      <td>2536482.0</td>\n",
       "      <td>125799.0</td>\n",
       "      <td>49.710</td>\n",
       "      <td>47.91</td>\n",
       "      <td>2.38</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2457799.0</td>\n",
       "      <td>2454916.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>50.02930</td>\n",
       "      <td>49.97070</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9657.0</td>\n",
       "      <td>9618.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>50.1010</td>\n",
       "      <td>49.8990</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>432</td>\n",
       "      <td>1141232</td>\n",
       "      <td>2.64174</td>\n",
       "      <td>625138</td>\n",
       "      <td>485026</td>\n",
       "      <td>1068</td>\n",
       "      <td>5549</td>\n",
       "      <td>10684</td>\n",
       "      <td>13767</td>\n",
       "      <td>170205</td>\n",
       "      <td>175628</td>\n",
       "      <td>139466</td>\n",
       "      <td>139839</td>\n",
       "      <td>44288</td>\n",
       "      <td>55822</td>\n",
       "      <td>46203</td>\n",
       "      <td>55142</td>\n",
       "      <td>46710</td>\n",
       "      <td>84563</td>\n",
       "      <td>74624</td>\n",
       "      <td>77674</td>\n",
       "      <td>8028</td>\n",
       "      <td>15504</td>\n",
       "      <td>2809</td>\n",
       "      <td>2959</td>\n",
       "      <td>915</td>\n",
       "      <td>853</td>\n",
       "      <td>1.44708</td>\n",
       "      <td>1.12275</td>\n",
       "      <td>0.00247</td>\n",
       "      <td>0.01284</td>\n",
       "      <td>54.78</td>\n",
       "      <td>42.50</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.49</td>\n",
       "      <td>0.94</td>\n",
       "      <td>1.21</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4350.0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>528.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>339</td>\n",
       "      <td>0.0096</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.7</td>\n",
       "      <td>9.4</td>\n",
       "      <td>10.1</td>\n",
       "      <td>12244.0</td>\n",
       "      <td>10038.0</td>\n",
       "      <td>9500.0</td>\n",
       "      <td>8213.0</td>\n",
       "      <td>3195.0</td>\n",
       "      <td>447.0</td>\n",
       "      <td>10038.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10038.0</td>\n",
       "      <td>10038.0</td>\n",
       "      <td>10038.0</td>\n",
       "      <td>10038.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>460.342</td>\n",
       "      <td>395.168</td>\n",
       "      <td>371.811</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1577516</th>\n",
       "      <td>NaN</td>\n",
       "      <td>10168.0</td>\n",
       "      <td>50.879</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SRR1577516</td>\n",
       "      <td>SRR1577516</td>\n",
       "      <td>Df(2L)ED50001/+ female rep2, whole animal, w[1...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RF2pw</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Df(2L)ED50001/+ female rep2, whole animal, w[1...</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Whole_organism</td>\n",
       "      <td>Lee H, Russell S, Oliver B, NIDDK, NIH, Bethes...</td>\n",
       "      <td>Drosophila melanogaster (taxid:7227)</td>\n",
       "      <td>Expression profiling pooled Drosophila melanog...</td>\n",
       "      <td>We performed mRNA transcriptional profiling on...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RNA, Unspecified_RNA, RNA-Seq, cDNA</td>\n",
       "      <td>Flies were homogenized in 500 ul of RNAlater s...</td>\n",
       "      <td>Illumina, HiSeq 2000</td>\n",
       "      <td>sraRNA, sraUnspecified_RNA, RNA-Seq, cDNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6560264 spots, 498580064 bases, average read l...</td>\n",
       "      <td>Single-end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>f1:76</td>\n",
       "      <td>1.86</td>\n",
       "      <td>3.525</td>\n",
       "      <td>6.560</td>\n",
       "      <td>498.577</td>\n",
       "      <td>24.72</td>\n",
       "      <td>24.52</td>\n",
       "      <td>25.24</td>\n",
       "      <td>25.33</td>\n",
       "      <td>0.19</td>\n",
       "      <td>50.57</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>106.25</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.46</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13.34</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.492</td>\n",
       "      <td>74.72</td>\n",
       "      <td>485.088</td>\n",
       "      <td>98.96</td>\n",
       "      <td>97.29</td>\n",
       "      <td>98.32</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6560264</td>\n",
       "      <td>6492023</td>\n",
       "      <td>6190697</td>\n",
       "      <td>158716</td>\n",
       "      <td>6349413</td>\n",
       "      <td>142610</td>\n",
       "      <td>6527</td>\n",
       "      <td>434</td>\n",
       "      <td>43</td>\n",
       "      <td>37200</td>\n",
       "      <td>0</td>\n",
       "      <td>24037</td>\n",
       "      <td>98.96</td>\n",
       "      <td>94.37</td>\n",
       "      <td>2.42</td>\n",
       "      <td>96.79</td>\n",
       "      <td>2.17</td>\n",
       "      <td>0.0995</td>\n",
       "      <td>0.00662</td>\n",
       "      <td>0.000655</td>\n",
       "      <td>0.567</td>\n",
       "      <td>0</td>\n",
       "      <td>0.366</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5731</td>\n",
       "      <td>23670</td>\n",
       "      <td>465968</td>\n",
       "      <td>392353</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5478360</td>\n",
       "      <td>0</td>\n",
       "      <td>5089242</td>\n",
       "      <td>6359524</td>\n",
       "      <td>126768</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>623010</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0874</td>\n",
       "      <td>0.361</td>\n",
       "      <td>7.10</td>\n",
       "      <td>5.98</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>83.51</td>\n",
       "      <td>0</td>\n",
       "      <td>77.58</td>\n",
       "      <td>96.94</td>\n",
       "      <td>1.93</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000152</td>\n",
       "      <td>9.50</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.43</td>\n",
       "      <td>1.78</td>\n",
       "      <td>35.13</td>\n",
       "      <td>29.51</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>408.94</td>\n",
       "      <td>0.0</td>\n",
       "      <td>381.22</td>\n",
       "      <td>475.30</td>\n",
       "      <td>9363.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>75.81</td>\n",
       "      <td>74.72</td>\n",
       "      <td>74.72</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>74.65</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>74.91</td>\n",
       "      <td>75.21</td>\n",
       "      <td>75.39</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>74.77</td>\n",
       "      <td>75.33</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>98.56</td>\n",
       "      <td>98.56</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>98.51</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>98.840</td>\n",
       "      <td>99.026</td>\n",
       "      <td>99.237</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>98.420</td>\n",
       "      <td>99.158</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>233089.0</td>\n",
       "      <td>232879.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>50.0225</td>\n",
       "      <td>49.9775</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2705334.0</td>\n",
       "      <td>2611876.0</td>\n",
       "      <td>161150.0</td>\n",
       "      <td>49.382</td>\n",
       "      <td>47.68</td>\n",
       "      <td>2.94</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2547121.0</td>\n",
       "      <td>2542121.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>50.04910</td>\n",
       "      <td>49.95090</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11831.0</td>\n",
       "      <td>11839.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>49.9831</td>\n",
       "      <td>50.0169</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>454</td>\n",
       "      <td>1171627</td>\n",
       "      <td>2.58068</td>\n",
       "      <td>640940</td>\n",
       "      <td>493785</td>\n",
       "      <td>1183</td>\n",
       "      <td>6558</td>\n",
       "      <td>12604</td>\n",
       "      <td>16557</td>\n",
       "      <td>177008</td>\n",
       "      <td>183103</td>\n",
       "      <td>139647</td>\n",
       "      <td>141182</td>\n",
       "      <td>44785</td>\n",
       "      <td>56011</td>\n",
       "      <td>45952</td>\n",
       "      <td>56248</td>\n",
       "      <td>49660</td>\n",
       "      <td>88413</td>\n",
       "      <td>73955</td>\n",
       "      <td>78761</td>\n",
       "      <td>9641</td>\n",
       "      <td>18363</td>\n",
       "      <td>3042</td>\n",
       "      <td>3777</td>\n",
       "      <td>1104</td>\n",
       "      <td>975</td>\n",
       "      <td>1.41176</td>\n",
       "      <td>1.08763</td>\n",
       "      <td>0.00261</td>\n",
       "      <td>0.01444</td>\n",
       "      <td>54.71</td>\n",
       "      <td>42.15</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.56</td>\n",
       "      <td>1.08</td>\n",
       "      <td>1.41</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4400.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>608.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>381</td>\n",
       "      <td>0.0108</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.7</td>\n",
       "      <td>9.5</td>\n",
       "      <td>10.1</td>\n",
       "      <td>12420.0</td>\n",
       "      <td>10168.0</td>\n",
       "      <td>9643.0</td>\n",
       "      <td>8217.0</td>\n",
       "      <td>3102.0</td>\n",
       "      <td>438.0</td>\n",
       "      <td>10168.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10168.0</td>\n",
       "      <td>10168.0</td>\n",
       "      <td>10168.0</td>\n",
       "      <td>10168.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>485.088</td>\n",
       "      <td>408.939</td>\n",
       "      <td>375.547</td>\n",
       "      <td>8.19641</td>\n",
       "      <td>381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1577403</th>\n",
       "      <td>NaN</td>\n",
       "      <td>10319.0</td>\n",
       "      <td>50.694</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SRR1577403</td>\n",
       "      <td>SRR1577403</td>\n",
       "      <td>Df(2L)ED5878/+ female rep1, whole animal, w[11...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RF3pw</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Df(2L)ED5878/+ female rep1, whole animal, w[11...</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Whole_organism</td>\n",
       "      <td>Lee H, Russell S, Oliver B, NIDDK, NIH, Bethes...</td>\n",
       "      <td>Drosophila melanogaster (taxid:7227)</td>\n",
       "      <td>Expression profiling pooled Drosophila melanog...</td>\n",
       "      <td>We performed mRNA transcriptional profiling on...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RNA, Unspecified_RNA, RNA-Seq, cDNA</td>\n",
       "      <td>Flies were homogenized in 500 ul of RNAlater s...</td>\n",
       "      <td>Illumina, HiSeq 2000</td>\n",
       "      <td>sraRNA, sraUnspecified_RNA, RNA-Seq, cDNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5748503 spots, 436886228 bases, average read l...</td>\n",
       "      <td>Single-end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>f1:76</td>\n",
       "      <td>1.96</td>\n",
       "      <td>2.938</td>\n",
       "      <td>5.749</td>\n",
       "      <td>436.867</td>\n",
       "      <td>25.39</td>\n",
       "      <td>25.14</td>\n",
       "      <td>24.73</td>\n",
       "      <td>24.73</td>\n",
       "      <td>0.01</td>\n",
       "      <td>49.46</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>174.31</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>14.10</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.700</td>\n",
       "      <td>74.95</td>\n",
       "      <td>427.194</td>\n",
       "      <td>99.16</td>\n",
       "      <td>97.79</td>\n",
       "      <td>98.62</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5748503</td>\n",
       "      <td>5700078</td>\n",
       "      <td>5483738</td>\n",
       "      <td>74298</td>\n",
       "      <td>5558036</td>\n",
       "      <td>142042</td>\n",
       "      <td>8362</td>\n",
       "      <td>484</td>\n",
       "      <td>252</td>\n",
       "      <td>20449</td>\n",
       "      <td>0</td>\n",
       "      <td>18878</td>\n",
       "      <td>99.158</td>\n",
       "      <td>95.39</td>\n",
       "      <td>1.29</td>\n",
       "      <td>96.69</td>\n",
       "      <td>2.47</td>\n",
       "      <td>0.145</td>\n",
       "      <td>0.00842</td>\n",
       "      <td>0.00438</td>\n",
       "      <td>0.356</td>\n",
       "      <td>0</td>\n",
       "      <td>0.328</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9455</td>\n",
       "      <td>26809</td>\n",
       "      <td>38325</td>\n",
       "      <td>744335</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4756622</td>\n",
       "      <td>0</td>\n",
       "      <td>4445117</td>\n",
       "      <td>5566025</td>\n",
       "      <td>124598</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>522264</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.164</td>\n",
       "      <td>0.466</td>\n",
       "      <td>0.667</td>\n",
       "      <td>12.95</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>82.75</td>\n",
       "      <td>0</td>\n",
       "      <td>77.33</td>\n",
       "      <td>96.83</td>\n",
       "      <td>2.17</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000522</td>\n",
       "      <td>9.09</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.71</td>\n",
       "      <td>2.03</td>\n",
       "      <td>2.89</td>\n",
       "      <td>56.20</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>356.12</td>\n",
       "      <td>0.0</td>\n",
       "      <td>334.04</td>\n",
       "      <td>417.23</td>\n",
       "      <td>9248.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>75.82</td>\n",
       "      <td>74.95</td>\n",
       "      <td>74.95</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>74.87</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>75.15</td>\n",
       "      <td>75.50</td>\n",
       "      <td>75.43</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>75.38</td>\n",
       "      <td>75.58</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>98.85</td>\n",
       "      <td>98.85</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>98.79</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>99.156</td>\n",
       "      <td>99.408</td>\n",
       "      <td>99.289</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>99.236</td>\n",
       "      <td>99.474</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>19377.0</td>\n",
       "      <td>18948.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>50.5600</td>\n",
       "      <td>49.4400</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2372479.0</td>\n",
       "      <td>2307496.0</td>\n",
       "      <td>76647.0</td>\n",
       "      <td>49.877</td>\n",
       "      <td>48.51</td>\n",
       "      <td>1.61</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2220549.0</td>\n",
       "      <td>2224568.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>49.95480</td>\n",
       "      <td>50.04520</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13321.0</td>\n",
       "      <td>13488.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>49.6890</td>\n",
       "      <td>50.3110</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>402</td>\n",
       "      <td>1037681</td>\n",
       "      <td>2.58130</td>\n",
       "      <td>588815</td>\n",
       "      <td>406331</td>\n",
       "      <td>1151</td>\n",
       "      <td>6837</td>\n",
       "      <td>10067</td>\n",
       "      <td>24480</td>\n",
       "      <td>152444</td>\n",
       "      <td>184047</td>\n",
       "      <td>124325</td>\n",
       "      <td>127999</td>\n",
       "      <td>39249</td>\n",
       "      <td>44398</td>\n",
       "      <td>43001</td>\n",
       "      <td>49927</td>\n",
       "      <td>37868</td>\n",
       "      <td>66661</td>\n",
       "      <td>63349</td>\n",
       "      <td>61878</td>\n",
       "      <td>8966</td>\n",
       "      <td>26842</td>\n",
       "      <td>1385</td>\n",
       "      <td>3518</td>\n",
       "      <td>867</td>\n",
       "      <td>957</td>\n",
       "      <td>1.46471</td>\n",
       "      <td>1.01077</td>\n",
       "      <td>0.00286</td>\n",
       "      <td>0.01701</td>\n",
       "      <td>56.74</td>\n",
       "      <td>39.16</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.66</td>\n",
       "      <td>0.97</td>\n",
       "      <td>2.36</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3150.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>483.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>329</td>\n",
       "      <td>0.0093</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.9</td>\n",
       "      <td>10.1</td>\n",
       "      <td>10.4</td>\n",
       "      <td>12433.0</td>\n",
       "      <td>10319.0</td>\n",
       "      <td>10041.0</td>\n",
       "      <td>8690.0</td>\n",
       "      <td>3334.0</td>\n",
       "      <td>481.0</td>\n",
       "      <td>10319.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10319.0</td>\n",
       "      <td>10319.0</td>\n",
       "      <td>10319.0</td>\n",
       "      <td>10319.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>427.194</td>\n",
       "      <td>356.124</td>\n",
       "      <td>337.417</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>329</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR1577404</th>\n",
       "      <td>NaN</td>\n",
       "      <td>10381.0</td>\n",
       "      <td>50.399</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SRR1577404</td>\n",
       "      <td>SRR1577404</td>\n",
       "      <td>Df(2L)ED5878/+ female rep2, whole animal, w[11...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RF3pw</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Df(2L)ED5878/+ female rep2, whole animal, w[11...</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Whole_organism</td>\n",
       "      <td>Lee H, Russell S, Oliver B, NIDDK, NIH, Bethes...</td>\n",
       "      <td>Drosophila melanogaster (taxid:7227)</td>\n",
       "      <td>Expression profiling pooled Drosophila melanog...</td>\n",
       "      <td>We performed mRNA transcriptional profiling on...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RNA, Unspecified_RNA, RNA-Seq, cDNA</td>\n",
       "      <td>Flies were homogenized in 500 ul of RNAlater s...</td>\n",
       "      <td>Illumina, HiSeq 2000</td>\n",
       "      <td>sraRNA, sraUnspecified_RNA, RNA-Seq, cDNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7623975 spots, 579422100 bases, average read l...</td>\n",
       "      <td>Single-end</td>\n",
       "      <td>NaN</td>\n",
       "      <td>f1:76</td>\n",
       "      <td>2.23</td>\n",
       "      <td>3.423</td>\n",
       "      <td>7.624</td>\n",
       "      <td>579.418</td>\n",
       "      <td>25.63</td>\n",
       "      <td>25.12</td>\n",
       "      <td>24.45</td>\n",
       "      <td>24.79</td>\n",
       "      <td>0.01</td>\n",
       "      <td>49.24</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>102.83</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15.49</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.931</td>\n",
       "      <td>74.96</td>\n",
       "      <td>519.522</td>\n",
       "      <td>90.91</td>\n",
       "      <td>89.66</td>\n",
       "      <td>98.63</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7623975</td>\n",
       "      <td>6930835</td>\n",
       "      <td>6662658</td>\n",
       "      <td>108637</td>\n",
       "      <td>6771295</td>\n",
       "      <td>159540</td>\n",
       "      <td>9583</td>\n",
       "      <td>600</td>\n",
       "      <td>54</td>\n",
       "      <td>622986</td>\n",
       "      <td>0</td>\n",
       "      <td>59917</td>\n",
       "      <td>90.91</td>\n",
       "      <td>87.39</td>\n",
       "      <td>1.42</td>\n",
       "      <td>88.82</td>\n",
       "      <td>2.09</td>\n",
       "      <td>0.126</td>\n",
       "      <td>0.00787</td>\n",
       "      <td>0.000708</td>\n",
       "      <td>8.17</td>\n",
       "      <td>0</td>\n",
       "      <td>0.786</td>\n",
       "      <td>NaN</td>\n",
       "      <td>14382</td>\n",
       "      <td>44969</td>\n",
       "      <td>483772</td>\n",
       "      <td>776271</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5448639</td>\n",
       "      <td>0</td>\n",
       "      <td>5110327</td>\n",
       "      <td>6752733</td>\n",
       "      <td>163720</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>603352</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.189</td>\n",
       "      <td>0.590</td>\n",
       "      <td>6.35</td>\n",
       "      <td>10.18</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>71.47</td>\n",
       "      <td>0</td>\n",
       "      <td>67.03</td>\n",
       "      <td>88.57</td>\n",
       "      <td>2.15</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000918</td>\n",
       "      <td>7.91</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.08</td>\n",
       "      <td>3.40</td>\n",
       "      <td>36.57</td>\n",
       "      <td>58.59</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>407.79</td>\n",
       "      <td>0.0</td>\n",
       "      <td>383.86</td>\n",
       "      <td>506.27</td>\n",
       "      <td>12167.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>75.82</td>\n",
       "      <td>74.96</td>\n",
       "      <td>74.96</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>74.84</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>75.11</td>\n",
       "      <td>75.47</td>\n",
       "      <td>75.59</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>75.41</td>\n",
       "      <td>75.55</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>98.87</td>\n",
       "      <td>98.87</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>98.75</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>99.103</td>\n",
       "      <td>99.368</td>\n",
       "      <td>99.500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>99.276</td>\n",
       "      <td>99.447</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>243757.0</td>\n",
       "      <td>240015.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>50.3870</td>\n",
       "      <td>49.6130</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2690006.0</td>\n",
       "      <td>2647449.0</td>\n",
       "      <td>111184.0</td>\n",
       "      <td>49.370</td>\n",
       "      <td>48.59</td>\n",
       "      <td>2.04</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2554700.0</td>\n",
       "      <td>2555627.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>49.99093</td>\n",
       "      <td>50.00907</td>\n",
       "      <td>0.0</td>\n",
       "      <td>22251.0</td>\n",
       "      <td>22718.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>49.4810</td>\n",
       "      <td>50.5190</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>489</td>\n",
       "      <td>1241053</td>\n",
       "      <td>2.53794</td>\n",
       "      <td>672545</td>\n",
       "      <td>510945</td>\n",
       "      <td>1508</td>\n",
       "      <td>11826</td>\n",
       "      <td>15243</td>\n",
       "      <td>28986</td>\n",
       "      <td>179191</td>\n",
       "      <td>207480</td>\n",
       "      <td>140849</td>\n",
       "      <td>145025</td>\n",
       "      <td>49932</td>\n",
       "      <td>57949</td>\n",
       "      <td>51290</td>\n",
       "      <td>61015</td>\n",
       "      <td>46625</td>\n",
       "      <td>86882</td>\n",
       "      <td>78678</td>\n",
       "      <td>78574</td>\n",
       "      <td>12395</td>\n",
       "      <td>33659</td>\n",
       "      <td>3367</td>\n",
       "      <td>5844</td>\n",
       "      <td>989</td>\n",
       "      <td>1309</td>\n",
       "      <td>1.37535</td>\n",
       "      <td>1.04488</td>\n",
       "      <td>0.00308</td>\n",
       "      <td>0.02418</td>\n",
       "      <td>54.19</td>\n",
       "      <td>41.17</td>\n",
       "      <td>0.12</td>\n",
       "      <td>0.95</td>\n",
       "      <td>1.23</td>\n",
       "      <td>2.34</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3200.0</td>\n",
       "      <td>64.0</td>\n",
       "      <td>614.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>391</td>\n",
       "      <td>0.0111</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.7</td>\n",
       "      <td>9.8</td>\n",
       "      <td>10.2</td>\n",
       "      <td>12643.0</td>\n",
       "      <td>10381.0</td>\n",
       "      <td>9988.0</td>\n",
       "      <td>8549.0</td>\n",
       "      <td>3210.0</td>\n",
       "      <td>434.0</td>\n",
       "      <td>10381.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10381.0</td>\n",
       "      <td>10381.0</td>\n",
       "      <td>10381.0</td>\n",
       "      <td>10381.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>519.522</td>\n",
       "      <td>407.787</td>\n",
       "      <td>385.491</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>391</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Unnamed: 1  Genes significantly expressed  \\\n",
       "### Run                                                 \n",
       "SRR332342          NaN                        10772.0   \n",
       "SRR1577515         NaN                        10038.0   \n",
       "SRR1577516         NaN                        10168.0   \n",
       "SRR1577403         NaN                        10319.0   \n",
       "SRR1577404         NaN                        10381.0   \n",
       "\n",
       "            % fragments on strand plus of genes  Unnamed: 4      RunId   \\\n",
       "### Run                                                                   \n",
       "SRR332342                                51.057         NaN   SRR332342   \n",
       "SRR1577515                               50.920         NaN  SRR1577515   \n",
       "SRR1577516                               50.879         NaN  SRR1577516   \n",
       "SRR1577403                               50.694         NaN  SRR1577403   \n",
       "SRR1577404                               50.399         NaN  SRR1577404   \n",
       "\n",
       "           Run accession in SRA  \\\n",
       "### Run                           \n",
       "SRR332342             SRR332342   \n",
       "SRR1577515           SRR1577515   \n",
       "SRR1577516           SRR1577516   \n",
       "SRR1577403           SRR1577403   \n",
       "SRR1577404           SRR1577404   \n",
       "\n",
       "                                                        Title  Other title  \\\n",
       "### Run                                                                      \n",
       "SRR332342   Wholeflies w1118 Female R3 Illumina, 5 day mat...          NaN   \n",
       "SRR1577515  Df(2L)ED50001/+ female rep1, whole animal, w[1...          NaN   \n",
       "SRR1577516  Df(2L)ED50001/+ female rep2, whole animal, w[1...          NaN   \n",
       "SRR1577403  Df(2L)ED5878/+ female rep1, whole animal, w[11...          NaN   \n",
       "SRR1577404  Df(2L)ED5878/+ female rep2, whole animal, w[11...          NaN   \n",
       "\n",
       "           Sorting_title 1  Sorting title 2  \\\n",
       "### Run                                       \n",
       "SRR332342            RF1pw              NaN   \n",
       "SRR1577515           RF2pw              NaN   \n",
       "SRR1577516           RF2pw              NaN   \n",
       "SRR1577403           RF3pw              NaN   \n",
       "SRR1577404           RF3pw              NaN   \n",
       "\n",
       "                 Sample (summarized from biosample or manual)  \\\n",
       "### Run                                                         \n",
       "SRR332342   Wholeflies w1118 Female R3 Illumina, 5 day mat...   \n",
       "SRR1577515  Df(2L)ED50001/+ female rep1, whole animal, w[1...   \n",
       "SRR1577516  Df(2L)ED50001/+ female rep2, whole animal, w[1...   \n",
       "SRR1577403  Df(2L)ED5878/+ female rep1, whole animal, w[11...   \n",
       "SRR1577404  Df(2L)ED5878/+ female rep2, whole animal, w[11...   \n",
       "\n",
       "           Stage and experimental summary System or Tissue  \\\n",
       "### Run                                                      \n",
       "SRR332342                           Adult   Whole_organism   \n",
       "SRR1577515                          Adult   Whole_organism   \n",
       "SRR1577516                          Adult   Whole_organism   \n",
       "SRR1577403                          Adult   Whole_organism   \n",
       "SRR1577404                          Adult   Whole_organism   \n",
       "\n",
       "           Laboratory, Author when available, submission date  \\\n",
       "### Run                                                         \n",
       "SRR332342   Malone JH, Cho DY, Mattiuzzo NR, Artieri CG, J...   \n",
       "SRR1577515  Lee H, Russell S, Oliver B, NIDDK, NIH, Bethes...   \n",
       "SRR1577516  Lee H, Russell S, Oliver B, NIDDK, NIH, Bethes...   \n",
       "SRR1577403  Lee H, Russell S, Oliver B, NIDDK, NIH, Bethes...   \n",
       "SRR1577404  Lee H, Russell S, Oliver B, NIDDK, NIH, Bethes...   \n",
       "\n",
       "                                         Species  \\\n",
       "### Run                                            \n",
       "SRR332342   Drosophila melanogaster (taxid:7227)   \n",
       "SRR1577515  Drosophila melanogaster (taxid:7227)   \n",
       "SRR1577516  Drosophila melanogaster (taxid:7227)   \n",
       "SRR1577403  Drosophila melanogaster (taxid:7227)   \n",
       "SRR1577404  Drosophila melanogaster (taxid:7227)   \n",
       "\n",
       "                                                      Project  \\\n",
       "### Run                                                         \n",
       "SRR332342   RNA-Seq on Illumina platform of Drosophila Dro...   \n",
       "SRR1577515  Expression profiling pooled Drosophila melanog...   \n",
       "SRR1577516  Expression profiling pooled Drosophila melanog...   \n",
       "SRR1577403  Expression profiling pooled Drosophila melanog...   \n",
       "SRR1577404  Expression profiling pooled Drosophila melanog...   \n",
       "\n",
       "                                          Project description  \\\n",
       "### Run                                                         \n",
       "SRR332342   RNA-Seq reads of DrosDel deficiency (Df/+) and...   \n",
       "SRR1577515  We performed mRNA transcriptional profiling on...   \n",
       "SRR1577516  We performed mRNA transcriptional profiling on...   \n",
       "SRR1577403  We performed mRNA transcriptional profiling on...   \n",
       "SRR1577404  We performed mRNA transcriptional profiling on...   \n",
       "\n",
       "                                                    Reference  Unnamed: 19  \\\n",
       "### Run                                                                      \n",
       "SRR332342   PMID: 22531030 : Mediation of Drosophila autos...          NaN   \n",
       "SRR1577515                                                NaN          NaN   \n",
       "SRR1577516                                                NaN          NaN   \n",
       "SRR1577403                                                NaN          NaN   \n",
       "SRR1577404                                                NaN          NaN   \n",
       "\n",
       "                            Sequencing protocol  \\\n",
       "### Run                                           \n",
       "SRR332342   RNA, Unspecified_RNA, RNA-Seq, cDNA   \n",
       "SRR1577515  RNA, Unspecified_RNA, RNA-Seq, cDNA   \n",
       "SRR1577516  RNA, Unspecified_RNA, RNA-Seq, cDNA   \n",
       "SRR1577403  RNA, Unspecified_RNA, RNA-Seq, cDNA   \n",
       "SRR1577404  RNA, Unspecified_RNA, RNA-Seq, cDNA   \n",
       "\n",
       "                                                       Design  \\\n",
       "### Run                                                         \n",
       "SRR332342                                                 NaN   \n",
       "SRR1577515  Flies were homogenized in 500 ul of RNAlater s...   \n",
       "SRR1577516  Flies were homogenized in 500 ul of RNAlater s...   \n",
       "SRR1577403  Flies were homogenized in 500 ul of RNAlater s...   \n",
       "SRR1577404  Flies were homogenized in 500 ul of RNAlater s...   \n",
       "\n",
       "           Sequencing platform and Machine type  \\\n",
       "### Run                                           \n",
       "SRR332342          Illumina, Genome Analyzer II   \n",
       "SRR1577515                 Illumina, HiSeq 2000   \n",
       "SRR1577516                 Illumina, HiSeq 2000   \n",
       "SRR1577403                 Illumina, HiSeq 2000   \n",
       "SRR1577404                 Illumina, HiSeq 2000   \n",
       "\n",
       "                                      Experiment type  Sequencing date  \\\n",
       "### Run                                                                  \n",
       "SRR332342   sraRNA, sraUnspecified_RNA, RNA-Seq, cDNA              NaN   \n",
       "SRR1577515  sraRNA, sraUnspecified_RNA, RNA-Seq, cDNA              NaN   \n",
       "SRR1577516  sraRNA, sraUnspecified_RNA, RNA-Seq, cDNA              NaN   \n",
       "SRR1577403  sraRNA, sraUnspecified_RNA, RNA-Seq, cDNA              NaN   \n",
       "SRR1577404  sraRNA, sraUnspecified_RNA, RNA-Seq, cDNA              NaN   \n",
       "\n",
       "           Submission date  Library preparation  RNA Integrity Number  \\\n",
       "### Run                                                                 \n",
       "SRR332342       2012-04-05                  NaN                   NaN   \n",
       "SRR1577515             NaN                  NaN                   NaN   \n",
       "SRR1577516             NaN                  NaN                   NaN   \n",
       "SRR1577403             NaN                  NaN                   NaN   \n",
       "SRR1577404             NaN                  NaN                   NaN   \n",
       "\n",
       "            Details  Corresponding microarray  \\\n",
       "### Run                                         \n",
       "SRR332342       NaN                       NaN   \n",
       "SRR1577515      NaN                       NaN   \n",
       "SRR1577516      NaN                       NaN   \n",
       "SRR1577403      NaN                       NaN   \n",
       "SRR1577404      NaN                       NaN   \n",
       "\n",
       "                   Spots, bases, average read length (in SRA)  \\\n",
       "### Run                                                         \n",
       "SRR332342   14403905 spots, 518540580 bases, average read ...   \n",
       "SRR1577515  6196373 spots, 470924348 bases, average read l...   \n",
       "SRR1577516  6560264 spots, 498580064 bases, average read l...   \n",
       "SRR1577403  5748503 spots, 436886228 bases, average read l...   \n",
       "SRR1577404  7623975 spots, 579422100 bases, average read l...   \n",
       "\n",
       "           Single or Paired end  Unnamed: 32 Average read length (nt)  \\\n",
       "### Run                                                                 \n",
       "SRR332342            Single-end          NaN                    f1:36   \n",
       "SRR1577515           Single-end          NaN                    f1:76   \n",
       "SRR1577516           Single-end          NaN                    f1:76   \n",
       "SRR1577403           Single-end          NaN                    f1:76   \n",
       "SRR1577404           Single-end          NaN                    f1:76   \n",
       "\n",
       "            Fragment multiplicity  Million distinct sequences  \\\n",
       "### Run                                                         \n",
       "SRR332342                    2.60                       5.534   \n",
       "SRR1577515                   1.78                       3.477   \n",
       "SRR1577516                   1.86                       3.525   \n",
       "SRR1577403                   1.96                       2.938   \n",
       "SRR1577404                   2.23                       3.423   \n",
       "\n",
       "            Million raw reads  Megabases sequenced     %A     %T     %G  \\\n",
       "### Run                                                                   \n",
       "SRR332342              14.404              518.471  25.21  24.26  25.59   \n",
       "SRR1577515              6.196              470.920  24.38  24.17  25.62   \n",
       "SRR1577516              6.560              498.577  24.72  24.52  25.24   \n",
       "SRR1577403              5.749              436.867  25.39  25.14  24.73   \n",
       "SRR1577404              7.624              579.418  25.63  25.12  24.45   \n",
       "\n",
       "               %C    %N    %GC  Unnamed: 44 Exit adaptor of read 1  \\\n",
       "### Run                                                              \n",
       "SRR332342   24.94  0.01  50.52          NaN         agatcggaagagct   \n",
       "SRR1577515  25.64  0.19  51.26          NaN                    NaN   \n",
       "SRR1577516  25.33  0.19  50.57          NaN                    NaN   \n",
       "SRR1577403  24.73  0.01  49.46          NaN                    NaN   \n",
       "SRR1577404  24.79  0.01  49.24          NaN                    NaN   \n",
       "\n",
       "           Exit adaptor of read 2  \\\n",
       "### Run                             \n",
       "SRR332342                     NaN   \n",
       "SRR1577515                    NaN   \n",
       "SRR1577516                    NaN   \n",
       "SRR1577403                    NaN   \n",
       "SRR1577404                    NaN   \n",
       "\n",
       "            PolyA index : fragments per million with at least 3 A13 motifs shifted by at least 6 bases  \\\n",
       "### Run                                                                                                  \n",
       "SRR332342                                               69.29                                            \n",
       "SRR1577515                                             110.87                                            \n",
       "SRR1577516                                             106.25                                            \n",
       "SRR1577403                                             174.31                                            \n",
       "SRR1577404                                             102.83                                            \n",
       "\n",
       "            Telomere index : fragments per million with at least 3 TTAGGG/CCCTAA motifs  \\\n",
       "### Run                                                                                   \n",
       "SRR332342                                                 0.0                             \n",
       "SRR1577515                                                0.0                             \n",
       "SRR1577516                                                0.0                             \n",
       "SRR1577403                                                0.0                             \n",
       "SRR1577404                                                0.0                             \n",
       "\n",
       "            Telomere index allowing G/C substitution  \\\n",
       "### Run                                                \n",
       "SRR332342                                        0.0   \n",
       "SRR1577515                                       0.0   \n",
       "SRR1577516                                       0.0   \n",
       "SRR1577403                                       0.0   \n",
       "SRR1577404                                       0.0   \n",
       "\n",
       "            Noise in telomere index : FPM with at least 3 AATCCC/GGGATT motifs  \\\n",
       "### Run                                                                          \n",
       "SRR332342                                                0.00                    \n",
       "SRR1577515                                               0.16                    \n",
       "SRR1577516                                               0.46                    \n",
       "SRR1577403                                               0.17                    \n",
       "SRR1577404                                               0.13                    \n",
       "\n",
       "            Unnamed: 51  Million reads aligned on all targets per CPU hour  \\\n",
       "### Run                                                                      \n",
       "SRR332342           NaN                                              35.51   \n",
       "SRR1577515          NaN                                              14.32   \n",
       "SRR1577516          NaN                                              13.34   \n",
       "SRR1577403          NaN                                              14.10   \n",
       "SRR1577404          NaN                                              15.49   \n",
       "\n",
       "            Maximum RAM used (GB)  Number of blocks  Unnamed: 55  \\\n",
       "### Run                                                            \n",
       "SRR332342                       5                 2          NaN   \n",
       "SRR1577515                      8                 1          NaN   \n",
       "SRR1577516                      8                 1          NaN   \n",
       "SRR1577403                      6                 1          NaN   \n",
       "SRR1577404                      8                 1          NaN   \n",
       "\n",
       "            Million reads aligned on any target  \\\n",
       "### Run                                           \n",
       "SRR332342                                13.201   \n",
       "SRR1577515                                6.162   \n",
       "SRR1577516                                6.492   \n",
       "SRR1577403                                5.700   \n",
       "SRR1577404                                6.931   \n",
       "\n",
       "            Average length aligned per read (nt)  Mb aligned on any target  \\\n",
       "### Run                                                                      \n",
       "SRR332342                                  35.17                   464.254   \n",
       "SRR1577515                                 74.71                   460.342   \n",
       "SRR1577516                                 74.72                   485.088   \n",
       "SRR1577403                                 74.95                   427.194   \n",
       "SRR1577404                                 74.96                   519.522   \n",
       "\n",
       "            %Reads aligned on any target  %Mb aligned on any target  \\\n",
       "### Run                                                               \n",
       "SRR332342                          91.65                      89.54   \n",
       "SRR1577515                         99.44                      97.75   \n",
       "SRR1577516                         98.96                      97.29   \n",
       "SRR1577403                         99.16                      97.79   \n",
       "SRR1577404                         90.91                      89.66   \n",
       "\n",
       "            % length aligned on average  Unnamed: 62  Raw reads  \\\n",
       "### Run                                                           \n",
       "SRR332342                         97.69          NaN   14403900   \n",
       "SRR1577515                        98.30          NaN    6196373   \n",
       "SRR1577516                        98.32          NaN    6560264   \n",
       "SRR1577403                        98.62          NaN    5748503   \n",
       "SRR1577404                        98.63          NaN    7623975   \n",
       "\n",
       "            Well mapped reads  \\\n",
       "### Run                         \n",
       "SRR332342            13200970   \n",
       "SRR1577515            6161529   \n",
       "SRR1577516            6492023   \n",
       "SRR1577403            5700078   \n",
       "SRR1577404            6930835   \n",
       "\n",
       "            Reads mapping uniquely  to a single locus (genomic site) and maximum 1 gene  \\\n",
       "### Run                                                                                   \n",
       "SRR332342                                            12808340                             \n",
       "SRR1577515                                            5884460                             \n",
       "SRR1577516                                            6190697                             \n",
       "SRR1577403                                            5483738                             \n",
       "SRR1577404                                            6662658                             \n",
       "\n",
       "            Reads mapping uniquely to a single locus, but to 2 genes in antisense   \\\n",
       "### Run                                                                              \n",
       "SRR332342                                              117307                        \n",
       "SRR1577515                                             123491                        \n",
       "SRR1577516                                             158716                        \n",
       "SRR1577403                                              74298                        \n",
       "SRR1577404                                             108637                        \n",
       "\n",
       "            Reads mapping uniquely to a single locus  \\\n",
       "### Run                                                \n",
       "SRR332342                                   12925647   \n",
       "SRR1577515                                   6007951   \n",
       "SRR1577516                                   6349413   \n",
       "SRR1577403                                   5558036   \n",
       "SRR1577404                                   6771295   \n",
       "\n",
       "            Reads mapping to 2 to 9 sites  \\\n",
       "### Run                                     \n",
       "SRR332342                          275322   \n",
       "SRR1577515                         153578   \n",
       "SRR1577516                         142610   \n",
       "SRR1577403                         142042   \n",
       "SRR1577404                         159540   \n",
       "\n",
       "            Reads rejected because they map to 10 sites or more  \\\n",
       "### Run                                                           \n",
       "SRR332342                                               24910     \n",
       "SRR1577515                                               5852     \n",
       "SRR1577516                                               6527     \n",
       "SRR1577403                                               8362     \n",
       "SRR1577404                                               9583     \n",
       "\n",
       "            Reads rejected by alignment quality filter  \\\n",
       "### Run                                                  \n",
       "SRR332342                                       107511   \n",
       "SRR1577515                                         460   \n",
       "SRR1577516                                         434   \n",
       "SRR1577403                                         484   \n",
       "SRR1577404                                         600   \n",
       "\n",
       "            Reads rejected before alignment because of low entropy (<= 16 bp)  \\\n",
       "### Run                                                                         \n",
       "SRR332342                                                1933                   \n",
       "SRR1577515                                                 63                   \n",
       "SRR1577516                                                 43                   \n",
       "SRR1577403                                                252                   \n",
       "SRR1577404                                                 54                   \n",
       "\n",
       "            Reads with no insert (adaptor only)  \\\n",
       "### Run                                           \n",
       "SRR332342                                     0   \n",
       "SRR1577515                                 9605   \n",
       "SRR1577516                                37200   \n",
       "SRR1577403                                20449   \n",
       "SRR1577404                               622986   \n",
       "\n",
       "            Reads unaligned from half aligned fragments  \\\n",
       "### Run                                                   \n",
       "SRR332342                                             0   \n",
       "SRR1577515                                            0   \n",
       "SRR1577516                                            0   \n",
       "SRR1577403                                            0   \n",
       "SRR1577404                                            0   \n",
       "\n",
       "            Reads of good quality from unaligned fragments (missing genome, microbiome or mapping difficulty)  \\\n",
       "### Run                                                                                                         \n",
       "SRR332342                                             1068577                                                   \n",
       "SRR1577515                                              18864                                                   \n",
       "SRR1577516                                              24037                                                   \n",
       "SRR1577403                                              18878                                                   \n",
       "SRR1577404                                              59917                                                   \n",
       "\n",
       "           % well mapped reads  \\\n",
       "### Run                          \n",
       "SRR332342                91.65   \n",
       "SRR1577515              99.438   \n",
       "SRR1577516               98.96   \n",
       "SRR1577403              99.158   \n",
       "SRR1577404               90.91   \n",
       "\n",
       "           % reads mapping uniquely  to a single locus (genomic site) and maximum 1 gene  \\\n",
       "### Run                                                                                    \n",
       "SRR332342                                               88.92                              \n",
       "SRR1577515                                              94.97                              \n",
       "SRR1577516                                              94.37                              \n",
       "SRR1577403                                              95.39                              \n",
       "SRR1577404                                              87.39                              \n",
       "\n",
       "           % reads mapping uniquely to a single locus, but to 2 genes in antisense   \\\n",
       "### Run                                                                               \n",
       "SRR332342                                               0.814                         \n",
       "SRR1577515                                               1.99                         \n",
       "SRR1577516                                               2.42                         \n",
       "SRR1577403                                               1.29                         \n",
       "SRR1577404                                               1.42                         \n",
       "\n",
       "           % reads mapping uniquely to a single locus  \\\n",
       "### Run                                                 \n",
       "SRR332342                                       89.74   \n",
       "SRR1577515                                      96.96   \n",
       "SRR1577516                                      96.79   \n",
       "SRR1577403                                      96.69   \n",
       "SRR1577404                                      88.82   \n",
       "\n",
       "           % reads mapping to 2 to 9 sites  \\\n",
       "### Run                                      \n",
       "SRR332342                             1.91   \n",
       "SRR1577515                            2.48   \n",
       "SRR1577516                            2.17   \n",
       "SRR1577403                            2.47   \n",
       "SRR1577404                            2.09   \n",
       "\n",
       "           % reads rejected because they map to 10 sites or more  \\\n",
       "### Run                                                            \n",
       "SRR332342                                               0.173      \n",
       "SRR1577515                                             0.0944      \n",
       "SRR1577516                                             0.0995      \n",
       "SRR1577403                                              0.145      \n",
       "SRR1577404                                              0.126      \n",
       "\n",
       "           % reads rejected by alignment quality filter  \\\n",
       "### Run                                                   \n",
       "SRR332342                                         0.746   \n",
       "SRR1577515                                      0.00742   \n",
       "SRR1577516                                      0.00662   \n",
       "SRR1577403                                      0.00842   \n",
       "SRR1577404                                      0.00787   \n",
       "\n",
       "           % reads rejected before alignment because of low entropy (<= 16 bp)  \\\n",
       "### Run                                                                          \n",
       "SRR332342                                              0.0134                    \n",
       "SRR1577515                                            0.00102                    \n",
       "SRR1577516                                           0.000655                    \n",
       "SRR1577403                                            0.00438                    \n",
       "SRR1577404                                           0.000708                    \n",
       "\n",
       "           % reads with no insert (adaptor only)  \\\n",
       "### Run                                            \n",
       "SRR332342                                      0   \n",
       "SRR1577515                                 0.155   \n",
       "SRR1577516                                 0.567   \n",
       "SRR1577403                                 0.356   \n",
       "SRR1577404                                  8.17   \n",
       "\n",
       "           % reads unaligned from half aligned fragments  \\\n",
       "### Run                                                    \n",
       "SRR332342                                              0   \n",
       "SRR1577515                                             0   \n",
       "SRR1577516                                             0   \n",
       "SRR1577403                                             0   \n",
       "SRR1577404                                             0   \n",
       "\n",
       "           % reads of good quality from unaligned fragments (missing genome, microbiome or mapping difficulty)  \\\n",
       "### Run                                                                                                          \n",
       "SRR332342                                                7.42                                                    \n",
       "SRR1577515                                              0.304                                                    \n",
       "SRR1577516                                              0.366                                                    \n",
       "SRR1577403                                              0.328                                                    \n",
       "SRR1577404                                              0.786                                                    \n",
       "\n",
       "            Unnamed: 86  Reads aligned in PhiX DNA spike-in  \\\n",
       "### Run                                                       \n",
       "SRR332342           NaN                                   5   \n",
       "SRR1577515          NaN                                6890   \n",
       "SRR1577516          NaN                                5731   \n",
       "SRR1577403          NaN                                9455   \n",
       "SRR1577404          NaN                               14382   \n",
       "\n",
       "            Reads aligned in ERCC RNA spike-in  \\\n",
       "### Run                                          \n",
       "SRR332342                                   42   \n",
       "SRR1577515                               19275   \n",
       "SRR1577516                               23670   \n",
       "SRR1577403                               26809   \n",
       "SRR1577404                               44969   \n",
       "\n",
       "            Reads aligned in ribosomal RNA  Reads aligned in mitochondria  \\\n",
       "### Run                                                                     \n",
       "SRR332342                          5723251                          86672   \n",
       "SRR1577515                          408392                         319951   \n",
       "SRR1577516                          465968                         392353   \n",
       "SRR1577403                           38325                         744335   \n",
       "SRR1577404                          483772                         776271   \n",
       "\n",
       "            Reads aligned in other small genes  Reads aligned in AceView  \\\n",
       "### Run                                                                    \n",
       "SRR332342                                    0                         0   \n",
       "SRR1577515                                   0                         0   \n",
       "SRR1577516                                   0                         0   \n",
       "SRR1577403                                   0                         0   \n",
       "SRR1577404                                   0                         0   \n",
       "\n",
       "            Reads aligned in RefSeq  Reads aligned in Encode  \\\n",
       "### Run                                                        \n",
       "SRR332342                         0                  7061319   \n",
       "SRR1577515                        0                  5293878   \n",
       "SRR1577516                        0                  5478360   \n",
       "SRR1577403                        0                  4756622   \n",
       "SRR1577404                        0                  5448639   \n",
       "\n",
       "            Reads aligned in Magic  Reads aligned in genome  \\\n",
       "### Run                                                       \n",
       "SRR332342                        0                  6859694   \n",
       "SRR1577515                       0                  4912715   \n",
       "SRR1577516                       0                  5089242   \n",
       "SRR1577403                       0                  4445117   \n",
       "SRR1577404                       0                  5110327   \n",
       "\n",
       "            Reads aligned in any previously annotated transcripts  \\\n",
       "### Run                                                             \n",
       "SRR332342                                            12869513       \n",
       "SRR1577515                                            6040827       \n",
       "SRR1577516                                            6359524       \n",
       "SRR1577403                                            5566025       \n",
       "SRR1577404                                            6752733       \n",
       "\n",
       "            Reads inside introns, new exons and intergenic  \\\n",
       "### Run                                                      \n",
       "SRR332342                                           331452   \n",
       "SRR1577515                                          113812   \n",
       "SRR1577516                                          126768   \n",
       "SRR1577403                                          124598   \n",
       "SRR1577404                                          163720   \n",
       "\n",
       "            Reads aligned in microbiome  Reads aligned in viruses  \\\n",
       "### Run                                                             \n",
       "SRR332342                             0                         0   \n",
       "SRR1577515                            0                         0   \n",
       "SRR1577516                            0                         0   \n",
       "SRR1577403                            0                         0   \n",
       "SRR1577404                            0                         0   \n",
       "\n",
       "            Reads aligned on the imaginary genome (mapping specificity control)  \\\n",
       "### Run                                                                           \n",
       "SRR332342                                                  25                     \n",
       "SRR1577515                                                  7                     \n",
       "SRR1577516                                                 10                     \n",
       "SRR1577403                                                  3                     \n",
       "SRR1577404                                                  7                     \n",
       "\n",
       "            Reads supporting known exons-junctions  Unnamed: 103  \\\n",
       "### Run                                                            \n",
       "SRR332342                                   279034           NaN   \n",
       "SRR1577515                                  582715           NaN   \n",
       "SRR1577516                                  623010           NaN   \n",
       "SRR1577403                                  522264           NaN   \n",
       "SRR1577404                                  603352           NaN   \n",
       "\n",
       "           % Reads aligned in PhiX DNA spike-in  \\\n",
       "### Run                                           \n",
       "SRR332342                             0.0000347   \n",
       "SRR1577515                                0.111   \n",
       "SRR1577516                               0.0874   \n",
       "SRR1577403                                0.164   \n",
       "SRR1577404                                0.189   \n",
       "\n",
       "           % Reads aligned in ERCC RNA spike-in  \\\n",
       "### Run                                           \n",
       "SRR332342                              0.000292   \n",
       "SRR1577515                                0.311   \n",
       "SRR1577516                                0.361   \n",
       "SRR1577403                                0.466   \n",
       "SRR1577404                                0.590   \n",
       "\n",
       "           % Reads aligned in ribosomal RNA % Reads aligned in mitochondria  \\\n",
       "### Run                                                                       \n",
       "SRR332342                             39.73                           0.602   \n",
       "SRR1577515                             6.59                            5.16   \n",
       "SRR1577516                             7.10                            5.98   \n",
       "SRR1577403                            0.667                           12.95   \n",
       "SRR1577404                             6.35                           10.18   \n",
       "\n",
       "           % Reads aligned in other small genes % Reads aligned in AceView  \\\n",
       "### Run                                                                      \n",
       "SRR332342                                     0                          0   \n",
       "SRR1577515                                    0                          0   \n",
       "SRR1577516                                    0                          0   \n",
       "SRR1577403                                    0                          0   \n",
       "SRR1577404                                    0                          0   \n",
       "\n",
       "           % Reads aligned in RefSeq % Reads aligned in Encode  \\\n",
       "### Run                                                          \n",
       "SRR332342                          0                    49.024   \n",
       "SRR1577515                         0                     85.44   \n",
       "SRR1577516                         0                     83.51   \n",
       "SRR1577403                         0                     82.75   \n",
       "SRR1577404                         0                     71.47   \n",
       "\n",
       "           % Reads aligned in Magic % Reads aligned in genome  \\\n",
       "### Run                                                         \n",
       "SRR332342                         0                     47.62   \n",
       "SRR1577515                        0                     79.28   \n",
       "SRR1577516                        0                     77.58   \n",
       "SRR1577403                        0                     77.33   \n",
       "SRR1577404                        0                     67.03   \n",
       "\n",
       "           % Reads aligned in any previously annotated transcripts  \\\n",
       "### Run                                                              \n",
       "SRR332342                                               89.35        \n",
       "SRR1577515                                              97.49        \n",
       "SRR1577516                                              96.94        \n",
       "SRR1577403                                              96.83        \n",
       "SRR1577404                                              88.57        \n",
       "\n",
       "           % Reads inside introns, new exons and intergenic  \\\n",
       "### Run                                                       \n",
       "SRR332342                                              2.30   \n",
       "SRR1577515                                             1.84   \n",
       "SRR1577516                                             1.93   \n",
       "SRR1577403                                             2.17   \n",
       "SRR1577404                                             2.15   \n",
       "\n",
       "           % Reads aligned in microbiome % Reads aligned in viruses  \\\n",
       "### Run                                                               \n",
       "SRR332342                              0                          0   \n",
       "SRR1577515                             0                          0   \n",
       "SRR1577516                             0                          0   \n",
       "SRR1577403                             0                          0   \n",
       "SRR1577404                             0                          0   \n",
       "\n",
       "           % Reads aligned on the imaginary genome (mapping specificity control)  \\\n",
       "### Run                                                                            \n",
       "SRR332342                                            0.000174                      \n",
       "SRR1577515                                           0.000113                      \n",
       "SRR1577516                                           0.000152                      \n",
       "SRR1577403                                          0.0000522                      \n",
       "SRR1577404                                          0.0000918                      \n",
       "\n",
       "           % Reads supporting known exons-junctions  Unnamed: 120  \\\n",
       "### Run                                                             \n",
       "SRR332342                                      1.94           NaN   \n",
       "SRR1577515                                     9.40           NaN   \n",
       "SRR1577516                                     9.50           NaN   \n",
       "SRR1577403                                     9.09           NaN   \n",
       "SRR1577404                                     7.91           NaN   \n",
       "\n",
       "            Mb aligned in PhiX DNA spike-in  Mb aligned in ERCC RNA spike-in  \\\n",
       "### Run                                                                        \n",
       "SRR332342                              0.00                             0.00   \n",
       "SRR1577515                             0.52                             1.45   \n",
       "SRR1577516                             0.43                             1.78   \n",
       "SRR1577403                             0.71                             2.03   \n",
       "SRR1577404                             1.08                             3.40   \n",
       "\n",
       "            Mb aligned in ribosomal RNA  Mb aligned in mitochondria  \\\n",
       "### Run                                                               \n",
       "SRR332342                        203.06                        3.08   \n",
       "SRR1577515                        30.79                       24.06   \n",
       "SRR1577516                        35.13                       29.51   \n",
       "SRR1577403                         2.89                       56.20   \n",
       "SRR1577404                        36.57                       58.59   \n",
       "\n",
       "            Mb aligned in other small genes  Mb aligned in AceView  \\\n",
       "### Run                                                              \n",
       "SRR332342                               0.0                    0.0   \n",
       "SRR1577515                              0.0                    0.0   \n",
       "SRR1577516                              0.0                    0.0   \n",
       "SRR1577403                              0.0                    0.0   \n",
       "SRR1577404                              0.0                    0.0   \n",
       "\n",
       "            Mb aligned in RefSeq  Mb aligned in Encode  Mb aligned in Magic  \\\n",
       "### Run                                                                       \n",
       "SRR332342                    0.0                246.61                  0.0   \n",
       "SRR1577515                   0.0                395.17                  0.0   \n",
       "SRR1577516                   0.0                408.94                  0.0   \n",
       "SRR1577403                   0.0                356.12                  0.0   \n",
       "SRR1577404                   0.0                407.79                  0.0   \n",
       "\n",
       "            Mb aligned in genome  \\\n",
       "### Run                            \n",
       "SRR332342                 241.21   \n",
       "SRR1577515                368.00   \n",
       "SRR1577516                381.22   \n",
       "SRR1577403                334.04   \n",
       "SRR1577404                383.86   \n",
       "\n",
       "            Mb aligned in any previously annotated transcripts  \\\n",
       "### Run                                                          \n",
       "SRR332342                                              452.69    \n",
       "SRR1577515                                             451.42    \n",
       "SRR1577516                                             475.30    \n",
       "SRR1577403                                             417.23    \n",
       "SRR1577404                                             506.27    \n",
       "\n",
       "            Mb inside introns, new exons and intergenic  \\\n",
       "### Run                                                   \n",
       "SRR332342                                       11559.0   \n",
       "SRR1577515                                       8406.0   \n",
       "SRR1577516                                       9363.0   \n",
       "SRR1577403                                       9248.0   \n",
       "SRR1577404                                      12167.0   \n",
       "\n",
       "            Mb aligned in microbiome  Mb aligned in viruses  \\\n",
       "### Run                                                       \n",
       "SRR332342                        0.0                    0.0   \n",
       "SRR1577515                       0.0                    0.0   \n",
       "SRR1577516                       0.0                    0.0   \n",
       "SRR1577403                       0.0                    0.0   \n",
       "SRR1577404                       0.0                    0.0   \n",
       "\n",
       "            Mb aligned on the imaginary genome (mapping specificity control)  \\\n",
       "### Run                                                                        \n",
       "SRR332342                                                 0.0                  \n",
       "SRR1577515                                                0.0                  \n",
       "SRR1577516                                                0.0                  \n",
       "SRR1577403                                                0.0                  \n",
       "SRR1577404                                                0.0                  \n",
       "\n",
       "            Mb supporting known exons-junctions  Unnamed: 137  \\\n",
       "### Run                                                         \n",
       "SRR332342                                   NaN           NaN   \n",
       "SRR1577515                                  NaN           NaN   \n",
       "SRR1577516                                  NaN           NaN   \n",
       "SRR1577403                                  NaN           NaN   \n",
       "SRR1577404                                  NaN           NaN   \n",
       "\n",
       "            Number of aligned_fragments Compatible pairs Non compatible pairs  \\\n",
       "### Run                                                                         \n",
       "SRR332342                           NaN                -                    -   \n",
       "SRR1577515                          NaN                -                    -   \n",
       "SRR1577516                          NaN                -                    -   \n",
       "SRR1577403                          NaN                -                    -   \n",
       "SRR1577404                          NaN                -                    -   \n",
       "\n",
       "           Paired end fragments with only one read aligned % Compatible pairs  \\\n",
       "### Run                                                                         \n",
       "SRR332342                                                -                  -   \n",
       "SRR1577515                                               -                  -   \n",
       "SRR1577516                                               -                  -   \n",
       "SRR1577403                                               -                  -   \n",
       "SRR1577404                                               -                  -   \n",
       "\n",
       "           % Non compatible pairs  \\\n",
       "### Run                             \n",
       "SRR332342                       -   \n",
       "SRR1577515                      -   \n",
       "SRR1577516                      -   \n",
       "SRR1577403                      -   \n",
       "SRR1577404                      -   \n",
       "\n",
       "           % Paired end fragments with only one read aligned  \\\n",
       "### Run                                                        \n",
       "SRR332342                                                  -   \n",
       "SRR1577515                                                 -   \n",
       "SRR1577516                                                 -   \n",
       "SRR1577403                                                 -   \n",
       "SRR1577404                                                 -   \n",
       "\n",
       "            1% fragments in library are shorter than (nt)  \\\n",
       "### Run                                                     \n",
       "SRR332342                                             NaN   \n",
       "SRR1577515                                            NaN   \n",
       "SRR1577516                                            NaN   \n",
       "SRR1577403                                            NaN   \n",
       "SRR1577404                                            NaN   \n",
       "\n",
       "            5% fragments in library are shorter than (nt)  \\\n",
       "### Run                                                     \n",
       "SRR332342                                             NaN   \n",
       "SRR1577515                                            NaN   \n",
       "SRR1577516                                            NaN   \n",
       "SRR1577403                                            NaN   \n",
       "SRR1577404                                            NaN   \n",
       "\n",
       "            mode of fragment lengths  median fragment lengths  \\\n",
       "### Run                                                         \n",
       "SRR332342                        NaN                      NaN   \n",
       "SRR1577515                       NaN                      NaN   \n",
       "SRR1577516                       NaN                      NaN   \n",
       "SRR1577403                       NaN                      NaN   \n",
       "SRR1577404                       NaN                      NaN   \n",
       "\n",
       "            average of fragment lengths  \\\n",
       "### Run                                   \n",
       "SRR332342                           NaN   \n",
       "SRR1577515                          NaN   \n",
       "SRR1577516                          NaN   \n",
       "SRR1577403                          NaN   \n",
       "SRR1577404                          NaN   \n",
       "\n",
       "            5% fragments in library are longer than (nt)  \\\n",
       "### Run                                                    \n",
       "SRR332342                                            NaN   \n",
       "SRR1577515                                           NaN   \n",
       "SRR1577516                                           NaN   \n",
       "SRR1577403                                           NaN   \n",
       "SRR1577404                                           NaN   \n",
       "\n",
       "            1% fragments in library are longer than (nt)  Unnamed: 152  \\\n",
       "### Run                                                                  \n",
       "SRR332342                                            NaN           NaN   \n",
       "SRR1577515                                           NaN           NaN   \n",
       "SRR1577516                                           NaN           NaN   \n",
       "SRR1577403                                           NaN           NaN   \n",
       "SRR1577404                                           NaN           NaN   \n",
       "\n",
       "            Average length after clipping adaptors and barcodes (nt)  \\\n",
       "### Run                                                                \n",
       "SRR332342                                               35.72          \n",
       "SRR1577515                                              75.82          \n",
       "SRR1577516                                              75.81          \n",
       "SRR1577403                                              75.82          \n",
       "SRR1577404                                              75.82          \n",
       "\n",
       "            Average length aligned (nt)  \\\n",
       "### Run                                   \n",
       "SRR332342                         35.17   \n",
       "SRR1577515                        74.71   \n",
       "SRR1577516                        74.72   \n",
       "SRR1577403                        74.95   \n",
       "SRR1577404                        74.96   \n",
       "\n",
       "            Average length aligned in read 1 (nt)  \\\n",
       "### Run                                             \n",
       "SRR332342                                   35.17   \n",
       "SRR1577515                                  74.71   \n",
       "SRR1577516                                  74.72   \n",
       "SRR1577403                                  74.95   \n",
       "SRR1577404                                  74.96   \n",
       "\n",
       "            Average length aligned in read 2 (paired end)  \\\n",
       "### Run                                                     \n",
       "SRR332342                                             NaN   \n",
       "SRR1577515                                            NaN   \n",
       "SRR1577516                                            NaN   \n",
       "SRR1577403                                            NaN   \n",
       "SRR1577404                                            NaN   \n",
       "\n",
       "            Average length aligned on RefSeq (nt)  \\\n",
       "### Run                                             \n",
       "SRR332342                                     NaN   \n",
       "SRR1577515                                    NaN   \n",
       "SRR1577516                                    NaN   \n",
       "SRR1577403                                    NaN   \n",
       "SRR1577404                                    NaN   \n",
       "\n",
       "            Average length aligned on EBI (nt)  \\\n",
       "### Run                                          \n",
       "SRR332342                                34.92   \n",
       "SRR1577515                               74.65   \n",
       "SRR1577516                               74.65   \n",
       "SRR1577403                               74.87   \n",
       "SRR1577404                               74.84   \n",
       "\n",
       "            Average length aligned on AceView (nt)  \\\n",
       "### Run                                              \n",
       "SRR332342                                      NaN   \n",
       "SRR1577515                                     NaN   \n",
       "SRR1577516                                     NaN   \n",
       "SRR1577403                                     NaN   \n",
       "SRR1577404                                     NaN   \n",
       "\n",
       "            Average length aligned on magic (nt)  \\\n",
       "### Run                                            \n",
       "SRR332342                                    NaN   \n",
       "SRR1577515                                   NaN   \n",
       "SRR1577516                                   NaN   \n",
       "SRR1577403                                   NaN   \n",
       "SRR1577404                                   NaN   \n",
       "\n",
       "            Average length aligned on Other (nt)  \\\n",
       "### Run                                            \n",
       "SRR332342                                    NaN   \n",
       "SRR1577515                                   NaN   \n",
       "SRR1577516                                   NaN   \n",
       "SRR1577403                                   NaN   \n",
       "SRR1577404                                   NaN   \n",
       "\n",
       "            Average length aligned on Genome (nt)  \\\n",
       "### Run                                             \n",
       "SRR332342                                   35.16   \n",
       "SRR1577515                                  74.91   \n",
       "SRR1577516                                  74.91   \n",
       "SRR1577403                                  75.15   \n",
       "SRR1577404                                  75.11   \n",
       "\n",
       "            Average length aligned on Mitochondria (nt)  \\\n",
       "### Run                                                   \n",
       "SRR332342                                         35.53   \n",
       "SRR1577515                                        75.21   \n",
       "SRR1577516                                        75.21   \n",
       "SRR1577403                                        75.50   \n",
       "SRR1577404                                        75.47   \n",
       "\n",
       "            Average length aligned on rRNA (nt)  \\\n",
       "### Run                                           \n",
       "SRR332342                                 35.48   \n",
       "SRR1577515                                75.39   \n",
       "SRR1577516                                75.39   \n",
       "SRR1577403                                75.43   \n",
       "SRR1577404                                75.59   \n",
       "\n",
       "            Average length aligned on Chloroplast (nt)  \\\n",
       "### Run                                                  \n",
       "SRR332342                                          NaN   \n",
       "SRR1577515                                         NaN   \n",
       "SRR1577516                                         NaN   \n",
       "SRR1577403                                         NaN   \n",
       "SRR1577404                                         NaN   \n",
       "\n",
       "            Average length aligned on small RNAs (nt)  \\\n",
       "### Run                                                 \n",
       "SRR332342                                         NaN   \n",
       "SRR1577515                                        NaN   \n",
       "SRR1577516                                        NaN   \n",
       "SRR1577403                                        NaN   \n",
       "SRR1577404                                        NaN   \n",
       "\n",
       "            Average length aligned on DNA spikeIn (nt)  \\\n",
       "### Run                                                  \n",
       "SRR332342                                          NaN   \n",
       "SRR1577515                                       74.79   \n",
       "SRR1577516                                       74.77   \n",
       "SRR1577403                                       75.38   \n",
       "SRR1577404                                       75.41   \n",
       "\n",
       "            Average length aligned on RNA spikeIn (nt)  \\\n",
       "### Run                                                  \n",
       "SRR332342                                          NaN   \n",
       "SRR1577515                                       75.35   \n",
       "SRR1577516                                       75.33   \n",
       "SRR1577403                                       75.58   \n",
       "SRR1577404                                       75.55   \n",
       "\n",
       "            Average length aligned on Imaginary genome specificity control (nt)  \\\n",
       "### Run                                                                           \n",
       "SRR332342                                                 NaN                     \n",
       "SRR1577515                                                NaN                     \n",
       "SRR1577516                                                NaN                     \n",
       "SRR1577403                                                NaN                     \n",
       "SRR1577404                                                NaN                     \n",
       "\n",
       "            Unnamed: 170  Average % length aligned (nt)  \\\n",
       "### Run                                                   \n",
       "SRR332342            NaN                          98.46   \n",
       "SRR1577515           NaN                          98.54   \n",
       "SRR1577516           NaN                          98.56   \n",
       "SRR1577403           NaN                          98.85   \n",
       "SRR1577404           NaN                          98.87   \n",
       "\n",
       "            Average % length aligned in read 1 (nt)  \\\n",
       "### Run                                               \n",
       "SRR332342                                     98.46   \n",
       "SRR1577515                                    98.54   \n",
       "SRR1577516                                    98.56   \n",
       "SRR1577403                                    98.85   \n",
       "SRR1577404                                    98.87   \n",
       "\n",
       "            Average % length aligned in read 2 (paired end)  \\\n",
       "### Run                                                       \n",
       "SRR332342                                               NaN   \n",
       "SRR1577515                                              NaN   \n",
       "SRR1577516                                              NaN   \n",
       "SRR1577403                                              NaN   \n",
       "SRR1577404                                              NaN   \n",
       "\n",
       "            Average % length aligned on RefSeq  \\\n",
       "### Run                                          \n",
       "SRR332342                                  NaN   \n",
       "SRR1577515                                 NaN   \n",
       "SRR1577516                                 NaN   \n",
       "SRR1577403                                 NaN   \n",
       "SRR1577404                                 NaN   \n",
       "\n",
       "            Average % length aligned on EBI  \\\n",
       "### Run                                       \n",
       "SRR332342                             98.03   \n",
       "SRR1577515                            98.50   \n",
       "SRR1577516                            98.51   \n",
       "SRR1577403                            98.79   \n",
       "SRR1577404                            98.75   \n",
       "\n",
       "            Average % length aligned on AceView  \\\n",
       "### Run                                           \n",
       "SRR332342                                   NaN   \n",
       "SRR1577515                                  NaN   \n",
       "SRR1577516                                  NaN   \n",
       "SRR1577403                                  NaN   \n",
       "SRR1577404                                  NaN   \n",
       "\n",
       "            Average % length aligned on magic  \\\n",
       "### Run                                         \n",
       "SRR332342                                 NaN   \n",
       "SRR1577515                                NaN   \n",
       "SRR1577516                                NaN   \n",
       "SRR1577403                                NaN   \n",
       "SRR1577404                                NaN   \n",
       "\n",
       "            Average % length aligned on Other  \\\n",
       "### Run                                         \n",
       "SRR332342                                 NaN   \n",
       "SRR1577515                                NaN   \n",
       "SRR1577516                                NaN   \n",
       "SRR1577403                                NaN   \n",
       "SRR1577404                                NaN   \n",
       "\n",
       "            Average % length aligned on Genome  \\\n",
       "### Run                                          \n",
       "SRR332342                               98.710   \n",
       "SRR1577515                              98.830   \n",
       "SRR1577516                              98.840   \n",
       "SRR1577403                              99.156   \n",
       "SRR1577404                              99.103   \n",
       "\n",
       "            Average % length aligned on Mitochondria  \\\n",
       "### Run                                                \n",
       "SRR332342                                     98.830   \n",
       "SRR1577515                                    99.026   \n",
       "SRR1577516                                    99.026   \n",
       "SRR1577403                                    99.408   \n",
       "SRR1577404                                    99.368   \n",
       "\n",
       "            Average % length aligned on rRNA  \\\n",
       "### Run                                        \n",
       "SRR332342                             98.970   \n",
       "SRR1577515                            99.223   \n",
       "SRR1577516                            99.237   \n",
       "SRR1577403                            99.289   \n",
       "SRR1577404                            99.500   \n",
       "\n",
       "            Average length aligned on Chloroplast  \\\n",
       "### Run                                             \n",
       "SRR332342                                     NaN   \n",
       "SRR1577515                                    NaN   \n",
       "SRR1577516                                    NaN   \n",
       "SRR1577403                                    NaN   \n",
       "SRR1577404                                    NaN   \n",
       "\n",
       "            Average length aligned on small RNAs  \\\n",
       "### Run                                            \n",
       "SRR332342                                    NaN   \n",
       "SRR1577515                                   NaN   \n",
       "SRR1577516                                   NaN   \n",
       "SRR1577403                                   NaN   \n",
       "SRR1577404                                   NaN   \n",
       "\n",
       "            Average length aligned on DNA spikeIn  \\\n",
       "### Run                                             \n",
       "SRR332342                                     NaN   \n",
       "SRR1577515                                 98.450   \n",
       "SRR1577516                                 98.420   \n",
       "SRR1577403                                 99.236   \n",
       "SRR1577404                                 99.276   \n",
       "\n",
       "            Average length aligned on RNA spikeIn  \\\n",
       "### Run                                             \n",
       "SRR332342                                     NaN   \n",
       "SRR1577515                                 99.184   \n",
       "SRR1577516                                 99.158   \n",
       "SRR1577403                                 99.474   \n",
       "SRR1577404                                 99.447   \n",
       "\n",
       "            Average length aligned on Imaginary genome specificity control  \\\n",
       "### Run                                                                      \n",
       "SRR332342                                                 NaN                \n",
       "SRR1577515                                                NaN                \n",
       "SRR1577516                                                NaN                \n",
       "SRR1577403                                                NaN                \n",
       "SRR1577404                                                NaN                \n",
       "\n",
       "            Unnamed: 187  Reads mapping on plus strand of rRNA  \\\n",
       "### Run                                                          \n",
       "SRR332342            NaN                             2954663.0   \n",
       "SRR1577515           NaN                              206934.0   \n",
       "SRR1577516           NaN                              233089.0   \n",
       "SRR1577403           NaN                               19377.0   \n",
       "SRR1577404           NaN                              243757.0   \n",
       "\n",
       "            Reads mapping on minus strand of rRNA  \\\n",
       "### Run                                             \n",
       "SRR332342                               2768588.0   \n",
       "SRR1577515                               201458.0   \n",
       "SRR1577516                               232879.0   \n",
       "SRR1577403                                18948.0   \n",
       "SRR1577404                               240015.0   \n",
       "\n",
       "            Reads mapping on both strands of rRNA  \\\n",
       "### Run                                             \n",
       "SRR332342                                     0.0   \n",
       "SRR1577515                                    0.0   \n",
       "SRR1577516                                    0.0   \n",
       "SRR1577403                                    0.0   \n",
       "SRR1577404                                    0.0   \n",
       "\n",
       "            % reads mapping on plus strand of rRNA  \\\n",
       "### Run                                              \n",
       "SRR332342                                  51.6300   \n",
       "SRR1577515                                 50.6700   \n",
       "SRR1577516                                 50.0225   \n",
       "SRR1577403                                 50.5600   \n",
       "SRR1577404                                 50.3870   \n",
       "\n",
       "            % reads mapping on minus strand of rRNA  \\\n",
       "### Run                                               \n",
       "SRR332342                                   48.3700   \n",
       "SRR1577515                                  49.3300   \n",
       "SRR1577516                                  49.9775   \n",
       "SRR1577403                                  49.4400   \n",
       "SRR1577404                                  49.6130   \n",
       "\n",
       "            % reads mapping on both strand of rRNA  \\\n",
       "### Run                                              \n",
       "SRR332342                                      0.0   \n",
       "SRR1577515                                     0.0   \n",
       "SRR1577516                                     0.0   \n",
       "SRR1577403                                     0.0   \n",
       "SRR1577404                                     0.0   \n",
       "\n",
       "            Reads mapping on plus strand of RefSeq  \\\n",
       "### Run                                              \n",
       "SRR332342                                      NaN   \n",
       "SRR1577515                                     NaN   \n",
       "SRR1577516                                     NaN   \n",
       "SRR1577403                                     NaN   \n",
       "SRR1577404                                     NaN   \n",
       "\n",
       "            Reads mapping on minus strand of RefSeq  \\\n",
       "### Run                                               \n",
       "SRR332342                                       NaN   \n",
       "SRR1577515                                      NaN   \n",
       "SRR1577516                                      NaN   \n",
       "SRR1577403                                      NaN   \n",
       "SRR1577404                                      NaN   \n",
       "\n",
       "            Reads mapping on both strands of RefSeq  \\\n",
       "### Run                                               \n",
       "SRR332342                                       NaN   \n",
       "SRR1577515                                      NaN   \n",
       "SRR1577516                                      NaN   \n",
       "SRR1577403                                      NaN   \n",
       "SRR1577404                                      NaN   \n",
       "\n",
       "            % reads mapping on plus strand of RefSeq  \\\n",
       "### Run                                                \n",
       "SRR332342                                        NaN   \n",
       "SRR1577515                                       NaN   \n",
       "SRR1577516                                       NaN   \n",
       "SRR1577403                                       NaN   \n",
       "SRR1577404                                       NaN   \n",
       "\n",
       "            % reads mapping on minus strand of RefSeq  \\\n",
       "### Run                                                 \n",
       "SRR332342                                         NaN   \n",
       "SRR1577515                                        NaN   \n",
       "SRR1577516                                        NaN   \n",
       "SRR1577403                                        NaN   \n",
       "SRR1577404                                        NaN   \n",
       "\n",
       "            % reads mapping on both strand of RefSeq  \\\n",
       "### Run                                                \n",
       "SRR332342                                        NaN   \n",
       "SRR1577515                                       NaN   \n",
       "SRR1577516                                       NaN   \n",
       "SRR1577403                                       NaN   \n",
       "SRR1577404                                       NaN   \n",
       "\n",
       "            Reads mapping on plus strand of EBI  \\\n",
       "### Run                                           \n",
       "SRR332342                             3541738.0   \n",
       "SRR1577515                            2631597.0   \n",
       "SRR1577516                            2705334.0   \n",
       "SRR1577403                            2372479.0   \n",
       "SRR1577404                            2690006.0   \n",
       "\n",
       "            Reads mapping on minus strand of EBI  \\\n",
       "### Run                                            \n",
       "SRR332342                              3395040.0   \n",
       "SRR1577515                             2536482.0   \n",
       "SRR1577516                             2611876.0   \n",
       "SRR1577403                             2307496.0   \n",
       "SRR1577404                             2647449.0   \n",
       "\n",
       "            Reads mapping on both strands of EBI  \\\n",
       "### Run                                            \n",
       "SRR332342                               124541.0   \n",
       "SRR1577515                              125799.0   \n",
       "SRR1577516                              161150.0   \n",
       "SRR1577403                               76647.0   \n",
       "SRR1577404                              111184.0   \n",
       "\n",
       "            % reads mapping on plus strand of EBI  \\\n",
       "### Run                                             \n",
       "SRR332342                                  50.157   \n",
       "SRR1577515                                 49.710   \n",
       "SRR1577516                                 49.382   \n",
       "SRR1577403                                 49.877   \n",
       "SRR1577404                                 49.370   \n",
       "\n",
       "            % reads mapping on minus strand of EBI  \\\n",
       "### Run                                              \n",
       "SRR332342                                    48.08   \n",
       "SRR1577515                                   47.91   \n",
       "SRR1577516                                   47.68   \n",
       "SRR1577403                                   48.51   \n",
       "SRR1577404                                   48.59   \n",
       "\n",
       "            % reads mapping on both strand of EBI  \\\n",
       "### Run                                             \n",
       "SRR332342                                    1.76   \n",
       "SRR1577515                                   2.38   \n",
       "SRR1577516                                   2.94   \n",
       "SRR1577403                                   1.61   \n",
       "SRR1577404                                   2.04   \n",
       "\n",
       "            Reads mapping on plus strand of av  \\\n",
       "### Run                                          \n",
       "SRR332342                                  NaN   \n",
       "SRR1577515                                 NaN   \n",
       "SRR1577516                                 NaN   \n",
       "SRR1577403                                 NaN   \n",
       "SRR1577404                                 NaN   \n",
       "\n",
       "            Reads mapping on minus strand of av  \\\n",
       "### Run                                           \n",
       "SRR332342                                   NaN   \n",
       "SRR1577515                                  NaN   \n",
       "SRR1577516                                  NaN   \n",
       "SRR1577403                                  NaN   \n",
       "SRR1577404                                  NaN   \n",
       "\n",
       "            Reads mapping on both strands of av  \\\n",
       "### Run                                           \n",
       "SRR332342                                   NaN   \n",
       "SRR1577515                                  NaN   \n",
       "SRR1577516                                  NaN   \n",
       "SRR1577403                                  NaN   \n",
       "SRR1577404                                  NaN   \n",
       "\n",
       "            % reads mapping on plus strand of av  \\\n",
       "### Run                                            \n",
       "SRR332342                                    NaN   \n",
       "SRR1577515                                   NaN   \n",
       "SRR1577516                                   NaN   \n",
       "SRR1577403                                   NaN   \n",
       "SRR1577404                                   NaN   \n",
       "\n",
       "            % reads mapping on minus strand of av  \\\n",
       "### Run                                             \n",
       "SRR332342                                     NaN   \n",
       "SRR1577515                                    NaN   \n",
       "SRR1577516                                    NaN   \n",
       "SRR1577403                                    NaN   \n",
       "SRR1577404                                    NaN   \n",
       "\n",
       "            % reads mapping on both strand of av  \\\n",
       "### Run                                            \n",
       "SRR332342                                    NaN   \n",
       "SRR1577515                                   NaN   \n",
       "SRR1577516                                   NaN   \n",
       "SRR1577403                                   NaN   \n",
       "SRR1577404                                   NaN   \n",
       "\n",
       "            Reads mapping on plus strand of genome  \\\n",
       "### Run                                              \n",
       "SRR332342                                3431228.0   \n",
       "SRR1577515                               2457799.0   \n",
       "SRR1577516                               2547121.0   \n",
       "SRR1577403                               2220549.0   \n",
       "SRR1577404                               2554700.0   \n",
       "\n",
       "            Reads mapping on minus strand of genome  \\\n",
       "### Run                                               \n",
       "SRR332342                                 3428466.0   \n",
       "SRR1577515                                2454916.0   \n",
       "SRR1577516                                2542121.0   \n",
       "SRR1577403                                2224568.0   \n",
       "SRR1577404                                2555627.0   \n",
       "\n",
       "            Reads mapping on both strands of genome  \\\n",
       "### Run                                               \n",
       "SRR332342                                       0.0   \n",
       "SRR1577515                                      0.0   \n",
       "SRR1577516                                      0.0   \n",
       "SRR1577403                                      0.0   \n",
       "SRR1577404                                      0.0   \n",
       "\n",
       "            % reads mapping on plus strand of genome  \\\n",
       "### Run                                                \n",
       "SRR332342                                   50.02010   \n",
       "SRR1577515                                  50.02930   \n",
       "SRR1577516                                  50.04910   \n",
       "SRR1577403                                  49.95480   \n",
       "SRR1577404                                  49.99093   \n",
       "\n",
       "            % reads mapping on minus strand of genome  \\\n",
       "### Run                                                 \n",
       "SRR332342                                    49.97990   \n",
       "SRR1577515                                   49.97070   \n",
       "SRR1577516                                   49.95090   \n",
       "SRR1577403                                   50.04520   \n",
       "SRR1577404                                   50.00907   \n",
       "\n",
       "            % reads mapping on both strand of genome  \\\n",
       "### Run                                                \n",
       "SRR332342                                        0.0   \n",
       "SRR1577515                                       0.0   \n",
       "SRR1577516                                       0.0   \n",
       "SRR1577403                                       0.0   \n",
       "SRR1577404                                       0.0   \n",
       "\n",
       "            Reads mapping on plus strand of SpikeIn  \\\n",
       "### Run                                               \n",
       "SRR332342                                       NaN   \n",
       "SRR1577515                                   9657.0   \n",
       "SRR1577516                                  11831.0   \n",
       "SRR1577403                                  13321.0   \n",
       "SRR1577404                                  22251.0   \n",
       "\n",
       "            Reads mapping on minus strand of SpikeIn  \\\n",
       "### Run                                                \n",
       "SRR332342                                        NaN   \n",
       "SRR1577515                                    9618.0   \n",
       "SRR1577516                                   11839.0   \n",
       "SRR1577403                                   13488.0   \n",
       "SRR1577404                                   22718.0   \n",
       "\n",
       "            Reads mapping on both strands of SpikeIn  \\\n",
       "### Run                                                \n",
       "SRR332342                                        NaN   \n",
       "SRR1577515                                       0.0   \n",
       "SRR1577516                                       0.0   \n",
       "SRR1577403                                       0.0   \n",
       "SRR1577404                                       0.0   \n",
       "\n",
       "            % reads mapping on plus strand of SpikeIn  \\\n",
       "### Run                                                 \n",
       "SRR332342                                         NaN   \n",
       "SRR1577515                                    50.1010   \n",
       "SRR1577516                                    49.9831   \n",
       "SRR1577403                                    49.6890   \n",
       "SRR1577404                                    49.4810   \n",
       "\n",
       "            % reads mapping on minus strand of SpikeIn  \\\n",
       "### Run                                                  \n",
       "SRR332342                                          NaN   \n",
       "SRR1577515                                     49.8990   \n",
       "SRR1577516                                     50.0169   \n",
       "SRR1577403                                     50.3110   \n",
       "SRR1577404                                     50.5190   \n",
       "\n",
       "            % reads mapping on both strand of SpikeIn  Unnamed: 224  \\\n",
       "### Run                                                               \n",
       "SRR332342                                         NaN           NaN   \n",
       "SRR1577515                                        0.0           NaN   \n",
       "SRR1577516                                        0.0           NaN   \n",
       "SRR1577403                                        0.0           NaN   \n",
       "SRR1577404                                        0.0           NaN   \n",
       "\n",
       "            Megabases uniquely aligned and used for mismatch counts  \\\n",
       "### Run                                                               \n",
       "SRR332342                                                 439         \n",
       "SRR1577515                                                432         \n",
       "SRR1577516                                                454         \n",
       "SRR1577403                                                402         \n",
       "SRR1577404                                                489         \n",
       "\n",
       "            Total number of mismatches  Mismatches per kb aligned  \\\n",
       "### Run                                                             \n",
       "SRR332342                       587469                    1.33820   \n",
       "SRR1577515                     1141232                    2.64174   \n",
       "SRR1577516                     1171627                    2.58068   \n",
       "SRR1577403                     1037681                    2.58130   \n",
       "SRR1577404                     1241053                    2.53794   \n",
       "\n",
       "            Transitions  Transversions  \\\n",
       "### Run                                  \n",
       "SRR332342        287857         281897   \n",
       "SRR1577515       625138         485026   \n",
       "SRR1577516       640940         493785   \n",
       "SRR1577403       588815         406331   \n",
       "SRR1577404       672545         510945   \n",
       "\n",
       "            1, 2 or 3 bp insertions not in polymers  \\\n",
       "### Run                                               \n",
       "SRR332342                                       410   \n",
       "SRR1577515                                     1068   \n",
       "SRR1577516                                     1183   \n",
       "SRR1577403                                     1151   \n",
       "SRR1577404                                     1508   \n",
       "\n",
       "            1, 2 or 3 bp deletions not in polymers  \\\n",
       "### Run                                              \n",
       "SRR332342                                     4186   \n",
       "SRR1577515                                    5549   \n",
       "SRR1577516                                    6558   \n",
       "SRR1577403                                    6837   \n",
       "SRR1577404                                   11826   \n",
       "\n",
       "            1, 2 or 3 bp insertions in polymers  \\\n",
       "### Run                                           \n",
       "SRR332342                                  7043   \n",
       "SRR1577515                                10684   \n",
       "SRR1577516                                12604   \n",
       "SRR1577403                                10067   \n",
       "SRR1577404                                15243   \n",
       "\n",
       "            1, 2 or 3 bp deletions in polymers     A>G     T>C     G>A  \\\n",
       "### Run                                                                  \n",
       "SRR332342                                 6076   90084   80492   51011   \n",
       "SRR1577515                               13767  170205  175628  139466   \n",
       "SRR1577516                               16557  177008  183103  139647   \n",
       "SRR1577403                               24480  152444  184047  124325   \n",
       "SRR1577404                               28986  179191  207480  140849   \n",
       "\n",
       "               C>T    A>T    T>A    G>C    C>G    A>C    T>G    G>T    C>A  \\\n",
       "### Run                                                                      \n",
       "SRR332342    66270  32086  38624  18388  21903  59697  36659  30693  43847   \n",
       "SRR1577515  139839  44288  55822  46203  55142  46710  84563  74624  77674   \n",
       "SRR1577516  141182  44785  56011  45952  56248  49660  88413  73955  78761   \n",
       "SRR1577403  127999  39249  44398  43001  49927  37868  66661  63349  61878   \n",
       "SRR1577404  145025  49932  57949  51290  61015  46625  86882  78678  78574   \n",
       "\n",
       "            Single Insertions  Single Deletions  Double insertions  \\\n",
       "### Run                                                              \n",
       "SRR332342                2210              8138               5228   \n",
       "SRR1577515               8028             15504               2809   \n",
       "SRR1577516               9641             18363               3042   \n",
       "SRR1577403               8966             26842               1385   \n",
       "SRR1577404              12395             33659               3367   \n",
       "\n",
       "            Double deletions  Triple insertions  Triple deletions  \\\n",
       "### Run                                                             \n",
       "SRR332342               1988                 15               136   \n",
       "SRR1577515              2959                915               853   \n",
       "SRR1577516              3777               1104               975   \n",
       "SRR1577403              3518                867               957   \n",
       "SRR1577404              5844                989              1309   \n",
       "\n",
       "            Transitions per kb  Transversions per kb  Insertions per kb  \\\n",
       "### Run                                                                   \n",
       "SRR332342              0.65571               0.64213            0.00093   \n",
       "SRR1577515             1.44708               1.12275            0.00247   \n",
       "SRR1577516             1.41176               1.08763            0.00261   \n",
       "SRR1577403             1.46471               1.01077            0.00286   \n",
       "SRR1577404             1.37535               1.04488            0.00308   \n",
       "\n",
       "            Deletions per kb  % transitions  % transversions  \\\n",
       "### Run                                                        \n",
       "SRR332342            0.00954          49.00            47.99   \n",
       "SRR1577515           0.01284          54.78            42.50   \n",
       "SRR1577516           0.01444          54.71            42.15   \n",
       "SRR1577403           0.01701          56.74            39.16   \n",
       "SRR1577404           0.02418          54.19            41.17   \n",
       "\n",
       "            % insertions not in polymers  % deletions not in polymers  \\\n",
       "### Run                                                                 \n",
       "SRR332342                           0.07                         0.71   \n",
       "SRR1577515                          0.09                         0.49   \n",
       "SRR1577516                          0.10                         0.56   \n",
       "SRR1577403                          0.11                         0.66   \n",
       "SRR1577404                          0.12                         0.95   \n",
       "\n",
       "            % insertions in polymers  % deletions in polymers  Unnamed: 262  \\\n",
       "### Run                                                                       \n",
       "SRR332342                       1.20                     1.03           NaN   \n",
       "SRR1577515                      0.94                     1.21           NaN   \n",
       "SRR1577516                      1.08                     1.41           NaN   \n",
       "SRR1577403                      0.97                     2.36           NaN   \n",
       "SRR1577404                      1.23                     2.34           NaN   \n",
       "\n",
       "            Accessible transcript length, estimated on transcripts longer than 8kb, usually limited by the 3' biais in poly-A selected experiments  \\\n",
       "### Run                                                                                                                                              \n",
       "SRR332342                                              4800.0                                                                                        \n",
       "SRR1577515                                             4350.0                                                                                        \n",
       "SRR1577516                                             4400.0                                                                                        \n",
       "SRR1577403                                             3150.0                                                                                        \n",
       "SRR1577404                                             3200.0                                                                                        \n",
       "\n",
       "            Number of well expressed transcripts longer than 8kb  \\\n",
       "### Run                                                            \n",
       "SRR332342                                                34.0      \n",
       "SRR1577515                                               42.0      \n",
       "SRR1577516                                               45.0      \n",
       "SRR1577403                                               60.0      \n",
       "SRR1577404                                               64.0      \n",
       "\n",
       "            Average coverage cumulated over these transcripts  Unnamed: 266  \\\n",
       "### Run                                                                       \n",
       "SRR332342                                               345.0           NaN   \n",
       "SRR1577515                                              528.0           NaN   \n",
       "SRR1577516                                              608.0           NaN   \n",
       "SRR1577403                                              483.0           NaN   \n",
       "SRR1577404                                              614.0           NaN   \n",
       "\n",
       "            Intergenic kb  Intergenic density  Unnamed: 269  Zero index  \\\n",
       "### Run                                                                   \n",
       "SRR332342             377              0.0107           NaN         9.3   \n",
       "SRR1577515            339              0.0096           NaN         9.7   \n",
       "SRR1577516            381              0.0108           NaN         9.7   \n",
       "SRR1577403            329              0.0093           NaN         9.9   \n",
       "SRR1577404            391              0.0111           NaN         9.7   \n",
       "\n",
       "            Low index  Cross over index  Genes touched  \\\n",
       "### Run                                                  \n",
       "SRR332342         9.4               9.9        12795.0   \n",
       "SRR1577515        9.4              10.1        12244.0   \n",
       "SRR1577516        9.5              10.1        12420.0   \n",
       "SRR1577403       10.1              10.4        12433.0   \n",
       "SRR1577404        9.8              10.2        12643.0   \n",
       "\n",
       "            Genes with reliable index  Genes with index >= 10  \\\n",
       "### Run                                                         \n",
       "SRR332342                     10772.0                 10193.0   \n",
       "SRR1577515                    10038.0                  9500.0   \n",
       "SRR1577516                    10168.0                  9643.0   \n",
       "SRR1577403                    10319.0                 10041.0   \n",
       "SRR1577404                    10381.0                  9988.0   \n",
       "\n",
       "            Genes with index >= 12  Genes with index >= 15  \\\n",
       "### Run                                                      \n",
       "SRR332342                   8737.0                  3973.0   \n",
       "SRR1577515                  8213.0                  3195.0   \n",
       "SRR1577516                  8217.0                  3102.0   \n",
       "SRR1577403                  8690.0                  3334.0   \n",
       "SRR1577404                  8549.0                  3210.0   \n",
       "\n",
       "            Genes with index >= 18  Genes expressed in at least one run  \\\n",
       "### Run                                                                   \n",
       "SRR332342                    428.0                              10772.0   \n",
       "SRR1577515                   447.0                              10038.0   \n",
       "SRR1577516                   438.0                              10168.0   \n",
       "SRR1577403                   481.0                              10319.0   \n",
       "SRR1577404                   434.0                              10381.0   \n",
       "\n",
       "            RefSeq genes significantly expressed in at least one run  \\\n",
       "### Run                                                                \n",
       "SRR332342                                                 NaN          \n",
       "SRR1577515                                                NaN          \n",
       "SRR1577516                                                NaN          \n",
       "SRR1577403                                                NaN          \n",
       "SRR1577404                                                NaN          \n",
       "\n",
       "            RefSeq genes significantly expressed in majority of runs  \\\n",
       "### Run                                                                \n",
       "SRR332342                                                 NaN          \n",
       "SRR1577515                                                NaN          \n",
       "SRR1577516                                                NaN          \n",
       "SRR1577403                                                NaN          \n",
       "SRR1577404                                                NaN          \n",
       "\n",
       "            ENCODE genes significantly expressed in at least one run  \\\n",
       "### Run                                                                \n",
       "SRR332342                                             10772.0          \n",
       "SRR1577515                                            10038.0          \n",
       "SRR1577516                                            10168.0          \n",
       "SRR1577403                                            10319.0          \n",
       "SRR1577404                                            10381.0          \n",
       "\n",
       "            ENCODE genes significantly expressed in majority of runs  \\\n",
       "### Run                                                                \n",
       "SRR332342                                             10772.0          \n",
       "SRR1577515                                            10038.0          \n",
       "SRR1577516                                            10168.0          \n",
       "SRR1577403                                            10319.0          \n",
       "SRR1577404                                            10381.0          \n",
       "\n",
       "            AceView genes significantly expressed in at least one run  \\\n",
       "### Run                                                                 \n",
       "SRR332342                                             10772.0           \n",
       "SRR1577515                                            10038.0           \n",
       "SRR1577516                                            10168.0           \n",
       "SRR1577403                                            10319.0           \n",
       "SRR1577404                                            10381.0           \n",
       "\n",
       "            AceView genes significantly expressed in majority of runs  \\\n",
       "### Run                                                                 \n",
       "SRR332342                                             10772.0           \n",
       "SRR1577515                                            10038.0           \n",
       "SRR1577516                                            10168.0           \n",
       "SRR1577403                                            10319.0           \n",
       "SRR1577404                                            10381.0           \n",
       "\n",
       "            AceView transcripts significantly expressed in  at least one run  \\\n",
       "### Run                                                                        \n",
       "SRR332342                                                 NaN                  \n",
       "SRR1577515                                                NaN                  \n",
       "SRR1577516                                                NaN                  \n",
       "SRR1577403                                                NaN                  \n",
       "SRR1577404                                                NaN                  \n",
       "\n",
       "             Mb_aligned   Mb_in_genes  \\\n",
       "### Run                                 \n",
       "SRR332342       464.254       246.611   \n",
       "SRR1577515      460.342       395.168   \n",
       "SRR1577516      485.088       408.939   \n",
       "SRR1577403      427.194       356.124   \n",
       "SRR1577404      519.522       407.787   \n",
       "\n",
       "            Mb_in_genes_with_GeneId_minus_high_genes  Mb_in_high_genes  \\\n",
       "### Run                                                                  \n",
       "SRR332342                                    229.655           0.00000   \n",
       "SRR1577515                                   371.811           0.00000   \n",
       "SRR1577516                                   375.547           8.19641   \n",
       "SRR1577403                                   337.417           0.00000   \n",
       "SRR1577404                                   385.491           0.00000   \n",
       "\n",
       "            Intergenic  \n",
       "### Run                 \n",
       "SRR332342          377  \n",
       "SRR1577515         339  \n",
       "SRR1577516         381  \n",
       "SRR1577403         329  \n",
       "SRR1577404         391  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# I got this table from Jean and Danielle\n",
    "dfj = pd.read_table('../../data/jean/Mapping_QC/dd_any.Data_Summary.runs.txt', skiprows=2, low_memory=False, index_col=0)\n",
    "print(dfj.shape)\n",
    "dfj.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false,
    "hidden": true,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unnamed: 1\n",
      "Genes significantly expressed\n",
      "% fragments on strand plus of genes\n",
      "Unnamed: 4\n",
      "RunId \n",
      "Run accession in SRA\n",
      "Title\n",
      "Other title\n",
      "Sorting_title 1\n",
      "Sorting title 2\n",
      "Sample (summarized from biosample or manual)\n",
      "Stage and experimental summary\n",
      "System or Tissue\n",
      "Laboratory, Author when available, submission date\n",
      "Species\n",
      "Project\n",
      "Project description\n",
      "Reference\n",
      "Unnamed: 19\n",
      "Sequencing protocol\n",
      "Design\n",
      "Sequencing platform and Machine type\n",
      "Experiment type\n",
      "Sequencing date\n",
      "Submission date\n",
      "Library preparation\n",
      "RNA Integrity Number\n",
      "Details\n",
      "Corresponding microarray\n",
      "Spots, bases, average read length (in SRA)\n",
      "Single or Paired end\n",
      "Unnamed: 32\n",
      "Average read length (nt)\n",
      "Fragment multiplicity\n",
      "Million distinct sequences\n",
      "Million raw reads\n",
      "Megabases sequenced\n",
      "%A\n",
      "%T\n",
      "%G\n",
      "%C\n",
      "%N\n",
      "%GC\n",
      "Unnamed: 44\n",
      "Exit adaptor of read 1\n",
      "Exit adaptor of read 2\n",
      "PolyA index : fragments per million with at least 3 A13 motifs shifted by at least 6 bases\n",
      "Telomere index : fragments per million with at least 3 TTAGGG/CCCTAA motifs\n",
      "Telomere index allowing G/C substitution\n",
      "Noise in telomere index : FPM with at least 3 AATCCC/GGGATT motifs\n",
      "Unnamed: 51\n",
      "Million reads aligned on all targets per CPU hour\n",
      "Maximum RAM used (GB)\n",
      "Number of blocks\n",
      "Unnamed: 55\n",
      "Million reads aligned on any target\n",
      "Average length aligned per read (nt)\n",
      "Mb aligned on any target\n",
      "%Reads aligned on any target\n",
      "%Mb aligned on any target\n",
      "% length aligned on average\n",
      "Unnamed: 62\n",
      "Raw reads\n",
      "Well mapped reads\n",
      "Reads mapping uniquely  to a single locus (genomic site) and maximum 1 gene\n",
      "Reads mapping uniquely to a single locus, but to 2 genes in antisense \n",
      "Reads mapping uniquely to a single locus\n",
      "Reads mapping to 2 to 9 sites\n",
      "Reads rejected because they map to 10 sites or more\n",
      "Reads rejected by alignment quality filter\n",
      "Reads rejected before alignment because of low entropy (<= 16 bp)\n",
      "Reads with no insert (adaptor only)\n",
      "Reads unaligned from half aligned fragments\n",
      "Reads of good quality from unaligned fragments (missing genome, microbiome or mapping difficulty)\n",
      "% well mapped reads\n",
      "% reads mapping uniquely  to a single locus (genomic site) and maximum 1 gene\n",
      "% reads mapping uniquely to a single locus, but to 2 genes in antisense \n",
      "% reads mapping uniquely to a single locus\n",
      "% reads mapping to 2 to 9 sites\n",
      "% reads rejected because they map to 10 sites or more\n",
      "% reads rejected by alignment quality filter\n",
      "% reads rejected before alignment because of low entropy (<= 16 bp)\n",
      "% reads with no insert (adaptor only)\n",
      "% reads unaligned from half aligned fragments\n",
      "% reads of good quality from unaligned fragments (missing genome, microbiome or mapping difficulty)\n",
      "Unnamed: 86\n",
      "Reads aligned in PhiX DNA spike-in\n",
      "Reads aligned in ERCC RNA spike-in\n",
      "Reads aligned in ribosomal RNA\n",
      "Reads aligned in mitochondria\n",
      "Reads aligned in other small genes\n",
      "Reads aligned in AceView\n",
      "Reads aligned in RefSeq\n",
      "Reads aligned in Encode\n",
      "Reads aligned in Magic\n",
      "Reads aligned in genome\n",
      "Reads aligned in any previously annotated transcripts\n",
      "Reads inside introns, new exons and intergenic\n",
      "Reads aligned in microbiome\n",
      "Reads aligned in viruses\n",
      "Reads aligned on the imaginary genome (mapping specificity control)\n",
      "Reads supporting known exons-junctions\n",
      "Unnamed: 103\n",
      "% Reads aligned in PhiX DNA spike-in\n",
      "% Reads aligned in ERCC RNA spike-in\n",
      "% Reads aligned in ribosomal RNA\n",
      "% Reads aligned in mitochondria\n",
      "% Reads aligned in other small genes\n",
      "% Reads aligned in AceView\n",
      "% Reads aligned in RefSeq\n",
      "% Reads aligned in Encode\n",
      "% Reads aligned in Magic\n",
      "% Reads aligned in genome\n",
      "% Reads aligned in any previously annotated transcripts\n",
      "% Reads inside introns, new exons and intergenic\n",
      "% Reads aligned in microbiome\n",
      "% Reads aligned in viruses\n",
      "% Reads aligned on the imaginary genome (mapping specificity control)\n",
      "% Reads supporting known exons-junctions\n",
      "Unnamed: 120\n",
      "Mb aligned in PhiX DNA spike-in\n",
      "Mb aligned in ERCC RNA spike-in\n",
      "Mb aligned in ribosomal RNA\n",
      "Mb aligned in mitochondria\n",
      "Mb aligned in other small genes\n",
      "Mb aligned in AceView\n",
      "Mb aligned in RefSeq\n",
      "Mb aligned in Encode\n",
      "Mb aligned in Magic\n",
      "Mb aligned in genome\n",
      "Mb aligned in any previously annotated transcripts\n",
      "Mb inside introns, new exons and intergenic\n",
      "Mb aligned in microbiome\n",
      "Mb aligned in viruses\n",
      "Mb aligned on the imaginary genome (mapping specificity control)\n",
      "Mb supporting known exons-junctions\n",
      "Unnamed: 137\n",
      "Number of aligned_fragments\n",
      "Compatible pairs\n",
      "Non compatible pairs\n",
      "Paired end fragments with only one read aligned\n",
      "% Compatible pairs\n",
      "% Non compatible pairs\n",
      "% Paired end fragments with only one read aligned\n",
      "1% fragments in library are shorter than (nt)\n",
      "5% fragments in library are shorter than (nt)\n",
      "mode of fragment lengths\n",
      "median fragment lengths\n",
      "average of fragment lengths\n",
      "5% fragments in library are longer than (nt)\n",
      "1% fragments in library are longer than (nt)\n",
      "Unnamed: 152\n",
      "Average length after clipping adaptors and barcodes (nt)\n",
      "Average length aligned (nt)\n",
      "Average length aligned in read 1 (nt)\n",
      "Average length aligned in read 2 (paired end)\n",
      "Average length aligned on RefSeq (nt)\n",
      "Average length aligned on EBI (nt)\n",
      "Average length aligned on AceView (nt)\n",
      "Average length aligned on magic (nt)\n",
      "Average length aligned on Other (nt)\n",
      "Average length aligned on Genome (nt)\n",
      "Average length aligned on Mitochondria (nt)\n",
      "Average length aligned on rRNA (nt)\n",
      "Average length aligned on Chloroplast (nt)\n",
      "Average length aligned on small RNAs (nt)\n",
      "Average length aligned on DNA spikeIn (nt)\n",
      "Average length aligned on RNA spikeIn (nt)\n",
      "Average length aligned on Imaginary genome specificity control (nt)\n",
      "Unnamed: 170\n",
      "Average % length aligned (nt)\n",
      "Average % length aligned in read 1 (nt)\n",
      "Average % length aligned in read 2 (paired end)\n",
      "Average % length aligned on RefSeq\n",
      "Average % length aligned on EBI\n",
      "Average % length aligned on AceView\n",
      "Average % length aligned on magic\n",
      "Average % length aligned on Other\n",
      "Average % length aligned on Genome\n",
      "Average % length aligned on Mitochondria\n",
      "Average % length aligned on rRNA\n",
      "Average length aligned on Chloroplast\n",
      "Average length aligned on small RNAs\n",
      "Average length aligned on DNA spikeIn\n",
      "Average length aligned on RNA spikeIn\n",
      "Average length aligned on Imaginary genome specificity control\n",
      "Unnamed: 187\n",
      "Reads mapping on plus strand of rRNA\n",
      "Reads mapping on minus strand of rRNA\n",
      "Reads mapping on both strands of rRNA\n",
      "% reads mapping on plus strand of rRNA\n",
      "% reads mapping on minus strand of rRNA\n",
      "% reads mapping on both strand of rRNA\n",
      "Reads mapping on plus strand of RefSeq\n",
      "Reads mapping on minus strand of RefSeq\n",
      "Reads mapping on both strands of RefSeq\n",
      "% reads mapping on plus strand of RefSeq\n",
      "% reads mapping on minus strand of RefSeq\n",
      "% reads mapping on both strand of RefSeq\n",
      "Reads mapping on plus strand of EBI\n",
      "Reads mapping on minus strand of EBI\n",
      "Reads mapping on both strands of EBI\n",
      "% reads mapping on plus strand of EBI\n",
      "% reads mapping on minus strand of EBI\n",
      "% reads mapping on both strand of EBI\n",
      "Reads mapping on plus strand of av\n",
      "Reads mapping on minus strand of av\n",
      "Reads mapping on both strands of av\n",
      "% reads mapping on plus strand of av\n",
      "% reads mapping on minus strand of av\n",
      "% reads mapping on both strand of av\n",
      "Reads mapping on plus strand of genome\n",
      "Reads mapping on minus strand of genome\n",
      "Reads mapping on both strands of genome\n",
      "% reads mapping on plus strand of genome\n",
      "% reads mapping on minus strand of genome\n",
      "% reads mapping on both strand of genome\n",
      "Reads mapping on plus strand of SpikeIn\n",
      "Reads mapping on minus strand of SpikeIn\n",
      "Reads mapping on both strands of SpikeIn\n",
      "% reads mapping on plus strand of SpikeIn\n",
      "% reads mapping on minus strand of SpikeIn\n",
      "% reads mapping on both strand of SpikeIn\n",
      "Unnamed: 224\n",
      "Megabases uniquely aligned and used for mismatch counts\n",
      "Total number of mismatches\n",
      "Mismatches per kb aligned\n",
      "Transitions\n",
      "Transversions\n",
      "1, 2 or 3 bp insertions not in polymers\n",
      "1, 2 or 3 bp deletions not in polymers\n",
      "1, 2 or 3 bp insertions in polymers\n",
      "1, 2 or 3 bp deletions in polymers\n",
      "A>G\n",
      "T>C\n",
      "G>A\n",
      "C>T\n",
      "A>T\n",
      "T>A\n",
      "G>C\n",
      "C>G\n",
      "A>C\n",
      "T>G\n",
      "G>T\n",
      "C>A\n",
      "Single Insertions\n",
      "Single Deletions\n",
      "Double insertions\n",
      "Double deletions\n",
      "Triple insertions\n",
      "Triple deletions\n",
      "Transitions per kb\n",
      "Transversions per kb\n",
      "Insertions per kb\n",
      "Deletions per kb\n",
      "% transitions\n",
      "% transversions\n",
      "% insertions not in polymers\n",
      "% deletions not in polymers\n",
      "% insertions in polymers\n",
      "% deletions in polymers\n",
      "Unnamed: 262\n",
      "Accessible transcript length, estimated on transcripts longer than 8kb, usually limited by the 3' biais in poly-A selected experiments\n",
      "Number of well expressed transcripts longer than 8kb\n",
      "Average coverage cumulated over these transcripts\n",
      "Unnamed: 266\n",
      "Intergenic kb\n",
      "Intergenic density\n",
      "Unnamed: 269\n",
      "Zero index\n",
      "Low index\n",
      "Cross over index\n",
      "Genes touched\n",
      "Genes with reliable index\n",
      "Genes with index >= 10\n",
      "Genes with index >= 12\n",
      "Genes with index >= 15\n",
      "Genes with index >= 18\n",
      "Genes expressed in at least one run\n",
      "RefSeq genes significantly expressed in at least one run\n",
      "RefSeq genes significantly expressed in majority of runs\n",
      "ENCODE genes significantly expressed in at least one run\n",
      "ENCODE genes significantly expressed in majority of runs\n",
      "AceView genes significantly expressed in at least one run\n",
      "AceView genes significantly expressed in majority of runs\n",
      "AceView transcripts significantly expressed in  at least one run\n",
      " Mb_aligned\n",
      " Mb_in_genes\n",
      "Mb_in_genes_with_GeneId_minus_high_genes\n",
      "Mb_in_high_genes\n",
      "Intergenic\n"
     ]
    }
   ],
   "source": [
    "for c in dfj.columns:\n",
    "    print(c)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Compare Z vs J"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "zList = dfZ.BioSample.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n",
      "<!DOCTYPE eInfoResult PUBLIC \"-//NLM//DTD einfo 20130322//EN\" \"http://eutils.ncbi.nlm.nih.gov/eutils/dtd/20130322/einfo.dtd\">\n",
      "<eInfoResult>\n",
      "<DbList>\n",
      "\n",
      "\t<DbName>pubmed</DbName>\n",
      "\t<DbName>protein</DbName>\n",
      "\t<DbName>nuccore</DbName>\n",
      "\t<DbName>nucleotide</DbName>\n",
      "\t<DbName>nucgss</DbName>\n",
      "\t<DbName>nucest</DbName>\n",
      "\t<DbName>structure</DbName>\n",
      "\t<DbName>genome</DbName>\n",
      "\t<DbName>annotinfo</DbName>\n",
      "\t<DbName>assembly</DbName>\n",
      "\t<DbName>bioproject</DbName>\n",
      "\t<DbName>biosample</DbName>\n",
      "\t<DbName>blastdbinfo</DbName>\n",
      "\t<DbName>books</DbName>\n",
      "\t<DbName>cdd</DbName>\n",
      "\t<DbName>clinvar</DbName>\n",
      "\t<DbName>clone</DbName>\n",
      "\t<DbName>gap</DbName>\n",
      "\t<DbName>gapplus</DbName>\n",
      "\t<DbName>grasp</DbName>\n",
      "\t<DbName>dbvar</DbName>\n",
      "\t<DbName>gene</DbName>\n",
      "\t<DbName>gds</DbName>\n",
      "\t<DbName>geoprofiles</DbName>\n",
      "\t<DbName>homologene</DbName>\n",
      "\t<DbName>medgen</DbName>\n",
      "\t<DbName>mesh</DbName>\n",
      "\t<DbName>ncbisearch</DbName>\n",
      "\t<DbName>nlmcatalog</DbName>\n",
      "\t<DbName>omim</DbName>\n",
      "\t<DbName>orgtrack</DbName>\n",
      "\t<DbName>pmc</DbName>\n",
      "\t<DbName>popset</DbName>\n",
      "\t<DbName>probe</DbName>\n",
      "\t<DbName>proteinclusters</DbName>\n",
      "\t<DbName>pcassay</DbName>\n",
      "\t<DbName>biosystems</DbName>\n",
      "\t<DbName>pccompound</DbName>\n",
      "\t<DbName>pcsubstance</DbName>\n",
      "\t<DbName>pubmedhealth</DbName>\n",
      "\t<DbName>seqannot</DbName>\n",
      "\t<DbName>snp</DbName>\n",
      "\t<DbName>sra</DbName>\n",
      "\t<DbName>taxonomy</DbName>\n",
      "\t<DbName>unigene</DbName>\n",
      "\t<DbName>gencoll</DbName>\n",
      "\t<DbName>gtr</DbName>\n",
      "</DbList>\n",
      "\n",
      "</eInfoResult>\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(Entrez.einfo().read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "r = Entrez.efetch(db='biosample', id=zList[:10].tolist(), retmode='xml')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "res = Entrez.parse(r, validate=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "The XML file does not represent a list. Please use Entrez.read instead of Entrez.parse",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-56-6a2407def4e3>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0;32mfor\u001b[0m \u001b[0mb\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mres\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m     \u001b[0;32mpass\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/fearjm/opt/miniconda3/lib/python3.5/site-packages/Bio/Entrez/Parser.py\u001b[0m in \u001b[0;36mparse\u001b[0;34m(self, handle)\u001b[0m\n\u001b[1;32m    294\u001b[0m             \u001b[0mrecords\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstack\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    295\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrecords\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 296\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"The XML file does not represent a list. Please use Entrez.read instead of Entrez.parse\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    297\u001b[0m             \u001b[0;32mwhile\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrecords\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m  \u001b[0;31m# Then the top record is finished\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    298\u001b[0m                 \u001b[0mrecord\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrecords\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpop\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: The XML file does not represent a list. Please use Entrez.read instead of Entrez.parse"
     ]
    }
   ],
   "source": [
    "for b in res:\n",
    "    pass"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
