{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There is some differences between Jean's and Zhenxia's alignments. I want to examine the differences closely. I am going to focus on:\n",
    "\n",
    "* DRR014573\n",
    "* SRR3086588\n",
    "* SRR942869"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "pd.options.display.max_columns=999"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 1</th>\n",
       "      <th>Total reads</th>\n",
       "      <th>Z Total reads</th>\n",
       "      <th>% Total reads</th>\n",
       "      <th>% Z Total reads</th>\n",
       "      <th>% difference</th>\n",
       "      <th>Mapped</th>\n",
       "      <th>Z Mapped</th>\n",
       "      <th>% Mapped</th>\n",
       "      <th>% Z Mapped</th>\n",
       "      <th>% difference.1</th>\n",
       "      <th>Mapped uniquely</th>\n",
       "      <th>Z Mapped uniquely</th>\n",
       "      <th>% Mapped uniquely</th>\n",
       "      <th>% Z Mapped uniquely</th>\n",
       "      <th>% difference.2</th>\n",
       "      <th>Mapped non uniquely</th>\n",
       "      <th>Z Mapped non_unique</th>\n",
       "      <th>% Mapped non uniquely</th>\n",
       "      <th>% Z non_uniquely</th>\n",
       "      <th>% difference.3</th>\n",
       "      <th>Unmapped</th>\n",
       "      <th>Z unmapped_reads</th>\n",
       "      <th>% Unmapped</th>\n",
       "      <th>% Z unmapped_reads</th>\n",
       "      <th>% difference.4</th>\n",
       "      <th>Intron support: a read supporting 2 introns counts 2</th>\n",
       "      <th>Z spliced read</th>\n",
       "      <th>% Intron supports: a read supporting 2 intron counts 2</th>\n",
       "      <th>% Z spliced read</th>\n",
       "      <th>% difference.5</th>\n",
       "      <th>Reads mapped in transcripts</th>\n",
       "      <th>Z Read fragments mapped in genes, a read touching 2 exons counts 2</th>\n",
       "      <th>% Reads mapped in transcripts</th>\n",
       "      <th>% Z Read fragments mapped in genes, a read touching 2 exons counts 2</th>\n",
       "      <th>% difference.6</th>\n",
       "      <th>Stranding</th>\n",
       "      <th>Z Stranding</th>\n",
       "      <th>% difference.7</th>\n",
       "      <th>Significantly expressed genes</th>\n",
       "      <th>Touched Genes</th>\n",
       "      <th>Z expressed genes : is it touched or significant ?</th>\n",
       "      <th>Difference</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Z sex</th>\n",
       "      <th>Stage</th>\n",
       "      <th>Z dev_stage</th>\n",
       "      <th>Tissue-System</th>\n",
       "      <th>Z tissue</th>\n",
       "      <th>Z cell_type</th>\n",
       "      <th>Z sample_type</th>\n",
       "      <th>Stage summary</th>\n",
       "      <th>Z genotype</th>\n",
       "      <th>Magic sample title</th>\n",
       "      <th>Sequencing length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>### Run</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DRR014573</th>\n",
       "      <td>NaN</td>\n",
       "      <td>91947.0</td>\n",
       "      <td>91947.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>91069.0</td>\n",
       "      <td>296.0</td>\n",
       "      <td>99.05</td>\n",
       "      <td>0.32</td>\n",
       "      <td>-98.72</td>\n",
       "      <td>89415.0</td>\n",
       "      <td>242.0</td>\n",
       "      <td>97.25</td>\n",
       "      <td>0.26</td>\n",
       "      <td>-96.98</td>\n",
       "      <td>1654.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>1.8</td>\n",
       "      <td>0.06</td>\n",
       "      <td>-1.74</td>\n",
       "      <td>878.0</td>\n",
       "      <td>91651.0</td>\n",
       "      <td>0.95</td>\n",
       "      <td>99.68</td>\n",
       "      <td>98.72</td>\n",
       "      <td>0.0</td>\n",
       "      <td>102.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.11</td>\n",
       "      <td>65371.0</td>\n",
       "      <td>219.0</td>\n",
       "      <td>71.10</td>\n",
       "      <td>0.24</td>\n",
       "      <td>-70.86</td>\n",
       "      <td>52.16</td>\n",
       "      <td>50.45</td>\n",
       "      <td>1.72</td>\n",
       "      <td>412</td>\n",
       "      <td>2201</td>\n",
       "      <td>110</td>\n",
       "      <td>2091</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unknown</td>\n",
       "      <td>L3</td>\n",
       "      <td>larva</td>\n",
       "      <td>Body_part; Other; salivary gland and fat body ...</td>\n",
       "      <td>whole_organism</td>\n",
       "      <td>mix</td>\n",
       "      <td>whole_organism</td>\n",
       "      <td>salivary gland and fat body dissected from 20 ...</td>\n",
       "      <td>Twenty 3rd instar larvae of D. melanogaster (C...</td>\n",
       "      <td>Penicillium-fungus uninfected 3rd instar larva...</td>\n",
       "      <td>294.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR3086588</th>\n",
       "      <td>NaN</td>\n",
       "      <td>29560480.0</td>\n",
       "      <td>29560480.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>28723030.0</td>\n",
       "      <td>3385.0</td>\n",
       "      <td>97.17</td>\n",
       "      <td>0.01</td>\n",
       "      <td>-97.16</td>\n",
       "      <td>28722320.0</td>\n",
       "      <td>426.0</td>\n",
       "      <td>97.16</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-97.16</td>\n",
       "      <td>710.0</td>\n",
       "      <td>2959.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>837450.0</td>\n",
       "      <td>29557090.0</td>\n",
       "      <td>2.83</td>\n",
       "      <td>99.99</td>\n",
       "      <td>97.16</td>\n",
       "      <td>117.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-0.00</td>\n",
       "      <td>28720688.0</td>\n",
       "      <td>289.0</td>\n",
       "      <td>97.16</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-97.16</td>\n",
       "      <td>86.28</td>\n",
       "      <td>77.97</td>\n",
       "      <td>8.31</td>\n",
       "      <td>1</td>\n",
       "      <td>22</td>\n",
       "      <td>153</td>\n",
       "      <td>-131</td>\n",
       "      <td>Male</td>\n",
       "      <td>male</td>\n",
       "      <td>Adult</td>\n",
       "      <td>adult</td>\n",
       "      <td>Germline_and_development; testis</td>\n",
       "      <td>testis</td>\n",
       "      <td>mix</td>\n",
       "      <td>whole_tissue</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HMS00184</td>\n",
       "      <td>Model organism or animal sample from, tudormt ...</td>\n",
       "      <td>50.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR942869</th>\n",
       "      <td>NaN</td>\n",
       "      <td>144718704.0</td>\n",
       "      <td>144718704.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>129969800.0</td>\n",
       "      <td>12657088.0</td>\n",
       "      <td>89.81</td>\n",
       "      <td>8.75</td>\n",
       "      <td>-81.06</td>\n",
       "      <td>128809800.0</td>\n",
       "      <td>524808.0</td>\n",
       "      <td>89.01</td>\n",
       "      <td>0.36</td>\n",
       "      <td>-88.64</td>\n",
       "      <td>1160000.0</td>\n",
       "      <td>12132280.0</td>\n",
       "      <td>0.8</td>\n",
       "      <td>8.38</td>\n",
       "      <td>7.58</td>\n",
       "      <td>14748904.0</td>\n",
       "      <td>132061600.0</td>\n",
       "      <td>10.19</td>\n",
       "      <td>91.25</td>\n",
       "      <td>81.06</td>\n",
       "      <td>708082.0</td>\n",
       "      <td>327149.0</td>\n",
       "      <td>0.49</td>\n",
       "      <td>0.23</td>\n",
       "      <td>-0.26</td>\n",
       "      <td>128550736.0</td>\n",
       "      <td>190612.0</td>\n",
       "      <td>88.83</td>\n",
       "      <td>0.13</td>\n",
       "      <td>-88.70</td>\n",
       "      <td>99.88</td>\n",
       "      <td>47.34</td>\n",
       "      <td>52.54</td>\n",
       "      <td>2379</td>\n",
       "      <td>5520</td>\n",
       "      <td>6266</td>\n",
       "      <td>-746</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Embryo</td>\n",
       "      <td>embryo</td>\n",
       "      <td>NaN</td>\n",
       "      <td>whole_organism</td>\n",
       "      <td>mix</td>\n",
       "      <td>whole_organism</td>\n",
       "      <td>NaN</td>\n",
       "      <td>yw</td>\n",
       "      <td>footprint 0-2 hour embryo replicate A, whole 0...</td>\n",
       "      <td>57.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Unnamed: 1  Total reads  Z Total reads   % Total reads  \\\n",
       "### Run                                                              \n",
       "DRR014573          NaN      91947.0        91947.0           100.0   \n",
       "SRR3086588         NaN   29560480.0     29560480.0           100.0   \n",
       "SRR942869          NaN  144718704.0    144718704.0           100.0   \n",
       "\n",
       "             % Z Total reads  % difference       Mapped    Z Mapped  % Mapped  \\\n",
       "### Run                                                                         \n",
       "DRR014573              100.0           0.0      91069.0       296.0     99.05   \n",
       "SRR3086588             100.0           0.0   28723030.0      3385.0     97.17   \n",
       "SRR942869              100.0           0.0  129969800.0  12657088.0     89.81   \n",
       "\n",
       "            % Z Mapped  % difference.1  Mapped uniquely  Z Mapped uniquely  \\\n",
       "### Run                                                                      \n",
       "DRR014573         0.32          -98.72          89415.0              242.0   \n",
       "SRR3086588        0.01          -97.16       28722320.0              426.0   \n",
       "SRR942869         8.75          -81.06      128809800.0           524808.0   \n",
       "\n",
       "            % Mapped uniquely  % Z Mapped uniquely  % difference.2  \\\n",
       "### Run                                                              \n",
       "DRR014573               97.25                 0.26          -96.98   \n",
       "SRR3086588              97.16                 0.00          -97.16   \n",
       "SRR942869               89.01                 0.36          -88.64   \n",
       "\n",
       "            Mapped non uniquely  Z Mapped non_unique  % Mapped non uniquely  \\\n",
       "### Run                                                                       \n",
       "DRR014573                1654.0                 54.0                    1.8   \n",
       "SRR3086588                710.0               2959.0                    0.0   \n",
       "SRR942869             1160000.0           12132280.0                    0.8   \n",
       "\n",
       "            % Z non_uniquely  % difference.3    Unmapped  Z unmapped_reads  \\\n",
       "### Run                                                                      \n",
       "DRR014573               0.06           -1.74       878.0           91651.0   \n",
       "SRR3086588              0.01            0.01    837450.0        29557090.0   \n",
       "SRR942869               8.38            7.58  14748904.0       132061600.0   \n",
       "\n",
       "            % Unmapped  % Z unmapped_reads  % difference.4  \\\n",
       "### Run                                                      \n",
       "DRR014573         0.95               99.68           98.72   \n",
       "SRR3086588        2.83               99.99           97.16   \n",
       "SRR942869        10.19               91.25           81.06   \n",
       "\n",
       "            Intron support: a read supporting 2 introns counts 2  \\\n",
       "### Run                                                            \n",
       "DRR014573                                                 0.0      \n",
       "SRR3086588                                              117.0      \n",
       "SRR942869                                            708082.0      \n",
       "\n",
       "            Z spliced read  \\\n",
       "### Run                      \n",
       "DRR014573            102.0   \n",
       "SRR3086588             9.0   \n",
       "SRR942869         327149.0   \n",
       "\n",
       "            % Intron supports: a read supporting 2 intron counts 2  \\\n",
       "### Run                                                              \n",
       "DRR014573                                                0.00        \n",
       "SRR3086588                                               0.00        \n",
       "SRR942869                                                0.49        \n",
       "\n",
       "            % Z spliced read  % difference.5  Reads mapped in transcripts  \\\n",
       "### Run                                                                     \n",
       "DRR014573               0.11            0.11                      65371.0   \n",
       "SRR3086588              0.00           -0.00                   28720688.0   \n",
       "SRR942869               0.23           -0.26                  128550736.0   \n",
       "\n",
       "            Z Read fragments mapped in genes, a read touching 2 exons counts 2  \\\n",
       "### Run                                                                          \n",
       "DRR014573                                               219.0                    \n",
       "SRR3086588                                              289.0                    \n",
       "SRR942869                                            190612.0                    \n",
       "\n",
       "            % Reads mapped in transcripts  \\\n",
       "### Run                                     \n",
       "DRR014573                           71.10   \n",
       "SRR3086588                          97.16   \n",
       "SRR942869                           88.83   \n",
       "\n",
       "            % Z Read fragments mapped in genes, a read touching 2 exons counts 2  \\\n",
       "### Run                                                                            \n",
       "DRR014573                                                0.24                      \n",
       "SRR3086588                                               0.00                      \n",
       "SRR942869                                                0.13                      \n",
       "\n",
       "            % difference.6  Stranding  Z Stranding  % difference.7  \\\n",
       "### Run                                                              \n",
       "DRR014573           -70.86      52.16        50.45            1.72   \n",
       "SRR3086588          -97.16      86.28        77.97            8.31   \n",
       "SRR942869           -88.70      99.88        47.34           52.54   \n",
       "\n",
       "            Significantly expressed genes  Touched Genes  \\\n",
       "### Run                                                    \n",
       "DRR014573                             412           2201   \n",
       "SRR3086588                              1             22   \n",
       "SRR942869                            2379           5520   \n",
       "\n",
       "            Z expressed genes : is it touched or significant ?  Difference  \\\n",
       "### Run                                                                      \n",
       "DRR014573                                                 110         2091   \n",
       "SRR3086588                                                153         -131   \n",
       "SRR942869                                                6266         -746   \n",
       "\n",
       "             Sex    Z sex   Stage Z dev_stage  \\\n",
       "### Run                                         \n",
       "DRR014573    NaN  unknown      L3       larva   \n",
       "SRR3086588  Male     male   Adult       adult   \n",
       "SRR942869    NaN  unknown  Embryo      embryo   \n",
       "\n",
       "                                                Tissue-System        Z tissue  \\\n",
       "### Run                                                                         \n",
       "DRR014573   Body_part; Other; salivary gland and fat body ...  whole_organism   \n",
       "SRR3086588                   Germline_and_development; testis          testis   \n",
       "SRR942869                                                 NaN  whole_organism   \n",
       "\n",
       "           Z cell_type   Z sample_type  \\\n",
       "### Run                                  \n",
       "DRR014573          mix  whole_organism   \n",
       "SRR3086588         mix    whole_tissue   \n",
       "SRR942869          mix  whole_organism   \n",
       "\n",
       "                                                Stage summary  \\\n",
       "### Run                                                         \n",
       "DRR014573   salivary gland and fat body dissected from 20 ...   \n",
       "SRR3086588                                                NaN   \n",
       "SRR942869                                                 NaN   \n",
       "\n",
       "                                                   Z genotype  \\\n",
       "### Run                                                         \n",
       "DRR014573   Twenty 3rd instar larvae of D. melanogaster (C...   \n",
       "SRR3086588                                           HMS00184   \n",
       "SRR942869                                                  yw   \n",
       "\n",
       "                                           Magic sample title  \\\n",
       "### Run                                                         \n",
       "DRR014573   Penicillium-fungus uninfected 3rd instar larva...   \n",
       "SRR3086588  Model organism or animal sample from, tudormt ...   \n",
       "SRR942869   footprint 0-2 hour embryo replicate A, whole 0...   \n",
       "\n",
       "            Sequencing length  \n",
       "### Run                        \n",
       "DRR014573               294.0  \n",
       "SRR3086588               50.0  \n",
       "SRR942869                57.0  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('../../data/jean/Mapping_QC/ZhenXia.Data_Summary.ZhenXia.txt', sep='\\t', skiprows=2, index_col=0)\n",
    "df.loc[['DRR014573', 'SRR3086588', 'SRR942869']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can see from this table that while the number of total reads is the same between Jean and Zhenxia, the number of mapped reads is completely different, with Zhenxia failing to map a large portion of reads."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "# Prep Temp Folder and References"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {
    "collapsed": false,
    "hidden": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/fearjm/opt/miniconda3/bin/hisat2-align-s version 2.0.4\n",
      "cutadapt 1.9.1\n",
      "samtools 1.3.1\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "hisat2 --version |head -n1 || conda install hisat2 -y\n",
    "(printf \"cutadapt \" && cutadapt --version) || conda install cutadapt -y\n",
    "samtools --version | head -n1 || conda install samtools -y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {
    "collapsed": false,
    "hidden": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "receiving file list ... done\n",
      "\n",
      "sent 20 bytes  received 302 bytes  49.54 bytes/sec\n",
      "total size is 457850364  speedup is 1421895.54\n"
     ]
    }
   ],
   "source": [
    "%%bash --err error\n",
    "REF=~/tmp/ncbi_remap/references\n",
    "if [ ! -e $REF ]; then\n",
    "    mkdir -p $REF\n",
    "fi\n",
    "\n",
    "rsync -av bio:/data/Oliverlab/data/FlyBase/FB2015_04/dmel_ERCC* $REF/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run Hisat2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First I want to look at using Hisat2 to align one of the above samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "91947 reads; of these:\n",
      "  91947 (100.00%) were unpaired; of these:\n",
      "    61151 (66.51%) aligned 0 times\n",
      "    16819 (18.29%) aligned exactly 1 time\n",
      "    13977 (15.20%) aligned >1 times\n",
      "33.49% overall alignment rate\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd ~/tmp/ncbi_remap\n",
    "\n",
    "hisat2 -p 4 \\\n",
    "    -x references/dmel_ERCC \\\n",
    "    --dta --sra-acc DRR014573 \\\n",
    "    --known-splicesite-infile references/dmel_ERCC.splicesites.txt \\\n",
    "    -S DRR014573.hisat.sra.sam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "discarding /Users/fearjm/opt/miniconda3/bin from PATH\n",
      "prepending /Users/fearjm/opt/miniconda3/envs/hisat2/bin to PATH\n",
      "Load SAM file ...  Done\n",
      "\n",
      "#==================================================\n",
      "#All numbers are READ count\n",
      "#==================================================\n",
      "\n",
      "Total records:                          97105\n",
      "\n",
      "QC failed:                              0\n",
      "Optical/PCR duplicate:                  0\n",
      "Non primary hits                        5158\n",
      "Unmapped reads:                         61151\n",
      "mapq < mapq_cut (non-unique):           3659\n",
      "\n",
      "mapq >= mapq_cut (unique):              27137\n",
      "Read-1:                                 0\n",
      "Read-2:                                 0\n",
      "Reads map to '+':                       12766\n",
      "Reads map to '-':                       14371\n",
      "Non-splice reads:                       17758\n",
      "Splice reads:                           9379\n",
      "Reads mapped in proper pairs:           0\n",
      "Proper-paired reads map to different chrom:0\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "source activate hisat2\n",
    "cd ~/tmp/ncbi_remap\n",
    "\n",
    "bam_stat.py -i DRR014573.hisat.sra.sam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As you can see, bam_stat reports strange counts. But, either way my alignment percentage is low (~18%). So next I am going to try to dump the SRA to a FASTQ and look at the FASTQ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Rejected 91947 READS because READLEN < 35\n",
      "Read 91947 spots for DRR014573\n",
      "Written 91947 spots for DRR014573\n"
     ]
    }
   ],
   "source": [
    "%%bash \n",
    "cd ~/tmp/ncbi_remap\n",
    "\n",
    "fastq-dump --split-spot -M 35 DRR014573"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "91947 reads; of these:\n",
      "  91947 (100.00%) were unpaired; of these:\n",
      "    61150 (66.51%) aligned 0 times\n",
      "    16813 (18.29%) aligned exactly 1 time\n",
      "    13984 (15.21%) aligned >1 times\n",
      "33.49% overall alignment rate\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd ~/tmp/ncbi_remap\n",
    "\n",
    "hisat2 -p 4 \\\n",
    "    -x references/dmel_ERCC \\\n",
    "    --dta -U DRR014573.fastq \\\n",
    "    --known-splicesite-infile references/dmel_ERCC.splicesites.txt \\\n",
    "    -S DRR014573.hisat.fq.sam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "91947 reads; of these:\n",
      "  91947 (100.00%) were unpaired; of these:\n",
      "    35778 (38.91%) aligned 0 times\n",
      "    29482 (32.06%) aligned exactly 1 time\n",
      "    26687 (29.02%) aligned >1 times\n",
      "61.09% overall alignment rate\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd ~/tmp/ncbi_remap\n",
    "\n",
    "bowtie2 -p 4 -x ~/references/genomes/Dmelanogaster/dm6/bowtie2/dm6-noextra \\\n",
    "    -U DRR014573.fastq > DRR014573.bt2.sam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "91947 reads; of these:\n",
      "  91947 (100.00%) were unpaired; of these:\n",
      "    825 (0.90%) aligned 0 times\n",
      "    36591 (39.80%) aligned exactly 1 time\n",
      "    54531 (59.31%) aligned >1 times\n",
      "99.10% overall alignment rate\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd ~/tmp/ncbi_remap\n",
    "\n",
    "bowtie2 -p 4 -x ~/references/genomes/Dmelanogaster/dm6/bowtie2/dm6-noextra \\\n",
    "    --local \\\n",
    "    -U DRR014573.fastq > DRR014573.bt2.sam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Just dumping the file using fastq-dump and aligning with bowtie2 we get better alignments, but still not close to Jean's alignments. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is cutadapt 1.9.1 with Python 3.5.1\n",
      "Command line parameters: -a AGATCGGAAGAGC -a A{50} -a T{50} --trim-n -o DRR014573_2.trim.fastq DRR014573_2.fastq\n",
      "Trimming 3 adapters with at most 10.0% errors in single-end mode ...\n",
      "Finished in 4.66 s (51 us/read; 1.18 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:                  91,947\n",
      "Reads with adapters:                        81 (0.1%)\n",
      "Reads written (passing filters):        91,947 (100.0%)\n",
      "\n",
      "Total basepairs processed:    26,703,586 bp\n",
      "Total written (filtered):     26,417,431 bp (98.9%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: AGATCGGAAGAGC; Type: regular 3'; Length: 13; Trimmed: 37 times.\n",
      "\n",
      "No. of allowed errors:\n",
      "0-9 bp: 0; 10-13 bp: 1\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 10.8%\n",
      "  C: 45.9%\n",
      "  G: 24.3%\n",
      "  T: 16.2%\n",
      "  none/other: 2.7%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "3\t12\t1436.7\t0\t12\n",
      "5\t1\t89.8\t0\t1\n",
      "6\t6\t22.4\t0\t6\n",
      "9\t1\t0.4\t0\t0 1\n",
      "47\t1\t0.0\t1\t0 1\n",
      "74\t1\t0.0\t1\t0 1\n",
      "84\t1\t0.0\t1\t0 1\n",
      "104\t1\t0.0\t1\t0 1\n",
      "147\t1\t0.0\t1\t0 1\n",
      "161\t1\t0.0\t1\t0 1\n",
      "163\t1\t0.0\t1\t0 1\n",
      "167\t1\t0.0\t1\t0 1\n",
      "199\t1\t0.0\t1\t0 1\n",
      "232\t1\t0.0\t1\t0 1\n",
      "241\t1\t0.0\t1\t0 1\n",
      "245\t1\t0.0\t1\t0 1\n",
      "358\t1\t0.0\t1\t0 1\n",
      "380\t1\t0.0\t1\t0 1\n",
      "383\t1\t0.0\t1\t0 1\n",
      "460\t1\t0.0\t1\t0 1\n",
      "473\t1\t0.0\t1\t0 1\n",
      "\n",
      "=== Adapter 2 ===\n",
      "\n",
      "Sequence: AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA; Type: regular 3'; Length: 50; Trimmed: 36 times.\n",
      "\n",
      "No. of allowed errors:\n",
      "0-9 bp: 0; 10-19 bp: 1; 20-29 bp: 2; 30-39 bp: 3; 40-49 bp: 4; 50 bp: 5\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 0.0%\n",
      "  C: 27.8%\n",
      "  G: 8.3%\n",
      "  T: 61.1%\n",
      "  none/other: 2.8%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "3\t20\t1436.7\t0\t20\n",
      "4\t12\t359.2\t0\t12\n",
      "5\t1\t89.8\t0\t1\n",
      "7\t1\t5.6\t0\t1\n",
      "22\t1\t0.0\t2\t0 0 1\n",
      "249\t1\t0.0\t5\t0 0 0 0 1\n",
      "\n",
      "=== Adapter 3 ===\n",
      "\n",
      "Sequence: TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT; Type: regular 3'; Length: 50; Trimmed: 8 times.\n",
      "\n",
      "No. of allowed errors:\n",
      "0-9 bp: 0; 10-19 bp: 1; 20-29 bp: 2; 30-39 bp: 3; 40-49 bp: 4; 50 bp: 5\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 50.0%\n",
      "  C: 25.0%\n",
      "  G: 0.0%\n",
      "  T: 0.0%\n",
      "  none/other: 25.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "5\t1\t89.8\t0\t1\n",
      "118\t1\t0.0\t5\t0 0 0 0 1\n",
      "205\t1\t0.0\t5\t0 0 0 0 0 1\n",
      "249\t1\t0.0\t5\t0 0 0 0 0 1\n",
      "261\t1\t0.0\t5\t0 0 0 0 0 1\n",
      "294\t1\t0.0\t5\t0 0 0 0 0 1\n",
      "380\t1\t0.0\t5\t0 0 1\n",
      "425\t1\t0.0\t5\t0 0 0 0 0 1\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd ~/tmp/ncbi_remap\n",
    "\n",
    "cutadapt -a AGATCGGAAGAGC -a \"A{50}\" -a \"T{50}\" --trim-n -o DRR014573_2.trim.fastq DRR014573_2.fastq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
