{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sample Metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I have done my first submission of RNA-Seq data to GEO, but they really want sample biological metadata to add for indexing purposes. This is not an easy problem solved for everything, but if I can get most things sorted out then that will be good enough. I want to focus on the categroies:\n",
    "\n",
    "* Sex\n",
    "* Developmental stage\n",
    "* strain\n",
    "* tissue\n",
    "* cell type\n",
    "\n",
    "I am not sure if I will be able to get a reasonable parsing of strain, but at least I can try the others."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "from pathlib import Path\n",
    "import re\n",
    "from collections import defaultdict\n",
    "from yaml import load\n",
    "\n",
    "from IPython.display import display, HTML, Markdown\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "# Project level imports\n",
    "sys.path.insert(0, '../lib')\n",
    "from ncbi_remap.notebook import Nb\n",
    "from ncbi_remap.plotting import make_figs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please check output/fbgn2chrom.tsv. If it does not exist, run bin/fbgn2chrom.py\n",
      "last updated: 2018-07-31 \n",
      "Git hash: 8506138704bf2e17010368949820110db52a4599\n"
     ]
    }
   ],
   "source": [
    "# Setup notebook\n",
    "nbconfig = Nb.setup_notebook()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pymongo import MongoClient\n",
    "try:\n",
    "    with open('../output/.mongodb_host', 'r') as fh:\n",
    "        host = fh.read().strip()\n",
    "except FileNotFoundError:\n",
    "    host = 'localhost'\n",
    "\n",
    "mongoClient = MongoClient(host=host, port=27017)\n",
    "db = mongoClient['sra']\n",
    "ncbi = db['ncbi']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get list of uploaded samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_list = pd.read_csv('/media/psf/Promise_Pegasus/fearjm/ncbi_remap/geo-wf/justin.fear@nih.gov/sample_section.tsv', sep='\\t', usecols=[0]).sample_name.unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sex</th>\n",
       "      <th>developmental stage</th>\n",
       "      <th>tissue</th>\n",
       "      <th>strain</th>\n",
       "      <th>cell type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DRX013093</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRX013094</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRX014765</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRX014766</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRX014767</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           sex developmental stage tissue strain cell type\n",
       "DRX013093  NaN                 NaN    NaN    NaN       NaN\n",
       "DRX013094  NaN                 NaN    NaN    NaN       NaN\n",
       "DRX014765  NaN                 NaN    NaN    NaN       NaN\n",
       "DRX014766  NaN                 NaN    NaN    NaN       NaN\n",
       "DRX014767  NaN                 NaN    NaN    NaN       NaN"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata = pd.DataFrame([], index=sample_list, columns=['sex', 'developmental stage', 'tissue', 'strain', 'cell type'])\n",
    "metadata.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Attribute dump"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "attrs = pd.DataFrame(list(ncbi.aggregate([\n",
    "    {\n",
    "        '$match': {\n",
    "            '_id': {'$in': sample_list}\n",
    "        }\n",
    "    },\n",
    "    {\n",
    "        '$unwind': {\n",
    "            'path': '$sra.sample.attributes'\n",
    "        }\n",
    "    },\n",
    "    {\n",
    "        '$project': {\n",
    "            '_id': 0,\n",
    "            'srx': '$_id',\n",
    "            'name': '$sra.sample.attributes.name',\n",
    "            'value': '$sra.sample.attributes.value'\n",
    "        }\n",
    "    },\n",
    "])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(70753, 3)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attrs.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initial normalization\n",
    "replace_dict = {\n",
    "    '_': ' ', \n",
    "    '; ': ';', \n",
    "    ' / ': '/',\n",
    "    'dev stage': 'developmental stage', \n",
    "    'dev-stage': 'developmental stage',\n",
    "    'developemntal stage': 'developmental stage', \n",
    "    'development stage': 'developmental stage', \n",
    "    'developmental stage': 'developmental stage', \n",
    "    'develpomental stage': 'developmental stage', \n",
    "    'cell/tissue type': 'tissue',\n",
    "    'tissue lib': 'tissue', \n",
    "    'tissue source': 'tissue',\n",
    "    'tissue type': 'tissue', \n",
    "    'tissue/cell type': 'tissue',\n",
    "}\n",
    "regex = re.compile('|'.join(replace_dict.keys()))\n",
    "\n",
    "attrs['name'] = attrs['name'].str.lower().apply(lambda s: regex.sub(lambda x: replace_dict[x.group()], s))\n",
    "attrs['value'] = attrs['value'].str.lower().apply(lambda s: regex.sub(lambda x: replace_dict[x.group()], s))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "attrsp = attrs.pivot_table(values='value', index='srx', columns='name', aggfunc='first').copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "bow = {}\n",
    "for srx, row in attrsp.iterrows():\n",
    "    bow[srx] = ' '.join(row.dropna().values.tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['adapter', 'adapter barcode', 'affinity purification', 'age', 'age of flies', 'age of fly in days post eclosion', 'agent', 'ages', 'alternate taxon id 1', 'alternate taxon id 2', 'amplification', 'antibody', 'assay', 'background strain', 'barcode', 'barcode-kit', 'batch', 'biological replicate', 'biological replicate number', 'biological replicates', 'biomarker', 'biomaterial provider', 'bioproject id', 'bioprojectid', 'biorep', 'biosamplemodel', 'birth date', 'birth location', 'bloomington stock center id', 'body site', 'breed', 'breeding history', 'breeding method', 'cage', 'cell class', 'cell line', 'cell line background', 'cell passages', 'cell subtype', 'cell type', 'cell type marker', 'cells', 'cells derived from', 'checksum', 'chip antibody', 'chip or ip antibody', 'clip antibody', 'colection date', 'collected by', 'collection date', 'compound', 'condition', 'crosses', 'cultivar', 'culture collection', 'custom name', 'days after eclosion', 'days after treatment', 'days at 29 c', 'days post infection', 'death date', 'dev stage', 'developmental stage', 'developmental temperature', 'developmental time', 'developmental time point', 'dgrp line', 'diet', 'directional', 'disease', 'disease stage', 'dissection', 'dmso', 'dose', 'drosdel deficiency', 'drosdel id', 'drug for synchronization', 'dstim knockdown status', 'ecotype', 'embryonic stage', 'embryonic temperature', 'ena-first-public', 'ena-last-update', 'environment (biome)', 'environment (feature)', 'environment (material)', 'ercc infomation', 'ercc pool', 'evolutionary regime', 'experiment', 'experimental condition', 'experimental population', 'experimental procedure', 'expression', 'expression vector', 'exptgroup', 'external rna control', 'extraction method', 'extraction protocol', 'facs-sorted cells', 'fastqs from', 'fed on', 'fed with', 'feeding type', 'female genotype', 'file no.', 'fixation', 'flag', 'flow cell', 'fluorescent cell marker', 'fly knockdown gene', 'fly knockdown vdrc line', 'fly line', 'fly/stock id', 'flybase species id', 'food', 'fraction', 'full genotype', 'gal4 line', 'gender', 'gender type', 'gender/age', 'generation', 'genetic background', 'genetic cross', 'genetic modification', 'genotype', 'genotype of male the female mated with', 'genotype/variaion', 'genotype/variation', 'geo loc name', 'geographic location', 'geographic location (country and/or sea)', 'germline knock-down', 'germline knockdown and other transgenes', 'grandparental obesity', 'group', 'growth condition', 'growth conditions', 'growth protocol', 'guide rna', 'health state', 'heat hours', 'hormone for culture', 'host', 'host-associated environmental package', 'hours at restrictive temperature', 'human knockdown gene', 'iclip barcode', 'illumina barcode', 'incubation time', 'index', 'infection', 'infection status', 'initial time point', 'injected with', 'injury', 'input rna volume', 'investigation type', 'ip antibody', 'isolate', 'isolation source', 'knockdown', 'label', 'lane', 'lat lon', 'latitude', 'length of culture', 'library prep protocol', 'library preparation method', 'library selection', 'library source', 'library strategy', 'library-kit', 'line', 'line source', 'longitude', 'lysis buffer ion concentrations', 'lysis strategy', 'machine', 'marker', 'maternal genotype', 'mating status', 'mating system', 'mean insert size', 'microbiome status', 'midgut part', 'midgut region', 'molecule subtype', 'monosome enrichment strategy', 'muscle type', 'normalization', 'notes', 'number cells', 'oleic acid', 'organism', 'organism part', 'original pipeline', 'paired or single-end', 'parental heterozygosity', 'parental obesity', 'passage', 'passages', 'pe read length (bp)', 'phenotype', 'plasmid', 'plate and well id', 'platform', 'population', 'population diet', 'population replicate', 'project name', 'protocol', 'pull-down tag', 'purified length', 'quality', 'qualscores', 'quantity', 'radiation dose', 'random barcode', 'reared on', 'repeat', 'replicate', 'replicate no.', 'replicates', 'resistance', 'rin', 'rip antibody', 'rna', 'rna extracted from', 'rna fraction', 'rna interference', 'rna quantity', 'rna source', 'rna subtype', 'rnai', 'rnai driver', 'rnai reagent from harvard drosophila rnai screening center', 'rnai target gene name', 'rnai treatment', 'rrna depletion method', 'rt primer with random barcode/index', 'run no.', 's2 treatment', 'sample comment', 'sample extraction method', 'sample group', 'sample name', 'sample note', 'sample title', 'sample type', 'selection', 'selection method', 'sequencer', 'sequencing method', 'sex', 'sex composition', 'sex/age', 'shrna', 'silencing', 'similar to', 'sirna knockdown of', 'source', 'source name', 'species', 'specimen voucher', 'specimen with known storage state', 'stable line', 'stage', 'starvation status', 'stimulation', 'stock', 'store cond', 'strain', 'strain background', 'strain information', 'strain origin', 'strain/background', 'strain/genotype', 'strain/sex', 'strandedness', 'stress', 'stud book number', 'subcellular fraction', 'subregion', 'tag', 'target molecule', 'technical replicate', 'technical replicates pooled together', 'techrep', 'temperature', 'temperature shift', 'temperature tested', 'time', 'time of treatment', 'time point', 'time post-treatment', 'time-point', 'timepoint', 'timepoint/developmental stage', 'tissue', 'tissue type', 'tissue-specific knockdown', 'training', 'transfected sirna', 'transfected with', 'transfecting vector', 'transfector', 'treated with', 'treatment', 'treatment duration', 'tribe expression', 'tribe protein expression', 'type', 'unknown', 'variation', 'vdrc transformant id', 'virus infection', 'x chromosome dose', 'zygotic genotype']\n"
     ]
    }
   ],
   "source": [
    "print(sorted(attrsp.columns.tolist()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 243,
   "metadata": {},
   "outputs": [],
   "source": [
    "#print(sorted(attrs.value.unique().tolist()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_sex(attrsp, metadata):\n",
    "    dat = attrsp['sex']\n",
    "\n",
    "    with open('../geo-wf/config/sex.yaml') as fh:\n",
    "        sex_norm = load(fh)\n",
    "\n",
    "    for k, v in sex_norm.items():\n",
    "        if v == 'None':\n",
    "            sex_norm[k] = np.nan\n",
    "\n",
    "    dat.replace(sex_norm, inplace=True)\n",
    "    metadata.update(dat.dropna())\n",
    "        \n",
    "    missing = metadata.sex.isna().index.tolist()\n",
    "    for srx in missing:\n",
    "        try:\n",
    "            string = bow[srx]\n",
    "        except KeyError:\n",
    "            continue\n",
    "\n",
    "        mf = re.search(r'\\bfemale\\b', string)\n",
    "        mm = re.search(r'\\bmale\\b', string)\n",
    "        if (mf is not None) & (mm is not None):\n",
    "            metadata.loc[srx, 'sex'] = 'mixed'\n",
    "        elif mf:\n",
    "            metadata.loc[srx, 'sex'] = 'female'\n",
    "        elif mm:\n",
    "            metadata.loc[srx, 'sex'] = 'male'\n",
    "            \n",
    "parse_sex(attrsp, metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_dev(attrsp, metadata):\n",
    "    dat = attrsp['developmental stage']\n",
    "\n",
    "    with open('../geo-wf/config/dev_stage.yaml') as fh:\n",
    "        dev_norm = load(fh)\n",
    "\n",
    "    for k, v in dev_norm.items():\n",
    "        if v == 'None':\n",
    "            dev_norm[k] = np.nan\n",
    "\n",
    "    dat.replace(dev_norm, inplace=True)\n",
    "    metadata.update(dat.dropna())\n",
    "            \n",
    "parse_dev(attrsp, metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_tissue(attrsp, metadata):\n",
    "    dat = attrsp['tissue']\n",
    "\n",
    "    with open('../geo-wf/config/tissue.yaml') as fh:\n",
    "        tissue_norm = load(fh)\n",
    "\n",
    "    for k, v in tissue_norm.items():\n",
    "        if v == 'None':\n",
    "            tissue_norm[k] = np.nan\n",
    "\n",
    "    dat.replace(tissue_norm, inplace=True)\n",
    "    metadata.update(dat.dropna())\n",
    "            \n",
    "parse_tissue(attrsp, metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "attrs = pd.DataFrame(list(ncbi.aggregate([\n",
    "    {\n",
    "        '$match': {\n",
    "            '_id': {'$in': sample_list}\n",
    "        }\n",
    "    },\n",
    "    {\n",
    "        '$unwind': {\n",
    "            'path': '$sra.sample.attributes'\n",
    "        }\n",
    "    },\n",
    "    {\n",
    "        '$project': {\n",
    "            '_id': 0,\n",
    "            'srx': '$_id',\n",
    "            'name': '$sra.sample.attributes.name',\n",
    "            'value': '$sra.sample.attributes.value'\n",
    "        }\n",
    "    },\n",
    "])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(70753, 3)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attrs.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initial normalization\n",
    "replace_dict = {\n",
    "    '_': ' ', \n",
    "    '; ': ';', \n",
    "    ' / ': '/',\n",
    "    'cell/tissue type': 'cell type',\n",
    "    'tissue/cell type': 'cell type',\n",
    "    'cell class': 'cell type', \n",
    "    'cell line': 'cell type', \n",
    "    'cell line background': 'cell type', \n",
    "    'cell subtype': 'cell type', \n",
    "    'cell type': 'cell type', \n",
    "    'cells': 'cell type', \n",
    "    'cells derived from': 'cell type',\n",
    "}\n",
    "regex = re.compile('|'.join(replace_dict.keys()))\n",
    "\n",
    "attrs['name'] = attrs['name'].str.lower().apply(lambda s: regex.sub(lambda x: replace_dict[x.group()], s))\n",
    "attrs['value'] = attrs['value'].str.lower().apply(lambda s: regex.sub(lambda x: replace_dict[x.group()], s))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "attrsp = attrs.pivot_table(values='value', index='srx', columns='name', aggfunc='first').copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "bow = {}\n",
    "for srx, row in attrsp.iterrows():\n",
    "    bow[srx] = ' '.join(row.d.opna().values.tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_cell_type(attrsp, metadata):\n",
    "    dat = attrsp['cell type']\n",
    "\n",
    "    with open('../geo-wf/config/cell_type.yaml') as fh:\n",
    "        norm = load(fh)\n",
    "\n",
    "    for k, v in norm.items():\n",
    "        if v == 'None':\n",
    "            norm[k] = np.nan\n",
    "\n",
    "    dat.replace(norm, inplace=True)\n",
    "    metadata.update(dat.dropna())\n",
    "            \n",
    "parse_cell_type(attrsp, metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ERX173563    S2\n",
       "ERX173565    S2\n",
       "ERX173570    S2\n",
       "ERX173571    S2\n",
       "ERX173572    S2\n",
       "ERX173578    S2\n",
       "ERX173581    S2\n",
       "ERX173588    S2\n",
       "ERX173594    S2\n",
       "ERX173597    S2\n",
       "ERX173598    S2\n",
       "ERX173601    S2\n",
       "ERX173605    S2\n",
       "ERX173607    S2\n",
       "ERX173611    S2\n",
       "ERX173614    S2\n",
       "ERX173615    S2\n",
       "ERX173617    S2\n",
       "ERX173618    S2\n",
       "ERX173625    S2\n",
       "ERX173631    S2\n",
       "ERX173640    S2\n",
       "ERX173642    S2\n",
       "ERX173647    S2\n",
       "ERX173651    S2\n",
       "ERX173658    S2\n",
       "ERX173659    S2\n",
       "ERX173661    S2\n",
       "ERX173664    S2\n",
       "ERX173668    S2\n",
       "             ..\n",
       "SRX976314    S3\n",
       "SRX976315    S3\n",
       "SRX976316    S3\n",
       "SRX976317    S3\n",
       "SRX976318    S3\n",
       "SRX976319    S3\n",
       "SRX976320    S3\n",
       "SRX976321    S3\n",
       "SRX976322    S3\n",
       "SRX976323    S3\n",
       "SRX976324    S3\n",
       "SRX976325    S3\n",
       "SRX976326    S3\n",
       "SRX981410    S2\n",
       "SRX981411    S2\n",
       "SRX981412    S2\n",
       "SRX981413    S2\n",
       "SRX981414    S2\n",
       "SRX981415    S2\n",
       "SRX982548    S2\n",
       "SRX982550    S2\n",
       "SRX982551    S2\n",
       "SRX982552    S2\n",
       "SRX982553    S2\n",
       "SRX984881    S2\n",
       "SRX984882    S2\n",
       "SRX984883    S2\n",
       "SRX984887    S2\n",
       "SRX984888    S2\n",
       "SRX984889    S2\n",
       "Name: cell type, Length: 3560, dtype: object"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata['cell type'].dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sex</th>\n",
       "      <th>developmental stage</th>\n",
       "      <th>tissue</th>\n",
       "      <th>strain</th>\n",
       "      <th>cell type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DRX013093</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRX013094</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRX014765</th>\n",
       "      <td>female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRX014766</th>\n",
       "      <td>female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRX014767</th>\n",
       "      <td>male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRX014768</th>\n",
       "      <td>male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRX014769</th>\n",
       "      <td>female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRX014770</th>\n",
       "      <td>female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRX014771</th>\n",
       "      <td>male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRX014772</th>\n",
       "      <td>male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRX014773</th>\n",
       "      <td>female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRX014774</th>\n",
       "      <td>female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRX014775</th>\n",
       "      <td>male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRX014776</th>\n",
       "      <td>male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRX014777</th>\n",
       "      <td>female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRX014778</th>\n",
       "      <td>female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRX014779</th>\n",
       "      <td>male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRX014780</th>\n",
       "      <td>male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRX014781</th>\n",
       "      <td>female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRX014782</th>\n",
       "      <td>female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRX014783</th>\n",
       "      <td>male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRX014784</th>\n",
       "      <td>male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRX014785</th>\n",
       "      <td>female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRX014786</th>\n",
       "      <td>female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRX014787</th>\n",
       "      <td>male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRX014788</th>\n",
       "      <td>male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRX014789</th>\n",
       "      <td>female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRX014790</th>\n",
       "      <td>female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRX014791</th>\n",
       "      <td>male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRX014792</th>\n",
       "      <td>male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX982552</th>\n",
       "      <td>female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX982553</th>\n",
       "      <td>female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX984881</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX984882</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX984883</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX984887</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX984888</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX984889</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX986397</th>\n",
       "      <td>female</td>\n",
       "      <td>adult</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX986398</th>\n",
       "      <td>female</td>\n",
       "      <td>adult</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX986399</th>\n",
       "      <td>female</td>\n",
       "      <td>adult</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX986400</th>\n",
       "      <td>female</td>\n",
       "      <td>adult</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX986401</th>\n",
       "      <td>female</td>\n",
       "      <td>adult</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX986402</th>\n",
       "      <td>male</td>\n",
       "      <td>adult</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX986403</th>\n",
       "      <td>male</td>\n",
       "      <td>adult</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX986404</th>\n",
       "      <td>male</td>\n",
       "      <td>adult</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX986405</th>\n",
       "      <td>male</td>\n",
       "      <td>adult</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX986406</th>\n",
       "      <td>male</td>\n",
       "      <td>adult</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX986407</th>\n",
       "      <td>male</td>\n",
       "      <td>adult</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX986408</th>\n",
       "      <td>female</td>\n",
       "      <td>adult</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX986409</th>\n",
       "      <td>female</td>\n",
       "      <td>adult</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX986410</th>\n",
       "      <td>female</td>\n",
       "      <td>adult</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX986411</th>\n",
       "      <td>female</td>\n",
       "      <td>adult</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX986412</th>\n",
       "      <td>female</td>\n",
       "      <td>adult</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX986413</th>\n",
       "      <td>male</td>\n",
       "      <td>adult</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX986414</th>\n",
       "      <td>male</td>\n",
       "      <td>adult</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX986415</th>\n",
       "      <td>male</td>\n",
       "      <td>adult</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX986416</th>\n",
       "      <td>male</td>\n",
       "      <td>adult</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX986417</th>\n",
       "      <td>male</td>\n",
       "      <td>adult</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRX986418</th>\n",
       "      <td>male</td>\n",
       "      <td>adult</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14423 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              sex developmental stage tissue strain cell type\n",
       "DRX013093     NaN                 NaN    NaN    NaN       NaN\n",
       "DRX013094     NaN                 NaN    NaN    NaN       NaN\n",
       "DRX014765  female                 NaN    NaN    NaN       NaN\n",
       "DRX014766  female                 NaN    NaN    NaN       NaN\n",
       "DRX014767    male                 NaN    NaN    NaN       NaN\n",
       "DRX014768    male                 NaN    NaN    NaN       NaN\n",
       "DRX014769  female                 NaN    NaN    NaN       NaN\n",
       "DRX014770  female                 NaN    NaN    NaN       NaN\n",
       "DRX014771    male                 NaN    NaN    NaN       NaN\n",
       "DRX014772    male                 NaN    NaN    NaN       NaN\n",
       "DRX014773  female                 NaN    NaN    NaN       NaN\n",
       "DRX014774  female                 NaN    NaN    NaN       NaN\n",
       "DRX014775    male                 NaN    NaN    NaN       NaN\n",
       "DRX014776    male                 NaN    NaN    NaN       NaN\n",
       "DRX014777  female                 NaN    NaN    NaN       NaN\n",
       "DRX014778  female                 NaN    NaN    NaN       NaN\n",
       "DRX014779    male                 NaN    NaN    NaN       NaN\n",
       "DRX014780    male                 NaN    NaN    NaN       NaN\n",
       "DRX014781  female                 NaN    NaN    NaN       NaN\n",
       "DRX014782  female                 NaN    NaN    NaN       NaN\n",
       "DRX014783    male                 NaN    NaN    NaN       NaN\n",
       "DRX014784    male                 NaN    NaN    NaN       NaN\n",
       "DRX014785  female                 NaN    NaN    NaN       NaN\n",
       "DRX014786  female                 NaN    NaN    NaN       NaN\n",
       "DRX014787    male                 NaN    NaN    NaN       NaN\n",
       "DRX014788    male                 NaN    NaN    NaN       NaN\n",
       "DRX014789  female                 NaN    NaN    NaN       NaN\n",
       "DRX014790  female                 NaN    NaN    NaN       NaN\n",
       "DRX014791    male                 NaN    NaN    NaN       NaN\n",
       "DRX014792    male                 NaN    NaN    NaN       NaN\n",
       "...           ...                 ...    ...    ...       ...\n",
       "SRX982552  female                 NaN    NaN    NaN        S2\n",
       "SRX982553  female                 NaN    NaN    NaN        S2\n",
       "SRX984881     NaN                 NaN    NaN    NaN        S2\n",
       "SRX984882     NaN                 NaN    NaN    NaN        S2\n",
       "SRX984883     NaN                 NaN    NaN    NaN        S2\n",
       "SRX984887     NaN                 NaN    NaN    NaN        S2\n",
       "SRX984888     NaN                 NaN    NaN    NaN        S2\n",
       "SRX984889     NaN                 NaN    NaN    NaN        S2\n",
       "SRX986397  female               adult    NaN    NaN       NaN\n",
       "SRX986398  female               adult    NaN    NaN       NaN\n",
       "SRX986399  female               adult    NaN    NaN       NaN\n",
       "SRX986400  female               adult    NaN    NaN       NaN\n",
       "SRX986401  female               adult    NaN    NaN       NaN\n",
       "SRX986402    male               adult    NaN    NaN       NaN\n",
       "SRX986403    male               adult    NaN    NaN       NaN\n",
       "SRX986404    male               adult    NaN    NaN       NaN\n",
       "SRX986405    male               adult    NaN    NaN       NaN\n",
       "SRX986406    male               adult    NaN    NaN       NaN\n",
       "SRX986407    male               adult    NaN    NaN       NaN\n",
       "SRX986408  female               adult    NaN    NaN       NaN\n",
       "SRX986409  female               adult    NaN    NaN       NaN\n",
       "SRX986410  female               adult    NaN    NaN       NaN\n",
       "SRX986411  female               adult    NaN    NaN       NaN\n",
       "SRX986412  female               adult    NaN    NaN       NaN\n",
       "SRX986413    male               adult    NaN    NaN       NaN\n",
       "SRX986414    male               adult    NaN    NaN       NaN\n",
       "SRX986415    male               adult    NaN    NaN       NaN\n",
       "SRX986416    male               adult    NaN    NaN       NaN\n",
       "SRX986417    male               adult    NaN    NaN       NaN\n",
       "SRX986418    male               adult    NaN    NaN       NaN\n",
       "\n",
       "[14423 rows x 5 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 315,
   "metadata": {},
   "outputs": [],
   "source": [
    "bowS = pd.Series(bow)\n",
    "bowS.name = 'bow'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 332,
   "metadata": {},
   "outputs": [],
   "source": [
    "helper = ''\n",
    "for dv, bb in devs.join(bowS, how='left').groupby('developmental stage'):\n",
    "    mylist = ' '.join(bb.bow.unique().tolist())\n",
    "    helper += f'{dv}\\t{mylist}\\n'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 333,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/tmp/helper.tsv', 'w') as fo:\n",
    "    fo.write(helper)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:ncbi_remap]",
   "language": "python",
   "name": "conda-env-ncbi_remap-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
