"""Downstream analysis for RNA-Seq data.

Workflow for generating downstream files on RNA-Seq data specifically.
"""
from ncbi_remap.snakemake import wrapper_for, put_flag, get_patterns

# Setup tempdir to work with lscratch
TMPDIR = os.path.join('/lscratch', os.getenv('SLURM_JOBID'))
shell.prefix("set -euo pipefail; export TMPDIR={};".format(TMPDIR))

# Set working dir
workdir: '.'

# import config
configfile: '../config/reference_config.yaml'

patterns = {
    'stringtie': '../output/rnaseq-analysis-wf/{srx}.stringtie.gtf',
}

rule targets:
    input: expand(patterns['stringtie'], srx=srxs)


rule stringTie:
    input:
        bam = '../output/aln-wf/samples/{srx}/{srx}.bam',
        strand = '../output/prealn-wf/samples/{srx}/STRAND',
        gtf = config['references']['dmel']['gtf'],
    output: patterns['stringtie']
    threads: 2
    resources:
      mem_gb=lambda wildcards, attempt: attempt * 1,
      time_hr=lambda wildcards, attempt: attempt * 1
    script: 'scripts/stringtie.py'

