{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the s2RNAi project I need a list of ChIP-seq samples. I am doing it here because this is related to SRA project."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "last updated: 2017-09-20 \n",
      "Git hash: 4bc9218af8c49c09524cdcb0c0e81f7e74f4da46\n"
     ]
    }
   ],
   "source": [
    "# %load ../start.py\n",
    "# Load useful extensions\n",
    "\n",
    "# Activate the autoreload extension for easy reloading of external packages\n",
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "# Trun on the water mark\n",
    "%reload_ext watermark\n",
    "%watermark -u -d -g\n",
    "\n",
    "# Load ipycache extension\n",
    "%reload_ext ipycache\n",
    "from ipycache import CacheMagics\n",
    "CacheMagics.cachedir = '../cachedir'\n",
    "\n",
    "# Add project library to path\n",
    "import sys\n",
    "sys.path.insert(0, '../../lib/python')\n",
    "\n",
    "# The usual suspects\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "# plotting\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "%matplotlib inline\n",
    "sns.set_context('poster')\n",
    "\n",
    "# Turn off scientific notation\n",
    "np.set_printoptions(precision=5, suppress=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "code_folding": [
     0
    ],
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# %load ../../bin/load.py\n",
    "from pymongo import MongoClient\n",
    "with open('/home/fearjm/Projects/ncbi_remap/output/.mongodb_host', 'r') as fh:\n",
    "    host = fh.read().strip()\n",
    "client = MongoClient(host=host, port=27022)\n",
    "db = client['sra2']\n",
    "ncbi = db['ncbi']\n",
    "remap = db['remap']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "code_folding": [],
    "collapsed": true,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "df = pd.DataFrame(list(ncbi.aggregate([\n",
    "    {'$unwind': '$biosample'},\n",
    "    {'$unwind': '$biosample.attributes'},\n",
    "    {\n",
    "        '$match': {\n",
    "            'biosample.attributes.value': {'$regex': '(S2[\\sr\\+\\-]+|.*Schneider.*)', '$options': 'i'},\n",
    "        }\n",
    "    },\n",
    "    {\n",
    "        '$project': {\n",
    "            '_id': 0,\n",
    "            'srx': '$_id',\n",
    "            'biosample': '$biosample.biosample_accn',\n",
    "            'name': '$biosample.attributes.name',\n",
    "            'value': '$biosample.attributes.value',\n",
    "            'sra_strategy': '$sra.experiment.library_strategy',\n",
    "            'sra_source': '$sra.experiment.library_source',\n",
    "            'sra_selection': '$sra.experiment.library_selection',\n",
    "        }\n",
    "    },\n",
    "])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3877, 7)\n",
      "(3877, 7)\n"
     ]
    }
   ],
   "source": [
    "print(df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biosample</th>\n",
       "      <th>name</th>\n",
       "      <th>sra_selection</th>\n",
       "      <th>sra_source</th>\n",
       "      <th>sra_strategy</th>\n",
       "      <th>srx</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>SAMN05710807</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>ChIP</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>SRX2055967</td>\n",
       "      <td>S2-DRSC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>SAMN05710808</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>ChIP</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>SRX2055966</td>\n",
       "      <td>S2-DRSC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>SAMN05710809</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>ChIP</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>SRX2055965</td>\n",
       "      <td>S2-DRSC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>SAMN05710810</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>ChIP</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>SRX2055964</td>\n",
       "      <td>S2-DRSC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>SAMN05710811</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>ChIP</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>SRX2055963</td>\n",
       "      <td>S2-DRSC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>SAMN05710812</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>ChIP</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>SRX2055962</td>\n",
       "      <td>S2-DRSC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>SAMN05710813</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>ChIP</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>SRX2055961</td>\n",
       "      <td>S2-DRSC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>SAMN05710814</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>ChIP</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>SRX2055960</td>\n",
       "      <td>S2-DRSC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>SAMN05710815</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>ChIP</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>SRX2055959</td>\n",
       "      <td>S2-DRSC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>SAMN05710816</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>ChIP</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>SRX2055958</td>\n",
       "      <td>S2-DRSC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>SAMN05710818</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>ChIP</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>SRX2055957</td>\n",
       "      <td>S2-DRSC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>SAMN05710819</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>ChIP</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>SRX2055956</td>\n",
       "      <td>S2-DRSC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>SAMN05710820</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>ChIP</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>SRX2055955</td>\n",
       "      <td>S2-DRSC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>SAMN05710821</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>ChIP</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>SRX2055954</td>\n",
       "      <td>S2-DRSC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>SAMN05710822</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>ChIP</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>SRX2055953</td>\n",
       "      <td>S2-DRSC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>SAMN05710817</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>ChIP</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>SRX2055952</td>\n",
       "      <td>S2-DRSC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>SAMN05710823</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>ChIP</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>SRX2055951</td>\n",
       "      <td>S2-DRSC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>SAMN05710824</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>ChIP</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>SRX2055950</td>\n",
       "      <td>S2-DRSC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>SAMN05710804</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>ChIP</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>SRX2055949</td>\n",
       "      <td>S2-DRSC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>SAMN05710829</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>ChIP</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>SRX2055948</td>\n",
       "      <td>S2-DRSC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>SAMN05710828</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>ChIP</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>SRX2055947</td>\n",
       "      <td>S2-DRSC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>SAMN05710826</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>ChIP</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>SRX2055946</td>\n",
       "      <td>S2-DRSC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>SAMN05710825</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>ChIP</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>SRX2055945</td>\n",
       "      <td>S2-DRSC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>SAMN05710805</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>ChIP</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>SRX2055944</td>\n",
       "      <td>S2-DRSC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>SAMN05710806</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>ChIP</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>SRX2055943</td>\n",
       "      <td>S2-DRSC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>SAMN05509059</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>other</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>OTHER</td>\n",
       "      <td>SRX1997361</td>\n",
       "      <td>S2 cell line</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>SAMN05509057</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>other</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>OTHER</td>\n",
       "      <td>SRX1997360</td>\n",
       "      <td>S2 cell line</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>SAMN05509058</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>other</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>OTHER</td>\n",
       "      <td>SRX1997359</td>\n",
       "      <td>S2 cell line</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>SAMN05509060</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>other</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>OTHER</td>\n",
       "      <td>SRX1997358</td>\n",
       "      <td>S2 cell line</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>SAMN05464811</td>\n",
       "      <td>cell_type</td>\n",
       "      <td>other</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>OTHER</td>\n",
       "      <td>SRX1989900</td>\n",
       "      <td>S2 cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3715</th>\n",
       "      <td>SAMN04503041</td>\n",
       "      <td>cell_type</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>SRX1592050</td>\n",
       "      <td>s2 cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3717</th>\n",
       "      <td>SAMN04503040</td>\n",
       "      <td>cell_type</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>SRX1592049</td>\n",
       "      <td>s2 cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3719</th>\n",
       "      <td>SAMN04503039</td>\n",
       "      <td>cell_type</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>SRX1592048</td>\n",
       "      <td>s2 cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3721</th>\n",
       "      <td>SAMN04503038</td>\n",
       "      <td>cell_type</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>SRX1592047</td>\n",
       "      <td>s2 cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3723</th>\n",
       "      <td>SAMN04503037</td>\n",
       "      <td>cell_type</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>SRX1592046</td>\n",
       "      <td>s2 cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3725</th>\n",
       "      <td>SAMN04503036</td>\n",
       "      <td>cell_type</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>SRX1592045</td>\n",
       "      <td>s2 cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3727</th>\n",
       "      <td>SAMN04503035</td>\n",
       "      <td>cell_type</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>SRX1592044</td>\n",
       "      <td>s2 cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3729</th>\n",
       "      <td>SAMN04503034</td>\n",
       "      <td>cell_type</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>SRX1592043</td>\n",
       "      <td>s2 cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3731</th>\n",
       "      <td>SAMN04503033</td>\n",
       "      <td>cell_type</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>SRX1592042</td>\n",
       "      <td>s2 cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3733</th>\n",
       "      <td>SAMN04503032</td>\n",
       "      <td>cell_type</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>SRX1592041</td>\n",
       "      <td>s2 cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3735</th>\n",
       "      <td>SAMN04503031</td>\n",
       "      <td>cell_type</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>SRX1592040</td>\n",
       "      <td>s2 cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3737</th>\n",
       "      <td>SAMN04503030</td>\n",
       "      <td>cell_type</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>SRX1592039</td>\n",
       "      <td>s2 cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3739</th>\n",
       "      <td>SAMN04503029</td>\n",
       "      <td>cell_type</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>SRX1592038</td>\n",
       "      <td>s2 cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3741</th>\n",
       "      <td>SAMN04503028</td>\n",
       "      <td>cell_type</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>SRX1592037</td>\n",
       "      <td>s2 cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3743</th>\n",
       "      <td>SAMN04503027</td>\n",
       "      <td>cell_type</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>SRX1592036</td>\n",
       "      <td>s2 cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3745</th>\n",
       "      <td>SAMN04503026</td>\n",
       "      <td>cell_type</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>SRX1592035</td>\n",
       "      <td>s2 cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3747</th>\n",
       "      <td>SAMN04503025</td>\n",
       "      <td>cell_type</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>SRX1592034</td>\n",
       "      <td>s2 cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3749</th>\n",
       "      <td>SAMN04503024</td>\n",
       "      <td>cell_type</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>SRX1592033</td>\n",
       "      <td>s2 cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3751</th>\n",
       "      <td>SAMN04503023</td>\n",
       "      <td>cell_type</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>SRX1592032</td>\n",
       "      <td>s2 cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3753</th>\n",
       "      <td>SAMN04503022</td>\n",
       "      <td>cell_type</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>SRX1592031</td>\n",
       "      <td>s2 cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3776</th>\n",
       "      <td>SAMN04351215</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>SRX1492036</td>\n",
       "      <td>S2R+ embryonic cell line</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3778</th>\n",
       "      <td>SAMN04351214</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>SRX1492035</td>\n",
       "      <td>S2R+ embryonic cell line</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3862</th>\n",
       "      <td>SAMEA2629803</td>\n",
       "      <td>strain</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>ERX517283</td>\n",
       "      <td>S2R+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3864</th>\n",
       "      <td>SAMEA2629800</td>\n",
       "      <td>strain</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>ERX517280</td>\n",
       "      <td>S2R+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3866</th>\n",
       "      <td>SAMEA2629798</td>\n",
       "      <td>strain</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>ERX517278</td>\n",
       "      <td>S2R+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3868</th>\n",
       "      <td>SAMEA2629797</td>\n",
       "      <td>strain</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>ERX517277</td>\n",
       "      <td>S2R+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3870</th>\n",
       "      <td>SAMEA2629795</td>\n",
       "      <td>strain</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>ERX517275</td>\n",
       "      <td>S2R+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3872</th>\n",
       "      <td>SAMEA2629794</td>\n",
       "      <td>strain</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>ERX517274</td>\n",
       "      <td>S2R+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3874</th>\n",
       "      <td>SAMEA2629793</td>\n",
       "      <td>strain</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>ERX517273</td>\n",
       "      <td>S2R+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3876</th>\n",
       "      <td>SAMEA2629792</td>\n",
       "      <td>strain</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>ERX517272</td>\n",
       "      <td>S2R+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>661 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         biosample       name sra_selection      sra_source sra_strategy  \\\n",
       "16    SAMN05710807  cell_line          ChIP         GENOMIC     ChIP-Seq   \n",
       "18    SAMN05710808  cell_line          ChIP         GENOMIC     ChIP-Seq   \n",
       "20    SAMN05710809  cell_line          ChIP         GENOMIC     ChIP-Seq   \n",
       "22    SAMN05710810  cell_line          ChIP         GENOMIC     ChIP-Seq   \n",
       "24    SAMN05710811  cell_line          ChIP         GENOMIC     ChIP-Seq   \n",
       "26    SAMN05710812  cell_line          ChIP         GENOMIC     ChIP-Seq   \n",
       "28    SAMN05710813  cell_line          ChIP         GENOMIC     ChIP-Seq   \n",
       "30    SAMN05710814  cell_line          ChIP         GENOMIC     ChIP-Seq   \n",
       "32    SAMN05710815  cell_line          ChIP         GENOMIC     ChIP-Seq   \n",
       "34    SAMN05710816  cell_line          ChIP         GENOMIC     ChIP-Seq   \n",
       "36    SAMN05710818  cell_line          ChIP         GENOMIC     ChIP-Seq   \n",
       "38    SAMN05710819  cell_line          ChIP         GENOMIC     ChIP-Seq   \n",
       "40    SAMN05710820  cell_line          ChIP         GENOMIC     ChIP-Seq   \n",
       "42    SAMN05710821  cell_line          ChIP         GENOMIC     ChIP-Seq   \n",
       "44    SAMN05710822  cell_line          ChIP         GENOMIC     ChIP-Seq   \n",
       "46    SAMN05710817  cell_line          ChIP         GENOMIC     ChIP-Seq   \n",
       "48    SAMN05710823  cell_line          ChIP         GENOMIC     ChIP-Seq   \n",
       "50    SAMN05710824  cell_line          ChIP         GENOMIC     ChIP-Seq   \n",
       "52    SAMN05710804  cell_line          ChIP         GENOMIC     ChIP-Seq   \n",
       "54    SAMN05710829  cell_line          ChIP         GENOMIC     ChIP-Seq   \n",
       "56    SAMN05710828  cell_line          ChIP         GENOMIC     ChIP-Seq   \n",
       "58    SAMN05710826  cell_line          ChIP         GENOMIC     ChIP-Seq   \n",
       "60    SAMN05710825  cell_line          ChIP         GENOMIC     ChIP-Seq   \n",
       "62    SAMN05710805  cell_line          ChIP         GENOMIC     ChIP-Seq   \n",
       "64    SAMN05710806  cell_line          ChIP         GENOMIC     ChIP-Seq   \n",
       "70    SAMN05509059  cell_line         other  TRANSCRIPTOMIC        OTHER   \n",
       "72    SAMN05509057  cell_line         other  TRANSCRIPTOMIC        OTHER   \n",
       "74    SAMN05509058  cell_line         other  TRANSCRIPTOMIC        OTHER   \n",
       "76    SAMN05509060  cell_line         other  TRANSCRIPTOMIC        OTHER   \n",
       "78    SAMN05464811  cell_type         other  TRANSCRIPTOMIC        OTHER   \n",
       "...            ...        ...           ...             ...          ...   \n",
       "3715  SAMN04503041  cell_type          cDNA  TRANSCRIPTOMIC      RNA-Seq   \n",
       "3717  SAMN04503040  cell_type          cDNA  TRANSCRIPTOMIC      RNA-Seq   \n",
       "3719  SAMN04503039  cell_type          cDNA  TRANSCRIPTOMIC      RNA-Seq   \n",
       "3721  SAMN04503038  cell_type          cDNA  TRANSCRIPTOMIC      RNA-Seq   \n",
       "3723  SAMN04503037  cell_type          cDNA  TRANSCRIPTOMIC      RNA-Seq   \n",
       "3725  SAMN04503036  cell_type          cDNA  TRANSCRIPTOMIC      RNA-Seq   \n",
       "3727  SAMN04503035  cell_type          cDNA  TRANSCRIPTOMIC      RNA-Seq   \n",
       "3729  SAMN04503034  cell_type          cDNA  TRANSCRIPTOMIC      RNA-Seq   \n",
       "3731  SAMN04503033  cell_type          cDNA  TRANSCRIPTOMIC      RNA-Seq   \n",
       "3733  SAMN04503032  cell_type          cDNA  TRANSCRIPTOMIC      RNA-Seq   \n",
       "3735  SAMN04503031  cell_type          cDNA  TRANSCRIPTOMIC      RNA-Seq   \n",
       "3737  SAMN04503030  cell_type          cDNA  TRANSCRIPTOMIC      RNA-Seq   \n",
       "3739  SAMN04503029  cell_type          cDNA  TRANSCRIPTOMIC      RNA-Seq   \n",
       "3741  SAMN04503028  cell_type          cDNA  TRANSCRIPTOMIC      RNA-Seq   \n",
       "3743  SAMN04503027  cell_type          cDNA  TRANSCRIPTOMIC      RNA-Seq   \n",
       "3745  SAMN04503026  cell_type          cDNA  TRANSCRIPTOMIC      RNA-Seq   \n",
       "3747  SAMN04503025  cell_type          cDNA  TRANSCRIPTOMIC      RNA-Seq   \n",
       "3749  SAMN04503024  cell_type          cDNA  TRANSCRIPTOMIC      RNA-Seq   \n",
       "3751  SAMN04503023  cell_type          cDNA  TRANSCRIPTOMIC      RNA-Seq   \n",
       "3753  SAMN04503022  cell_type          cDNA  TRANSCRIPTOMIC      RNA-Seq   \n",
       "3776  SAMN04351215  cell_line          cDNA  TRANSCRIPTOMIC      RNA-Seq   \n",
       "3778  SAMN04351214  cell_line          cDNA  TRANSCRIPTOMIC      RNA-Seq   \n",
       "3862  SAMEA2629803     strain        RANDOM  TRANSCRIPTOMIC      RNA-Seq   \n",
       "3864  SAMEA2629800     strain        RANDOM  TRANSCRIPTOMIC      RNA-Seq   \n",
       "3866  SAMEA2629798     strain        RANDOM  TRANSCRIPTOMIC      RNA-Seq   \n",
       "3868  SAMEA2629797     strain        RANDOM  TRANSCRIPTOMIC      RNA-Seq   \n",
       "3870  SAMEA2629795     strain        RANDOM  TRANSCRIPTOMIC      RNA-Seq   \n",
       "3872  SAMEA2629794     strain        RANDOM  TRANSCRIPTOMIC      RNA-Seq   \n",
       "3874  SAMEA2629793     strain        RANDOM  TRANSCRIPTOMIC      RNA-Seq   \n",
       "3876  SAMEA2629792     strain        RANDOM  TRANSCRIPTOMIC      RNA-Seq   \n",
       "\n",
       "             srx                     value  \n",
       "16    SRX2055967                   S2-DRSC  \n",
       "18    SRX2055966                   S2-DRSC  \n",
       "20    SRX2055965                   S2-DRSC  \n",
       "22    SRX2055964                   S2-DRSC  \n",
       "24    SRX2055963                   S2-DRSC  \n",
       "26    SRX2055962                   S2-DRSC  \n",
       "28    SRX2055961                   S2-DRSC  \n",
       "30    SRX2055960                   S2-DRSC  \n",
       "32    SRX2055959                   S2-DRSC  \n",
       "34    SRX2055958                   S2-DRSC  \n",
       "36    SRX2055957                   S2-DRSC  \n",
       "38    SRX2055956                   S2-DRSC  \n",
       "40    SRX2055955                   S2-DRSC  \n",
       "42    SRX2055954                   S2-DRSC  \n",
       "44    SRX2055953                   S2-DRSC  \n",
       "46    SRX2055952                   S2-DRSC  \n",
       "48    SRX2055951                   S2-DRSC  \n",
       "50    SRX2055950                   S2-DRSC  \n",
       "52    SRX2055949                   S2-DRSC  \n",
       "54    SRX2055948                   S2-DRSC  \n",
       "56    SRX2055947                   S2-DRSC  \n",
       "58    SRX2055946                   S2-DRSC  \n",
       "60    SRX2055945                   S2-DRSC  \n",
       "62    SRX2055944                   S2-DRSC  \n",
       "64    SRX2055943                   S2-DRSC  \n",
       "70    SRX1997361              S2 cell line  \n",
       "72    SRX1997360              S2 cell line  \n",
       "74    SRX1997359              S2 cell line  \n",
       "76    SRX1997358              S2 cell line  \n",
       "78    SRX1989900                  S2 cells  \n",
       "...          ...                       ...  \n",
       "3715  SRX1592050                   s2 cell  \n",
       "3717  SRX1592049                   s2 cell  \n",
       "3719  SRX1592048                   s2 cell  \n",
       "3721  SRX1592047                   s2 cell  \n",
       "3723  SRX1592046                   s2 cell  \n",
       "3725  SRX1592045                   s2 cell  \n",
       "3727  SRX1592044                   s2 cell  \n",
       "3729  SRX1592043                   s2 cell  \n",
       "3731  SRX1592042                   s2 cell  \n",
       "3733  SRX1592041                   s2 cell  \n",
       "3735  SRX1592040                   s2 cell  \n",
       "3737  SRX1592039                   s2 cell  \n",
       "3739  SRX1592038                   s2 cell  \n",
       "3741  SRX1592037                   s2 cell  \n",
       "3743  SRX1592036                   s2 cell  \n",
       "3745  SRX1592035                   s2 cell  \n",
       "3747  SRX1592034                   s2 cell  \n",
       "3749  SRX1592033                   s2 cell  \n",
       "3751  SRX1592032                   s2 cell  \n",
       "3753  SRX1592031                   s2 cell  \n",
       "3776  SRX1492036  S2R+ embryonic cell line  \n",
       "3778  SRX1492035  S2R+ embryonic cell line  \n",
       "3862   ERX517283                      S2R+  \n",
       "3864   ERX517280                      S2R+  \n",
       "3866   ERX517278                      S2R+  \n",
       "3868   ERX517277                      S2R+  \n",
       "3870   ERX517275                      S2R+  \n",
       "3872   ERX517274                      S2R+  \n",
       "3874   ERX517273                      S2R+  \n",
       "3876   ERX517272                      S2R+  \n",
       "\n",
       "[661 rows x 7 columns]"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df.srx.duplicated()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3877,)"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.srx.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "flip = df.pivot(values='value', columns='name')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>name</th>\n",
       "      <th>breed</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>dev_stage</th>\n",
       "      <th>genotype</th>\n",
       "      <th>growth_protocol</th>\n",
       "      <th>isolate</th>\n",
       "      <th>library source</th>\n",
       "      <th>sample group</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>source</th>\n",
       "      <th>source_name</th>\n",
       "      <th>strain</th>\n",
       "      <th>tissue</th>\n",
       "      <th>treatment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Drosophila S2 cells, dnaj1 KD</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Drosophila S2 cells, dnaj1 KD</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Drosophila S2 cells, mlf KD</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Drosophila S2 cells, mlf KD</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Drosophila S2 cells, lacZ KD</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "name breed cell_line cell_type dev_stage genotype growth_protocol isolate  \\\n",
       "0     None      None      None      None     None            None    None   \n",
       "1     None      None      None      None     None            None    None   \n",
       "2     None      None      None      None     None            None    None   \n",
       "3     None      None      None      None     None            None    None   \n",
       "4     None      None      None      None     None            None    None   \n",
       "\n",
       "name library source sample group sample_name sample_type source  \\\n",
       "0              None         None        None        None   None   \n",
       "1              None         None        None        None   None   \n",
       "2              None         None        None        None   None   \n",
       "3              None         None        None        None   None   \n",
       "4              None         None        None        None   None   \n",
       "\n",
       "name                    source_name strain tissue treatment  \n",
       "0     Drosophila S2 cells, dnaj1 KD   None   None      None  \n",
       "1     Drosophila S2 cells, dnaj1 KD   None   None      None  \n",
       "2       Drosophila S2 cells, mlf KD   None   None      None  \n",
       "3       Drosophila S2 cells, mlf KD   None   None      None  \n",
       "4      Drosophila S2 cells, lacZ KD   None   None      None  "
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "flip.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged = df.loc[:, ['srx', 'biosample', 'sra_selection', 'sra_source', 'sra_strategy']].join(flip)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>srx</th>\n",
       "      <th>biosample</th>\n",
       "      <th>sra_selection</th>\n",
       "      <th>sra_source</th>\n",
       "      <th>sra_strategy</th>\n",
       "      <th>breed</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>dev_stage</th>\n",
       "      <th>genotype</th>\n",
       "      <th>...</th>\n",
       "      <th>isolate</th>\n",
       "      <th>library source</th>\n",
       "      <th>sample group</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>source</th>\n",
       "      <th>source_name</th>\n",
       "      <th>strain</th>\n",
       "      <th>tissue</th>\n",
       "      <th>treatment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRX2172438</td>\n",
       "      <td>SAMN05784497</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Drosophila S2 cells, dnaj1 KD</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRX2172437</td>\n",
       "      <td>SAMN05784495</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Drosophila S2 cells, dnaj1 KD</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRX2172436</td>\n",
       "      <td>SAMN05784496</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Drosophila S2 cells, mlf KD</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRX2172435</td>\n",
       "      <td>SAMN05784491</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Drosophila S2 cells, mlf KD</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRX2172434</td>\n",
       "      <td>SAMN05784492</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Drosophila S2 cells, lacZ KD</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          srx     biosample sra_selection      sra_source sra_strategy breed  \\\n",
       "0  SRX2172438  SAMN05784497          cDNA  TRANSCRIPTOMIC      RNA-Seq  None   \n",
       "1  SRX2172437  SAMN05784495          cDNA  TRANSCRIPTOMIC      RNA-Seq  None   \n",
       "2  SRX2172436  SAMN05784496          cDNA  TRANSCRIPTOMIC      RNA-Seq  None   \n",
       "3  SRX2172435  SAMN05784491          cDNA  TRANSCRIPTOMIC      RNA-Seq  None   \n",
       "4  SRX2172434  SAMN05784492          cDNA  TRANSCRIPTOMIC      RNA-Seq  None   \n",
       "\n",
       "  cell_line cell_type dev_stage genotype    ...    isolate library source  \\\n",
       "0      None      None      None     None    ...       None           None   \n",
       "1      None      None      None     None    ...       None           None   \n",
       "2      None      None      None     None    ...       None           None   \n",
       "3      None      None      None     None    ...       None           None   \n",
       "4      None      None      None     None    ...       None           None   \n",
       "\n",
       "  sample group sample_name sample_type source                    source_name  \\\n",
       "0         None        None        None   None  Drosophila S2 cells, dnaj1 KD   \n",
       "1         None        None        None   None  Drosophila S2 cells, dnaj1 KD   \n",
       "2         None        None        None   None    Drosophila S2 cells, mlf KD   \n",
       "3         None        None        None   None    Drosophila S2 cells, mlf KD   \n",
       "4         None        None        None   None   Drosophila S2 cells, lacZ KD   \n",
       "\n",
       "  strain tissue treatment  \n",
       "0   None   None      None  \n",
       "1   None   None      None  \n",
       "2   None   None      None  \n",
       "3   None   None      None  \n",
       "4   None   None      None  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "150\n"
     ]
    }
   ],
   "source": [
    "isa = ['modENCODE_2626', 'modENCODE_2627', 'modENCODE_2629', 'modENCODE_2630',\n",
    "'modENCODE_2633', 'modENCODE_2634', 'modENCODE_2636', 'modENCODE_2638',\n",
    "'modENCODE_2639', 'modENCODE_2641', 'modENCODE_2642', 'modENCODE_3229',\n",
    "'modENCODE_3230', 'modENCODE_3232', 'modENCODE_3234', 'modENCODE_3237',\n",
    "'modENCODE_3238', 'modENCODE_3239', 'modENCODE_3240', 'modENCODE_3241',\n",
    "'modENCODE_3242', 'modENCODE_3243', 'modENCODE_3245', 'modENCODE_3390',\n",
    "'modENCODE_3391', 'modENCODE_3392', 'modENCODE_3393', 'modENCODE_3394',\n",
    "'modENCODE_3395', 'modENCODE_3396', 'modENCODE_3398', 'modENCODE_3399',\n",
    "'modENCODE_3400', 'modENCODE_3401', 'modENCODE_3402', 'modENCODE_3403',\n",
    "'modENCODE_3830', 'modENCODE_3956', 'modENCODE_3957', 'modENCODE_3958',\n",
    "'modENCODE_3959', 'modENCODE_3960', 'modENCODE_4069', 'modENCODE_4070',\n",
    "'modENCODE_4071', 'modENCODE_4074', 'modENCODE_4078', 'modENCODE_4080',\n",
    "'modENCODE_4081', 'modENCODE_4082', 'modENCODE_4089', 'modENCODE_4091',\n",
    "'modENCODE_4094', 'modENCODE_4095', 'modENCODE_4096', 'modENCODE_4098',\n",
    "'modENCODE_4099', 'modENCODE_4103', 'modENCODE_4104', 'modENCODE_4105',\n",
    "'modENCODE_4107', 'modENCODE_4113', 'modENCODE_4114', 'modENCODE_4119',\n",
    "'modENCODE_4352', 'modENCODE_4936', 'modENCODE_4943', 'modENCODE_4944',\n",
    "'modENCODE_4974', 'modENCODE_4976', 'modENCODE_4981', 'modENCODE_4982',\n",
    "'modENCODE_4998', 'modENCODE_5004', 'modENCODE_5005', 'modENCODE_5008',\n",
    "'modENCODE_5014', 'modENCODE_5017', 'modENCODE_5023', 'modENCODE_5024',\n",
    "'modENCODE_5025', 'modENCODE_5028', 'modENCODE_5029', 'modENCODE_5068',\n",
    "'modENCODE_5069', 'modENCODE_5070', 'modENCODE_5071', 'modENCODE_5072',\n",
    "'modENCODE_5110', 'modENCODE_5111', 'modENCODE_5112', 'modENCODE_5113',\n",
    "'modENCODE_5114', 'modENCODE_5115', 'modENCODE_5116', 'modENCODE_5117',\n",
    "'modENCODE_5118', 'modENCODE_5119', 'modENCODE_5120', 'modENCODE_5121',\n",
    "'modENCODE_5122', 'modENCODE_5123', 'modENCODE_5124', 'modENCODE_5125',\n",
    "'modENCODE_5126', 'modENCODE_5127', 'modENCODE_5128', 'modENCODE_5129',\n",
    "'modENCODE_5257', 'modENCODE_5264', 'modENCODE_5568', 'modENCODE_5569',\n",
    "'modENCODE_5570', 'modENCODE_5571', 'modENCODE_5574', 'modENCODE_5575',\n",
    "'modENCODE_5576', 'modENCODE_5577', 'modENCODE_5579', 'modENCODE_5580',\n",
    "'modENCODE_5587', 'modENCODE_5590', 'modENCODE_5591', 'modENCODE_5592',\n",
    "'modENCODE_5593', 'modENCODE_5594', 'modENCODE_5597', 'modENCODE_5598',\n",
    "'modENCODE_5599', 'modENCODE_5606', 'modENCODE_846', 'modENCODE_847',\n",
    "'modENCODE_848', 'modENCODE_849', 'modENCODE_850', 'modENCODE_851',\n",
    "'modENCODE_852', 'modENCODE_853', 'modENCODE_854', 'modENCODE_855',\n",
    "'modENCODE_856', 'modENCODE_857', 'modENCODE_858', 'modENCODE_859',\n",
    "'modENCODE_860', 'modENCODE_861', 'modENCODE_862', 'modENCODE_863',\n",
    "'modENCODE_984', 'modENCODE_985']\n",
    "print(len(isa))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "modencode = pd.read_table('../../output/modENCODE_sampletable.tsv')\n",
    "modencode.drop('sra_strategy', axis=1, inplace=True)\n",
    "modencode['modencode'] = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "805\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>srx</th>\n",
       "      <th>srr</th>\n",
       "      <th>modENCODE_id</th>\n",
       "      <th>modENCODE_assay</th>\n",
       "      <th>modENCODE_type</th>\n",
       "      <th>modencode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRX002599</td>\n",
       "      <td>SRR013488</td>\n",
       "      <td>modENCODE_1040</td>\n",
       "      <td>UTR</td>\n",
       "      <td>RACE</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRX002600</td>\n",
       "      <td>SRR013489</td>\n",
       "      <td>modENCODE_1040</td>\n",
       "      <td>UTR</td>\n",
       "      <td>RACE</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRX002600</td>\n",
       "      <td>SRR013490</td>\n",
       "      <td>modENCODE_1040</td>\n",
       "      <td>UTR</td>\n",
       "      <td>RACE</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRX002601</td>\n",
       "      <td>SRR013491</td>\n",
       "      <td>modENCODE_1040</td>\n",
       "      <td>UTR</td>\n",
       "      <td>RACE</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRX002602</td>\n",
       "      <td>SRR013492</td>\n",
       "      <td>modENCODE_1040</td>\n",
       "      <td>UTR</td>\n",
       "      <td>RACE</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         srx        srr    modENCODE_id modENCODE_assay modENCODE_type  \\\n",
       "0  SRX002599  SRR013488  modENCODE_1040             UTR           RACE   \n",
       "1  SRX002600  SRR013489  modENCODE_1040             UTR           RACE   \n",
       "2  SRX002600  SRR013490  modENCODE_1040             UTR           RACE   \n",
       "3  SRX002601  SRR013491  modENCODE_1040             UTR           RACE   \n",
       "4  SRX002602  SRR013492  modENCODE_1040             UTR           RACE   \n",
       "\n",
       "  modencode  \n",
       "0      True  \n",
       "1      True  \n",
       "2      True  \n",
       "3      True  \n",
       "4      True  "
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(len(modencode.modENCODE_id.unique()))\n",
    "modencode.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged2 = merged.merge(modencode, left_on='srx', right_on='srx', how='left')\n",
    "merged2.modencode = merged2.modencode.fillna(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4102, 27)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>srx</th>\n",
       "      <th>biosample</th>\n",
       "      <th>sra_selection</th>\n",
       "      <th>sra_source</th>\n",
       "      <th>sra_strategy</th>\n",
       "      <th>breed</th>\n",
       "      <th>cell_line</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>dev_stage</th>\n",
       "      <th>genotype</th>\n",
       "      <th>...</th>\n",
       "      <th>source_name</th>\n",
       "      <th>strain</th>\n",
       "      <th>tissue</th>\n",
       "      <th>treatment</th>\n",
       "      <th>srr</th>\n",
       "      <th>modENCODE_id</th>\n",
       "      <th>modENCODE_assay</th>\n",
       "      <th>modENCODE_type</th>\n",
       "      <th>modencode</th>\n",
       "      <th>isa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRX2172438</td>\n",
       "      <td>SAMN05784497</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>Drosophila S2 cells, dnaj1 KD</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRX2172437</td>\n",
       "      <td>SAMN05784495</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>Drosophila S2 cells, dnaj1 KD</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRX2172436</td>\n",
       "      <td>SAMN05784496</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>Drosophila S2 cells, mlf KD</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRX2172435</td>\n",
       "      <td>SAMN05784491</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>Drosophila S2 cells, mlf KD</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRX2172434</td>\n",
       "      <td>SAMN05784492</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>Drosophila S2 cells, lacZ KD</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          srx     biosample sra_selection      sra_source sra_strategy breed  \\\n",
       "0  SRX2172438  SAMN05784497          cDNA  TRANSCRIPTOMIC      RNA-Seq  None   \n",
       "1  SRX2172437  SAMN05784495          cDNA  TRANSCRIPTOMIC      RNA-Seq  None   \n",
       "2  SRX2172436  SAMN05784496          cDNA  TRANSCRIPTOMIC      RNA-Seq  None   \n",
       "3  SRX2172435  SAMN05784491          cDNA  TRANSCRIPTOMIC      RNA-Seq  None   \n",
       "4  SRX2172434  SAMN05784492          cDNA  TRANSCRIPTOMIC      RNA-Seq  None   \n",
       "\n",
       "  cell_line cell_type dev_stage genotype ...                     source_name  \\\n",
       "0      None      None      None     None ...   Drosophila S2 cells, dnaj1 KD   \n",
       "1      None      None      None     None ...   Drosophila S2 cells, dnaj1 KD   \n",
       "2      None      None      None     None ...     Drosophila S2 cells, mlf KD   \n",
       "3      None      None      None     None ...     Drosophila S2 cells, mlf KD   \n",
       "4      None      None      None     None ...    Drosophila S2 cells, lacZ KD   \n",
       "\n",
       "  strain tissue treatment  srr modENCODE_id modENCODE_assay modENCODE_type  \\\n",
       "0   None   None      None  NaN          NaN             NaN            NaN   \n",
       "1   None   None      None  NaN          NaN             NaN            NaN   \n",
       "2   None   None      None  NaN          NaN             NaN            NaN   \n",
       "3   None   None      None  NaN          NaN             NaN            NaN   \n",
       "4   None   None      None  NaN          NaN             NaN            NaN   \n",
       "\n",
       "  modencode  isa  \n",
       "0     False  NaN  \n",
       "1     False  NaN  \n",
       "2     False  NaN  \n",
       "3     False  NaN  \n",
       "4     False  NaN  \n",
       "\n",
       "[5 rows x 27 columns]"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(merged2.shape)\n",
    "merged2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3216"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(merged2.srx.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>modencode</th>\n",
       "      <th>False</th>\n",
       "      <th>True</th>\n",
       "      <th>All</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sra_strategy</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ChIP-Seq</th>\n",
       "      <td>568</td>\n",
       "      <td>62</td>\n",
       "      <td>630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DNase-Hypersensitivity</th>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EST</th>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MNase-Seq</th>\n",
       "      <td>104</td>\n",
       "      <td>0</td>\n",
       "      <td>104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OTHER</th>\n",
       "      <td>298</td>\n",
       "      <td>149</td>\n",
       "      <td>447</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>POOLCLONE</th>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RIP-Seq</th>\n",
       "      <td>82</td>\n",
       "      <td>0</td>\n",
       "      <td>82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNA-Seq</th>\n",
       "      <td>2384</td>\n",
       "      <td>349</td>\n",
       "      <td>2733</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WGS</th>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>miRNA-Seq</th>\n",
       "      <td>58</td>\n",
       "      <td>0</td>\n",
       "      <td>58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ncRNA-Seq</th>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>All</th>\n",
       "      <td>3529</td>\n",
       "      <td>573</td>\n",
       "      <td>4102</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "modencode               False  True   All\n",
       "sra_strategy                             \n",
       "ChIP-Seq                  568    62   630\n",
       "DNase-Hypersensitivity     14     0    14\n",
       "EST                         0    11    11\n",
       "MNase-Seq                 104     0   104\n",
       "OTHER                     298   149   447\n",
       "POOLCLONE                  12     0    12\n",
       "RIP-Seq                    82     0    82\n",
       "RNA-Seq                  2384   349  2733\n",
       "WGS                         4     2     6\n",
       "miRNA-Seq                  58     0    58\n",
       "ncRNA-Seq                   5     0     5\n",
       "All                      3529   573  4102"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.crosstab(merged2.sra_strategy, merged2.modencode, margins=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'_cls': 'Ncbi',\n",
       " '_id': 'SRX2581987',\n",
       " 'bioproject': {'bioproject_accn': 'PRJNA357269',\n",
       "  'bioproject_id': '357269',\n",
       "  'description': \"RNA molecules can attach to chromatin and thus provide a type of epigenomic information. It remains difficult to know what RNAs are associated with chromatin and where are the genomic target loci of these RNAs. Here, we present MARGI (Mapping RNA-genome interactions), a technology to massively reveal native RNA-chromatin interactions from unperturbed cells. The gist of this technology is to ligate chromatin-associated RNA (caRNA) with their target genomic sequence by proximity ligation, forming RNA-DNA chimeric sequences, which are converted to sequencing library for paired-end sequencing. Using MARGI, we produced RNA-genome interaction maps for human embryonic stem (ES) cells and HEK cells. MARGI revealed hundreds of chromatin-associated RNAs (caRNA), including previously known XIST, SNHG1, NEAT1, and MALAT1, as well as each caRNA's genomic interaction loci. Using a cross-species experiment, we estimated that approximately 2.2% of MARGI identified interactions are false positives. MARGI data suggested that 80-95% of interactions are proximal, where a caRNA is connected to its nearby genomic regions, 1-2% are distal, and 5-15% are inter-chromosomal. The majority of TSSs are associated with distal or inter-chromosomal caRNAs. ChIP-seq reported H3K27ac and H3K4me3 levels are positively correlated while H3K9me is negatively correlated with MARGI reported RNA attachment levels. The MARGI technology should facilitate revealing novel RNA functions and epigenomic events. Overall design: Two human cell lines were analysed: H9 and HEK. For each, we used two protocols, pxMARGI and diMARGI, having two replicates for the former and only one for the latter. Two samples from each replicate were sequenced. Whereas pxMARGI is designed to not differentiate passive and direct interactions, diMARGI is aimed to reveal protein tethered interactions. For the former, this was achieved by a combination of formaldehyde crosslinking and complete genome fragmentation, by overnight HaeIII digestion to ensure all genomic regions including heterochromatin were fragmented before the subsequent DNA ligation step. For the second, after crosslinking with formaldehyde and DSG, the chromatin was fragmented by sonication, and only the soluble fraction was passed onto the subsequent ligation steps.\",\n",
       "  'external_id': [],\n",
       "  'last_update': datetime.datetime(2017, 1, 26, 0, 0),\n",
       "  'name': 'Systematic mapping of RNA-chromatin interactions in vivo',\n",
       "  'submission_date': datetime.datetime(2016, 12, 13, 0, 0),\n",
       "  'submission_id': 'SUB2354705',\n",
       "  'title': 'Systematic mapping of RNA-chromatin interactions in vivo'},\n",
       " 'biosample': [{'GEO': 'GSM2427901',\n",
       "   'access': 'public',\n",
       "   'attributes': [{'name': 'source_name', 'value': 'HEK and DM cells'},\n",
       "    {'name': 'cell_line', 'value': 'HEK'},\n",
       "    {'name': 'cell_line', 'value': 'DM'},\n",
       "    {'name': 'protocol', 'value': 'pxMARGI'}],\n",
       "   'biosample_accn': 'SAMN06140184',\n",
       "   'biosample_id': '6140184',\n",
       "   'biosample_primary': 'SAMN06140184',\n",
       "   'contacts': [{'email': 'marcelorivas@gmail.com',\n",
       "     'first_name': 'Marcelo',\n",
       "     'last_name': 'Rivas'}],\n",
       "   'institute': 'Sheng Zhong, Bioengineering, University of California San Diego',\n",
       "   'last_update': '2017-02-22',\n",
       "   'models': ['Generic'],\n",
       "   'organism_name': 'Drosophila melanogaster',\n",
       "   'publication_date': '2017-01-27',\n",
       "   'sample_id': 'SRS1996012',\n",
       "   'submission_date': '2016-12-13',\n",
       "   'tax_id': '7227',\n",
       "   'tax_name': 'Drosophila melanogaster',\n",
       "   'title': 'HEK-dm_pxMARGI_rep1'}],\n",
       " 'sra': {'db_flags': ['RNASeq', 'PE'],\n",
       "  'db_imported': datetime.datetime(2017, 3, 1, 11, 59, 41, 213000),\n",
       "  'experiment': {'GEO_Dataset': '302427901',\n",
       "   'attributes': [{'name': 'GEO Accession', 'value': 'GSM2427901'}],\n",
       "   'ddbj_links': [],\n",
       "   'ena_links': [],\n",
       "   'entrez_links': [],\n",
       "   'experiment_id': 'SRX2581987',\n",
       "   'external_id': [],\n",
       "   'instrument_model': 'Illumina MiniSeq',\n",
       "   'library_construction_protocol': 'The major steps of MARGI technologies are: crosslinking, isolation of nuclear material, protein biotinylation, genome fragmentation, RNA fragmentation, stabilization of proteins on streptavidin beads, end modification and RNA ligation, DNA ligation, reverse crosslinking, purification of RNA and DNA, removal of biotin from non-ligated linkers, pulldown of ligation products, reverse transcription, DNA denaturation, circularization and re-linearization, generation of sequencing library, and paired-end sequencing. See Extended Materials and Methods in published paper for details. Sequencing library was generated by amplifying material using NEBnext Hi-fidelity PCR enzyme (New England Biolabs M0541S) with NEBnext multiplex oligos for Illumina sequencing (New England Biolabs E7335S) as the primers. Multiple test PCRs with 1 μL of sample and 10 μL reaction volume were run to identify the correct cycle number for the production run. The production PCR was run in a 50 μL reaction volume with 5 μL sample. After the PCR, size selection was performed using Ampure XP beads (Beckman Coulter A63880). Library between 200-2000 bp was selected. Library was run on a Bioanalyzer to determine the size distribution and then paired-end sequenced for 100 cycles at each end on Illumina HiSeq 2500 (PE100) or MiniSeq sequencers (PE150).',\n",
       "   'library_layout': 'PAIRED',\n",
       "   'library_selection': 'other',\n",
       "   'library_source': 'TRANSCRIPTOMIC',\n",
       "   'library_strategy': 'OTHER',\n",
       "   'platform': 'ILLUMINA',\n",
       "   'secondary_id': [],\n",
       "   'study_id': 'SRP100516',\n",
       "   'submitter_id': [{'db': 'GEO', 'id': 'GSM2427901'}],\n",
       "   'title': 'GSM2427901: HEK-dm_pxMARGI_rep1; Drosophila melanogaster; Homo sapiens; OTHER',\n",
       "   'url_links': [],\n",
       "   'uuid': [],\n",
       "   'xref_links': []},\n",
       "  'organization': {'abbreviation': 'GEO',\n",
       "   'email': 'geo-group@ncbi.nlm.nih.gov',\n",
       "   'first_name': 'Geo',\n",
       "   'last_name': 'Curators',\n",
       "   'name': 'NCBI',\n",
       "   'organization_type': 'center'},\n",
       "  'pool': ['SRS1996012'],\n",
       "  'run': [{'download_path': 'https://sra-download.ncbi.nlm.nih.gov/srapub/SRR5278103',\n",
       "    'experiment_id': 'SRX2581987',\n",
       "    'external_id': [],\n",
       "    'load_date': '2017-02-22',\n",
       "    'nbases': 4996041946,\n",
       "    'nreads': 2,\n",
       "    'nspots': 17428278,\n",
       "    'read_count_r1': 17428278.0,\n",
       "    'read_count_r2': 17428278.0,\n",
       "    'read_len_r1': 142.97,\n",
       "    'read_len_r2': 143.69,\n",
       "    'release_date': '2017-02-22',\n",
       "    'run_flags': ['PE'],\n",
       "    'run_id': 'SRR5278103',\n",
       "    'samples': ['SRS1996012'],\n",
       "    'secondary_id': [],\n",
       "    'size_MB': 1928,\n",
       "    'submitter_id': [{'db': 'GEO', 'id': 'GSM2427901_r1'}],\n",
       "    'uuid': []}],\n",
       "  'sample': {'BioProject': '357269',\n",
       "   'BioSample': 'SAMN06140184',\n",
       "   'GEO': 'GSM2427901',\n",
       "   'attributes': [{'name': 'source_name', 'value': 'HEK and DM cells'},\n",
       "    {'name': 'cell line', 'value': 'HEK'},\n",
       "    {'name': 'cell line', 'value': 'DM'},\n",
       "    {'name': 'protocol', 'value': 'pxMARGI'}],\n",
       "   'ddbj_links': [],\n",
       "   'ena_links': [],\n",
       "   'entrez_links': [],\n",
       "   'external_id': [],\n",
       "   'sample_id': 'SRS1996012',\n",
       "   'scientific_name': 'Drosophila melanogaster',\n",
       "   'secondary_id': [],\n",
       "   'submitter_id': [],\n",
       "   'taxon_id': '7227',\n",
       "   'title': 'HEK-dm_pxMARGI_rep1',\n",
       "   'url_links': [],\n",
       "   'uuid': [],\n",
       "   'xref_links': []},\n",
       "  'study': {'BioProject': 'PRJNA357269',\n",
       "   'GEO': 'GSE92345',\n",
       "   'abstract': \"RNA molecules can attach to chromatin and thus provide a type of epigenomic information. It remains difficult to know what RNAs are associated with chromatin and where are the genomic target loci of these RNAs. Here, we present MARGI (Mapping RNA-genome interactions), a technology to massively reveal native RNA-chromatin interactions from unperturbed cells. The gist of this technology is to ligate chromatin-associated RNA (caRNA) with their target genomic sequence by proximity ligation, forming RNA-DNA chimeric sequences, which are converted to sequencing library for paired-end sequencing. Using MARGI, we produced RNA-genome interaction maps for human embryonic stem (ES) cells and HEK cells. MARGI revealed hundreds of chromatin-associated RNAs (caRNA), including previously known XIST, SNHG1, NEAT1, and MALAT1, as well as each caRNA's genomic interaction loci. Using a cross-species experiment, we estimated that approximately 2.2% of MARGI identified interactions are false positives. MARGI data suggested that 80-95% of interactions are proximal, where a caRNA is connected to its nearby genomic regions, 1-2% are distal, and 5-15% are inter-chromosomal. The majority of TSSs are associated with distal or inter-chromosomal caRNAs. ChIP-seq reported H3K27ac and H3K4me3 levels are positively correlated while H3K9me is negatively correlated with MARGI reported RNA attachment levels. The MARGI technology should facilitate revealing novel RNA functions and epigenomic events. Overall design: Two human cell lines were analysed: H9 and HEK. For each, we used two protocols, pxMARGI and diMARGI, having two replicates for the former and only one for the latter. Two samples from each replicate were sequenced. Whereas pxMARGI is designed to not differentiate passive and direct interactions, diMARGI is aimed to reveal protein tethered interactions. For the former, this was achieved by a combination of formaldehyde crosslinking and complete genome fragmentation, by overnight HaeIII digestion to ensure all genomic regions including heterochromatin were fragmented before the subsequent DNA ligation step. For the second, after crosslinking with formaldehyde and DSG, the chromatin was fragmented by sonication, and only the soluble fraction was passed onto the subsequent ligation steps.\",\n",
       "   'center_name': 'GEO',\n",
       "   'center_project_name': 'GSE92345',\n",
       "   'ddbj_links': [],\n",
       "   'ena_links': [],\n",
       "   'entrez_links': [],\n",
       "   'external_id': [],\n",
       "   'related_studies': [],\n",
       "   'secondary_id': [],\n",
       "   'study_id': 'SRP100516',\n",
       "   'study_type': 'Other',\n",
       "   'submitter_id': [],\n",
       "   'title': 'Systematic mapping of RNA-chromatin interactions in vivo',\n",
       "   'url_links': [],\n",
       "   'uuid': [],\n",
       "   'xref_links': []},\n",
       "  'submission': {'broker': 'GEO',\n",
       "   'external_id': [],\n",
       "   'secondary_id': [],\n",
       "   'submission_id': 'SRA539405',\n",
       "   'submitter_id': [{'db': 'GEO', 'id': 'GEO: GSE92345'}],\n",
       "   'uuid': []}}}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi.find_one()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:ncbi_remap]",
   "language": "python",
   "name": "conda-env-ncbi_remap-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.3"
  },
  "toc": {
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": "block",
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
